---
title: "SlimeMold_SpatialTranscriptomics"
author: "Tobias Gerber"
date: "September 9, 2019"
output: html_document
---

##96 well format


```{r}
library(Seurat)

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/")

input.dataRaw <- read.delim("/mnt/SingleCellGenomics/scg_projects/SlimeMold/SpatialTranscriptomics/SlimeMold_SmartSeq2.SpatialTranscriptomics.tpm.final", header=T)

#log transform
tmp<-input.dataRaw[,c("TPM")]
tmp[tmp<1]<-NA
tmp<-log2(tmp)
tmp[is.na(tmp)]<-0
rm(input.dataRaw_log2)
input.dataRaw_log2<-input.dataRaw
input.dataRaw_log2$log2<-tmp
rm(tmp)

#cast
library(reshape)

data.reshape<-input.dataRaw_log2[,c("cellID","gene","log2")]
data <- as.data.frame(cast(data.reshape,cellID ~ gene,mean))
rm(data.reshape)

#add rownames
rownames(data) = data[,"cellID"]
data[,"cellID"] = NULL

#split based on experiment

number2 = data[rownames(data) %in% rownames(data)[grep("\\dA",rownames(data))],]

tmp = data[rownames(data) %in% rownames(data)[grep("\\dB",rownames(data))],]

star.death.cells = rownames(tmp)[grep("1B",rownames(tmp))]
star.death.cells = c(star.death.cells,rownames(tmp)[grep("12",rownames(tmp))])
star.death.cells = c(star.death.cells,rownames(tmp)[grep("8",rownames(tmp))])
star.death.cells = c(star.death.cells,rownames(tmp)[grep("9",rownames(tmp))])
star.death.cells = c(star.death.cells,rownames(tmp)[grep("10",rownames(tmp))])

star.death = tmp[star.death.cells,]

flower = tmp[!rownames(tmp) %in% rownames(star.death),]

#star death well IDs were shifted on the plate assign original well ID

rownames(star.death) = gsub("8","2",rownames(star.death))
rownames(star.death) = gsub("9","3",rownames(star.death))
rownames(star.death) = gsub("10","4",rownames(star.death))
rownames(star.death) = gsub("11","5",rownames(star.death))
rownames(star.death) = gsub("12","6",rownames(star.death))

#assign experiment ID to cell names and remove index A and B information


rownames(star.death) = gsub(pattern="(.*\\d).*", replacement="\\1",rownames(star.death),perl = T)
rownames(flower) = gsub(pattern="(.*\\d).*", replacement="\\1",rownames(flower),perl = T)
rownames(number2) = gsub(pattern="(.*\\d).*", replacement="\\1",rownames(number2),perl = T)

rownames(flower) = paste(rownames(flower),"flower",sep = "_")
rownames(star.death) = paste(rownames(star.death),"stardeath",sep = "_")
rownames(number2) = paste(rownames(number2),"number2",sep = "_")


#transpose
flower = as.matrix(t(flower))
star.death = as.matrix(t(star.death))
number2 = as.matrix(t(number2))

#save
saveRDS(flower,"data_flower_SS2_SpatialTranscriptomics.RDS")
saveRDS(star.death,"data_stardeath_SS2_SpatialTranscriptomics.RDS")
saveRDS(number2,"data_number2_SS2_SpatialTranscriptomics.RDS")


```

##Analyze 96 well data set

```{r}
library(Seurat) #version 3
library(Matrix)
library(tidyverse)
library(RColorBrewer)
library(ggplot2)
library(RANN)
library(stats)
library(BiocManager)
library(cowplot)
library(readxl)
library(viridis)
library(ComplexHeatmap)
library(circlize)
library(statmod)
library(Rtsne)
library(cluster)
library(factoextra)
library(dendextend)
library(Matrix.utils)

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/")
source("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/r-Analysis-TG/r-Analysis-TG/CustomFunctions.R")

# Load RDS raw matrix
data_StarDeath = readRDS("./data_stardeath_SS2_SpatialTranscriptomics.RDS")
data_NumberS = readRDS("./data_number2_SS2_SpatialTranscriptomics.RDS")
data_Flower = readRDS("./data_flower_SS2_SpatialTranscriptomics.RDS")

x <- merge.Matrix(data_StarDeath, data_NumberS,
                     by.x = rownames(data_StarDeath), by.y = rownames(data_NumberS), 
                     type = "full")
x <- merge.Matrix(x, data_Flower,
                     by.x = rownames(x), by.y = rownames(data_Flower), 
                     type = "full")

# 1- DATA PREPARATION
  # Prepare MetaData file
  MD_StarDeath <- read_excel("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/SpatialTranscriptomics.xlsx", sheet = "StarDeath", col_names = TRUE)
  MD_NumberS <- read_excel("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/SpatialTranscriptomics.xlsx", sheet = "NumberS", col_names = TRUE)
  MD_Flower <- read_excel("//home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/SpatialTranscriptomics.xlsx", sheet = "Flower", col_names = TRUE)
  MD_All <- rbind(MD_StarDeath, MD_NumberS, MD_Flower)
  MD_All <- MD_All %>% mutate(setID = paste0(MD_All$Position, "_", MD_All$Batch))

  QC.Data <- data.frame(wells = sapply(strsplit(colnames(x), "_"), "[", 1), 
                      nUMI = colSums(x), 
                      nGene = colSums(x!=0), 
                      origin = sapply(strsplit(colnames(x), "_"), "[", 2), 
                      ID = colnames(x))
  
  MD_All <- MD_All %>% left_join(QC.Data, by = c("setID"="ID"))
  rm(QC.Data, MD_Flower, MD_NumberS, MD_StarDeath)
  
  # Prepare Grid
  MD_OatFlakes <- read_excel("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/SpatialTranscriptomics.xlsx", sheet = "OatFlakesCoordinates", col_names = TRUE)
  
  grid_All <- expand.grid(row=c("A","B","C","D","E","F","G","H","I","J","K","L","M","N","O","P",
                                "Q","R","S","T","U","V","W","X","Z","AA","BB","CC","DD","EE",
                                "FF","GG","HH","II","JJ","KK","LL","MM","NN","OO","PP","QQ","RR",
                                "SS","TT","AAA","BBB","CCC","DDD","EEE","FFF","GGG","HHH","III",
                                "JJJ","KKK","LLL","MMM","NNN","OOO","PPP","QQQ","RRR","SSS","TTT"), 
                              column=c(1:26)) %>% mutate(well=paste0(row,column)) 
  
  grid_All = merge(grid_All,as.data.frame(MD_All,stringsAsFactors = F), by.x = "well", by.y = "Coordinates")
  
  grid_All <- grid_All %>% left_join(MD_All, by = c(well=Coordinates))
  grid_All$Piece <- ifelse(grid_All$Position!=is.na(grid_All$Position), 1, 0)
  grid_All$Piece[is.na(grid_All$Piece)] <- 0
  
  rm(data_Flower, data_NumberS, data_StarDeath)

```

```{r}
#BSS4 = readRDS("BSS4_final_data.RDS")
#BSS4 = as.data.frame(t(BSS4),stringsAsFactors = F)
#BSS4.meta = readRDS("BSS4_final_meta.RDS")


#Get annotated transcripts

anno = read.delim("/mnt/SingleCellGenomics/genome_data/10xSlimeMoldTranscriptome/41598_2017_12250_MOESM2_ESM.txt.csv",sep = "\t",header = T)
anno = anno[1:49403,]
colnames(anno) = c("transcript","ORF_length","database","ID","description","from","to","IPR_number","Domain_description","GOterm","Log2_FoldChange","pvalue","baseMean","IfcSE","stat","padj","UniProt_Name","UniProt_ID","UniProt_Acc","Regulation")

anno.gene.id = unique(as.character(anno$transcript))

anno$transcript = gsub(x = anno$transcript,pattern = "_",replacement = "-")

anno.gene = anno[,c("transcript","UniProt_Acc")]
anno.gene = unique(anno.gene)
anno.gene$transcript = as.character(anno.gene$transcript)
anno.gene$UniProt_Acc = as.character(anno.gene$UniProt_Acc)
rownames(anno.gene) = anno.gene$transcript
anno.gene$UniProt_Acc[is.na(anno.gene$UniProt_Acc)] = "noGene"

#Get GO information
anno.GO = anno[,c("transcript","GOterm","ORF_length","description")]

#remove rows without GO assignment
emptyfields <- rowSums(anno.GO == "") == 1
anno.GO <- anno.GO[!emptyfields,]
anno.GO = unique(anno.GO)

anno.GO$transcript = as.character(anno.GO$transcript)
anno.GO$GOterm = as.character(anno.GO$GOterm)
anno.GO$ORF_length = as.numeric(anno.GO$ORF_length)

id_list_GO = vector("list")
id_list_Domain = vector("list")
ORF_list = vector("list")

for (i in 1:nrow(anno.GO)) {
  k = anno.GO[i,1]
  if (k %in% names(id_list_GO)) {
    if (anno.GO[i,3] == ORF_list[[k]]) {
      ORF_list[[k]] = anno.GO[i,3]
      id_list_GO[[k]] = paste(id_list_GO[[k]],anno.GO[i,2],sep = "|")
      if (id_list_Domain[[k]] == anno.GO[i,4]) {
        next
      } else {
        id_list_Domain[[k]] = paste(id_list_Domain[[k]],anno.GO[i,4],sep = "|")
      }
    } else {
      if (anno.GO[i,3] > ORF_list[[k]]) {
        ORF_list[[k]] = anno.GO[i,3]
        id_list_GO[[k]] = anno.GO[i,2]
        id_list_Domain[[k]] = anno.GO[i,4]
      } else {
        next
      }
    }
  } else {
    id_list_GO[[k]] = anno.GO[i,2]
    ORF_list[[k]] = anno.GO[i,3]
    id_list_Domain[[k]] = anno.GO[i,4]
  }
}

anno.GO = data.frame(unlist(id_list_GO,use.names = TRUE),stringsAsFactors=F)
anno.GO$transcript = rownames(anno.GO)
colnames(anno.GO) = c("GOterm","transcript")

anno.Domain = data.frame(unlist(id_list_Domain,use.names = TRUE),stringsAsFactors=F)
anno.Domain$transcript = rownames(anno.Domain)
colnames(anno.Domain) = c("Domain","transcript")

anno.gene = merge(anno.gene,anno.GO, by = "transcript")
anno.gene = merge(anno.gene,anno.Domain, by = "transcript")

############# get data and remove genes

data96_All = x[rownames(x) %in% anno.gene.id,]

############# load in Seurat

data96_All <- CreateSeuratObject(counts = data96_All, project = "SpatialTrans96well", min.cells = 2, min.features = 10)
data96_All

############# load meta data into Seurat

data96_All_meta = as.data.frame(MD_All,stringsAsFactor = F)
data96_All_meta = data96_All_meta[match(colnames(data96_All),data96_All_meta$setID),]

dim = strsplit(data96_All_meta$Coordinates, "(?<=[^\\d])", perl = TRUE)
dim <- data.frame(matrix(unlist(dim), nrow=length(dim), byrow=TRUE),stringsAsFactors=FALSE)
rownames(dim) = rownames(data96_All_meta)
colnames(dim) = c("Spatial_1","Spatial_2")

myLetters <- toupper(letters[1:26])
dim$Spatial_1 = as.numeric(match(dim$Spatial_1, myLetters))
dim$Spatial_2 = as.numeric(dim$Spatial_2)


data96_All@meta.data$Spatial_1 = dim$Spatial_1
data96_All@meta.data$Spatial_2 = dim$Spatial_2

data96_All@meta.data$orig.ident = c(rep("StarDeath",48),rep("Number2",80),rep("Flower",48))

Idents(data96_All) <- data96_All@meta.data$orig.ident

StarDeath <- subset(data96_All, idents = c("StarDeath"))
Number2 <- subset(data96_All, idents = c("Number2"))
Flower <- subset(data96_All, idents = c("Flower"))

reads_number2 = read.delim("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/RawReadsBAM_Number2.csv", header = F,sep = ";")
colnames(reads_number2) = c("RawReads","ID")
rownames(reads_number2) = reads_number2$ID
rownames(reads_number2) = paste(rownames(reads_number2),"number2", sep = "_")
reads_number2 = reads_number2[match(colnames(Number2),rownames(reads_number2)),]

Number2@meta.data$reads = reads_number2$RawReads

reads_stardeath = read.delim("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/RawReadsBAM_StarDeath.csv", header = F,sep = ";")
colnames(reads_stardeath) = c("RawReads","ID")
rownames(reads_stardeath) = reads_stardeath$ID
rownames(reads_stardeath) = paste(rownames(reads_stardeath),"stardeath", sep = "_")
reads_stardeath = reads_stardeath[match(colnames(StarDeath),rownames(reads_stardeath)),]

StarDeath@meta.data$reads = reads_stardeath$RawReads


reads_flower = read.delim("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/RawReadsBAM_Flower.csv", header = F,sep = ";")
colnames(reads_flower) = c("RawReads","ID")
rownames(reads_flower) = reads_flower$ID
rownames(reads_flower) = paste(rownames(reads_flower),"flower", sep = "_")
reads_flower = reads_flower[match(colnames(Flower),rownames(reads_flower)),]

Flower@meta.data$reads = reads_flower$RawReads

```

#Star Death

```{r}


#parallize scaling
library(future)
plan("multiprocess", workers = 8)
plan()

#set maximum to 10GB for each worker
options(future.globals.maxSize = 10000 * 1024^2)

all.genes <- rownames(StarDeath)
StarDeath = ScaleData(  StarDeath, features = all.genes ,vars.to.regress = c("reads","nFeature_RNA"))

#run PCA on all genes
#RunPCA(object, assay = NULL, features = NULL,  npcs = 50, rev.pca = FALSE, weight.by.var = TRUE, verbose = TRUE,  ndims.print = 1:5, nfeatures.print = 30, reduction.name = "pca",  reduction.key = "PC_", seed.use = 42)

StarDeath = RunPCA(StarDeath,features = all.genes,npcs = 20)

#plot PCA heatmaps
#DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

library(RColorBrewer)
library(ggplot2)

pdf("StarDeath_PCAheatmaps.pdf",width=30,height=20)
DimHeatmap(StarDeath, dims = 1:16, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
dev.off()

#elbow plot
pdf("StarDeath_PCelbow.pdf",width=20,height=10)
ElbowPlot(StarDeath, ndims = 20)
dev.off()

StarDeath[["spatial"]] <- CreateDimReducObject(embeddings = as.matrix(StarDeath@meta.data[,c("Spatial_1","Spatial_2")]),  key = "Spatial_", assay = "RNA")

#Cluster cells
StarDeath<- FindNeighbors(StarDeath, dims = 1:10)
StarDeath <- FindClusters(StarDeath, resolution = 0.6)


# Perform Hierarchical Clustering
library(tidyverse)
Data.hc <- StarDeath@reductions$pca@cell.embeddings[,c(1:10)] %>% get_dist(method = "pearson") %>% hclust(method = 'ward.D') ## PCs
Hcls <- data.frame(Clusters = as.factor(cutree(Data.hc, k = 3)))
Hcls <- Hcls %>% mutate(Wells = colnames(StarDeath))

StarDeath@meta.data$Hclust = Hcls$Clusters

StarDeath = RunUMAP(StarDeath, dims = 1:10,min.dist = 0.03)

#plot result
#DimPlot(object, dims = c(1, 2), cells = NULL, cols = NULL,  pt.size = NULL, reduction = NULL, group.by = NULL,  split.by = NULL, shape.by = NULL, order = NULL, label = FALSE,  label.size = 4, repel = FALSE, cells.highlight = NULL,  cols.highlight = "red", sizes.highlight = 1, na.value = "grey50",  combine = TRUE)

pdf("StarDeath_Embedding_Grid_PC10_res0.6.pdf",width=7,height=6.5)
DimPlot(object = StarDeath, reduction = 'spatial', pt.size = 5.5, shape.by = "orig.ident") + scale_shape_manual(values = c("StarDeath" = 15))
DimPlot(object = StarDeath, reduction = 'spatial', pt.size = 5.5, shape.by = "orig.ident",group.by = "Hclust") + scale_shape_manual(values = c("StarDeath" = 15))
DimPlot(object = StarDeath, reduction = 'umap', pt.size = 5.5, shape.by = "orig.ident") + scale_shape_manual(values = c("StarDeath" = 15))
DimPlot(object = StarDeath, reduction = 'umap', pt.size = 5.5, shape.by = "orig.ident",group.by = "Hclust")  + scale_shape_manual(values = c("StarDeath" = 15)) 
dev.off()

#some QC features

pdf("StarDeath_feature_QC.pdf",width=10,height=10)
FeaturePlot(StarDeath, reduction = 'spatial', shape.by = "orig.ident", pt.size = 2, features = c("nFeature_RNA","nCount_RNA","reads"),order = F, cols = c("grey",rev(brewer.pal(11,"RdYlBu")[1:11]))) + scale_shape_manual(values = c("StarDeath" = 15))
dev.off()

Idents(object = StarDeath) <- StarDeath@meta.data$Hclust

plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(StarDeath, only.pos = TRUE,  logfc.threshold = 0.2)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(cluster , desc(avg_logFC))

write.csv(markers.anno,"StarDeath_Grid_allMarker.csv")

saveRDS(StarDeath, file = "StarDeath_Grid_SeuratObj.RDS")

				
gene_names = c(
"TOP2_DICDI",
"KI67_HUMAN",
"DPOLN_MOUSE",
"DNLI_HAEIN",
"TOLB_PSEA6",
"WDR54_MOUSE",
"ACTN4_RAT",
"CDC20_DICDI",
"DPOLB_RAT",
"RHOQ_HUMAN",
"FORD_DICDI",
"RASA2_MOUSE",
"CDKA1_ORYSJ",
"CCA11_ORYSJ",
"CCNB_DICDI",
"TBB2_PHYPO",
"MCM9_XENTR",
"PATS1_DICDI",
"CHX17_ARATH",
"APRX_BACSU")
gene_ids = c(
"Phypoly-transcript-00381",
"Phypoly-transcript-00261",
"Phypoly-transcript-01132",
"Phypoly-transcript-14390",
"Phypoly-transcript-04660",
"Phypoly-transcript-13888",
"Phypoly-transcript-00407",
"Phypoly-transcript-07151",
"Phypoly-transcript-04465",
"Phypoly-transcript-23095",
"Phypoly-transcript-02192",
"Phypoly-transcript-06376",
"Phypoly-transcript-06666",
"Phypoly-transcript-05693",
"Phypoly-transcript-08256",
"Phypoly-transcript-06678",
"Phypoly-transcript-00963",
"Phypoly-transcript-00502",
"Phypoly-transcript-13802",
"Phypoly-transcript-04762")


gg_Fig <- FeaturePlot(StarDeath, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("StarDeath" = 15)) })
	

pdf("StarDeath_Grid_feature_BSS1Marker.pdf",width=20.5,height=14)
CombinePlots( gg_Fig )
dev.off()
	

gene_names = c("HSE1_MAGO7",
"CBPB_DICDI",
"AVIL_HUMAN",
"SCP1_ASTLP",
"LIMD2_BOVIN",
"CYAA_ANACY",
"ALKB_ECOLI",
"GTAG_DICDI",
"XPR1_MUSMC",
"ABCGL_DICDI",
"RAB32_MOUSE",
"UFD1_DROME",
"MFSD8_DANRE",
"SPXS4_DICDI",
"CP3A6_RABIT",
"ARUS_PSEAE",
"PLAS_SYNY3",
"RACH_DICDI")
gene_ids = c("Phypoly-transcript-06951",
"Phypoly-transcript-16629",
"Phypoly-transcript-13968",
"Phypoly-transcript-17763",
"Phypoly-transcript-02119",
"Phypoly-transcript-01311",
"Phypoly-transcript-17043",
"Phypoly-transcript-09221",
"Phypoly-transcript-22816",
"Phypoly-transcript-16211",
"Phypoly-transcript-13603",
"Phypoly-transcript-23641",
"Phypoly-transcript-04492",
"Phypoly-transcript-02158",
"Phypoly-transcript-11480",
"Phypoly-transcript-04786",
"Phypoly-transcript-17808",
"Phypoly-transcript-16297")


gg_Fig <- FeaturePlot(StarDeath, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("StarDeath" = 15)) })
	

pdf("StarDeath_Grid_feature_BSS4Marker.pdf",width=20.5,height=14)
CombinePlots( gg_Fig )
dev.off()
	

#some photoreceptors
pdf("StarDeath_Grid_feature_PhotoReceptorsMarker.pdf",width=10,height=10)
FeaturePlot(StarDeath, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = c("Phypoly-transcript-25848" ,"Phypoly-transcript-24665" ,"Phypoly-transcript-20950", "Phypoly-transcript-16352", "Phypoly-transcript-16339" ,"Phypoly-transcript-13173", "Phypoly-transcript-12963", "Phypoly-transcript-11651"),order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
dev.off()

gene_names = c("VGA_BPS13",
"ZC3H3_HUMAN",
"RAC5_ORYSJ",
"GEFX_DICDI",
"POED2_ARATH",
"POLK_SCHPO",
"DNLI_HAEIN",
"EHD4_HUMAN",
"DPNP_DICDI",
"YR665_MIMIV",
"ANK1_HUMAN",
"BRE1_DICDI",
"GRLE_DICDI",
"PTP2_DICDI",
"MFSD8_HUMAN",
"RACH_DICDI")
gene_ids = c("Phypoly-transcript-00051",
"Phypoly-transcript-06479",
"Phypoly-transcript-14244",
"Phypoly-transcript-02579",
"Phypoly-transcript-15917",
"Phypoly-transcript-09316",
"Phypoly-transcript-02851",
"Phypoly-transcript-06523",
"Phypoly-transcript-06033",
"Phypoly-transcript-18282",
"Phypoly-transcript-02481",
"Phypoly-transcript-02710",
"Phypoly-transcript-11890",
"Phypoly-transcript-06659",
"Phypoly-transcript-07556",
"Phypoly-transcript-16297")


gg_Fig <- FeaturePlot(StarDeath, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("StarDeath" = 15)) })
	

pdf("StarDeath_Grid_feature_Top4ClustMarkerANDSclerotiumANDFanMarker.pdf",width=20.5,height=14.5)
CombinePlots( gg_Fig )
dev.off()
	

```

##Number2

```{r}


#parallize scaling
library(future)
plan("multiprocess", workers = 8)
plan()

#set maximum to 10GB for each worker
options(future.globals.maxSize = 10000 * 1024^2)

all.genes <- rownames(Number2)
Number2 = ScaleData(  Number2, features = all.genes ,vars.to.regress = c("reads","nFeature_RNA"))

#run PCA on all genes
#RunPCA(object, assay = NULL, features = NULL,  npcs = 50, rev.pca = FALSE, weight.by.var = TRUE, verbose = TRUE,  ndims.print = 1:5, nfeatures.print = 30, reduction.name = "pca",  reduction.key = "PC_", seed.use = 42)

Number2 = RunPCA(Number2,features = all.genes,npcs = 20)

#plot PCA heatmaps
#DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

library(RColorBrewer)
library(ggplot2)

pdf("Number2_PCAheatmaps.pdf",width=30,height=20)
DimHeatmap(Number2, dims = 1:16, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
dev.off()

#elbow plot
pdf("Number2_PCelbow.pdf",width=20,height=10)
ElbowPlot(Number2, ndims = 20)
dev.off()

Number2[["spatial"]] <- CreateDimReducObject(embeddings = as.matrix(Number2@meta.data[,c("Spatial_1","Spatial_2")]),  key = "Spatial_", assay = "RNA")

#Cluster cells
Number2<- FindNeighbors(Number2, dims = 1:10)
Number2 <- FindClusters(Number2, resolution = 0.6)


# Perform Hierarchical Clustering
library(tidyverse)
Data.hc <- Number2@reductions$pca@cell.embeddings[,c(1:10)] %>% get_dist(method = "pearson") %>% hclust(method = 'ward.D') ## PCs
Hcls <- data.frame(Clusters = as.factor(cutree(Data.hc, k = 3)))
Hcls <- Hcls %>% mutate(Wells = colnames(Number2))

Number2@meta.data$Hclust = Hcls$Clusters

Number2 = RunUMAP(Number2, dims = 1:10,min.dist = 0.03)

#plot result
#DimPlot(object, dims = c(1, 2), cells = NULL, cols = NULL,  pt.size = NULL, reduction = NULL, group.by = NULL,  split.by = NULL, shape.by = NULL, order = NULL, label = FALSE,  label.size = 4, repel = FALSE, cells.highlight = NULL,  cols.highlight = "red", sizes.highlight = 1, na.value = "grey50",  combine = TRUE)

pdf("Number2_Embedding_Grid_PC10_res0.6.pdf",width=6.4,height=6.2)
DimPlot(object = Number2, reduction = 'spatial', pt.size = 5.5, shape.by = "orig.ident") + scale_shape_manual(values = c("Number2" = 15))
DimPlot(object = Number2, reduction = 'spatial', pt.size = 5.5, shape.by = "orig.ident",group.by = "Hclust") + scale_shape_manual(values = c("Number2" = 15))
DimPlot(object = Number2, reduction = 'umap', pt.size = 5.5, shape.by = "orig.ident") + scale_shape_manual(values = c("Number2" = 15))
DimPlot(object = Number2, reduction = 'umap', pt.size = 5.5, shape.by = "orig.ident",group.by = "Hclust")  + scale_shape_manual(values = c("Number2" = 15)) 
dev.off()

#some QC features

pdf("Number2_feature_QC.pdf",width=10,height=10)
FeaturePlot(Number2, reduction = 'spatial', shape.by = "orig.ident", pt.size = 2, features = c("nFeature_RNA","nCount_RNA","reads"),order = F, cols = c("grey",rev(brewer.pal(11,"RdYlBu")[1:11]))) + scale_shape_manual(values = c("Number2" = 15))
dev.off()

#Idents(object = Number2) <- Number2@meta.data$Hclust

plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(Number2, only.pos = TRUE,  logfc.threshold = 0.2)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(cluster , desc(avg_logFC))

write.csv(markers.anno,"Number2_Grid_allMarker.csv")

saveRDS(Number2, file = "Number2_Grid_SeuratObj.RDS")

				
gene_names = c(
"TOP2_DICDI",
"KI67_HUMAN",
"DPOLN_MOUSE",
"DNLI_HAEIN",
"TOLB_PSEA6",
"WDR54_MOUSE",
"ACTN4_RAT",
"CDC20_DICDI",
"DPOLB_RAT",
"RHOQ_HUMAN",
"FORD_DICDI",
"RASA2_MOUSE",
"CDKA1_ORYSJ",
"CCA11_ORYSJ",
"CCNB_DICDI",
"TBB2_PHYPO",
"MCM9_XENTR",
"PATS1_DICDI")
gene_ids = c(
"Phypoly-transcript-00381",
"Phypoly-transcript-00261",
"Phypoly-transcript-01132",
"Phypoly-transcript-14390",
"Phypoly-transcript-04660",
"Phypoly-transcript-13888",
"Phypoly-transcript-00407",
"Phypoly-transcript-07151",
"Phypoly-transcript-04465",
"Phypoly-transcript-23095",
"Phypoly-transcript-02192",
"Phypoly-transcript-06376",
"Phypoly-transcript-06666",
"Phypoly-transcript-05693",
"Phypoly-transcript-08256",
"Phypoly-transcript-06678",
"Phypoly-transcript-00963",
"Phypoly-transcript-00502")
		


gg_Fig <- FeaturePlot(Number2, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("Number2" = 15)) })
	

pdf("Number2_Grid_feature_BSS1Marker.pdf",width=20.5,height=14.5)
CombinePlots( gg_Fig )
dev.off()

gene_names = c("CALM_SCHPO",
"MYOI_DICDI",
"PVA_PLAFA",
"PPA_ASPFI",
"ACSM3_HUMAN",
"SYK_MOUSE",
"E2FB_ARATH",
"HELQ_HUMAN",
"ARUS_PSEAE",
"KBL_SHIFL",
"GRLE_DICDI",
"PTP2_DICDI",
"MFSD8_HUMAN",
"RACH_DICDI")
gene_ids = c("Phypoly-transcript-18971",
"Phypoly-transcript-02773",
"Phypoly-transcript-28060",
"Phypoly-transcript-05083",
"Phypoly-transcript-06548",
"Phypoly-transcript-06219",
"Phypoly-transcript-07599",
"Phypoly-transcript-01058",
"Phypoly-transcript-04786",
"Phypoly-transcript-02440",
"Phypoly-transcript-11890",
"Phypoly-transcript-06659",
"Phypoly-transcript-07556",
"Phypoly-transcript-16297")

gg_Fig <- FeaturePlot(Number2, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("Number2" = 15)) })
	

pdf("Number2_Grid_feature_Top5ClustMarkerANDSclerotiumANDFanMarker.pdf",width=20.5,height=14.5)
CombinePlots( gg_Fig )
dev.off()
	

```

##Flower

```{r}


#parallize scaling
library(future)
plan("multiprocess", workers = 8)
plan()

#set maximum to 10GB for each worker
options(future.globals.maxSize = 10000 * 1024^2)

all.genes <- rownames(Flower)
Flower = ScaleData(  Flower, features = all.genes ,vars.to.regress = c("reads","nFeature_RNA"))

#run PCA on all genes
#RunPCA(object, assay = NULL, features = NULL,  npcs = 50, rev.pca = FALSE, weight.by.var = TRUE, verbose = TRUE,  ndims.print = 1:5, nfeatures.print = 30, reduction.name = "pca",  reduction.key = "PC_", seed.use = 42)

Flower = RunPCA(Flower,features = all.genes,npcs = 20)

#plot PCA heatmaps
#DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

library(RColorBrewer)
library(ggplot2)

pdf("Flower_PCAheatmaps.pdf",width=30,height=20)
DimHeatmap(Flower, dims = 1:16, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
dev.off()

#elbow plot
pdf("Flower_PCelbow.pdf",width=20,height=10)
ElbowPlot(Flower, ndims = 20)
dev.off()

Flower[["spatial"]] <- CreateDimReducObject(embeddings = as.matrix(Flower@meta.data[,c("Spatial_1","Spatial_2")]),  key = "Spatial_", assay = "RNA")

#Cluster cells
Flower<- FindNeighbors(Flower, dims = 1:5)
Flower <- FindClusters(Flower, resolution = 0.6)


# Perform Hierarchical Clustering
library(tidyverse)
Data.hc <- Flower@reductions$pca@cell.embeddings[,c(1:10)] %>% get_dist(method = "pearson") %>% hclust(method = 'ward.D') ## PCs
Hcls <- data.frame(Clusters = as.factor(cutree(Data.hc, k = 2)))
Hcls <- Hcls %>% mutate(Wells = colnames(Flower))

Flower@meta.data$Hclust = Hcls$Clusters

Flower = RunUMAP(Flower, dims = 1:5,min.dist = 0.03)

#plot result
#DimPlot(object, dims = c(1, 2), cells = NULL, cols = NULL,  pt.size = NULL, reduction = NULL, group.by = NULL,  split.by = NULL, shape.by = NULL, order = NULL, label = FALSE,  label.size = 4, repel = FALSE, cells.highlight = NULL,  cols.highlight = "red", sizes.highlight = 1, na.value = "grey50",  combine = TRUE)

pdf("Flower_Embedding_Grid_PC5_res0.6.pdf",width=6.4,height=6.2)
DimPlot(object = Flower, reduction = 'spatial', pt.size = 5.5, shape.by = "orig.ident") + scale_shape_manual(values = c("Flower" = 15))
DimPlot(object = Flower, reduction = 'spatial', pt.size = 5.5, shape.by = "orig.ident",group.by = "Hclust") + scale_shape_manual(values = c("Flower" = 15))
DimPlot(object = Flower, reduction = 'umap', pt.size = 5.5, shape.by = "orig.ident") + scale_shape_manual(values = c("Flower" = 15))
DimPlot(object = Flower, reduction = 'umap', pt.size = 5.5, shape.by = "orig.ident",group.by = "Hclust")  + scale_shape_manual(values = c("Flower" = 15)) 
dev.off()

#some QC features

pdf("Flower_feature_QC.pdf",width=10,height=10)
FeaturePlot(Flower, reduction = 'spatial', shape.by = "orig.ident", pt.size = 2, features = c("nFeature_RNA","nCount_RNA","reads"),order = F, cols = c("grey",rev(brewer.pal(11,"RdYlBu")[1:11]))) + scale_shape_manual(values = c("Flower" = 15))
dev.off()

#Idents(object = Number2) <- Number2@meta.data$Hclust

plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(Flower, only.pos = TRUE,  logfc.threshold = 0.2)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(cluster , desc(avg_logFC))

write.csv(markers.anno,"Flower_Grid_allMarker.csv")

saveRDS(Flower, file = "Flower_Grid_SeuratObj.RDS")

				
gene_names = c(
"TOP2_DICDI",
"KI67_HUMAN",
"DPOLN_MOUSE",
"DNLI_HAEIN",
"TOLB_PSEA6",
"WDR54_MOUSE",
"ACTN4_RAT",
"CDC20_DICDI",
"DPOLB_RAT",
"RHOQ_HUMAN",
"FORD_DICDI",
"RASA2_MOUSE",
"CDKA1_ORYSJ",
"CCA11_ORYSJ",
"CCNB_DICDI",
"TBB2_PHYPO",
"MCM9_XENTR",
"PATS1_DICDI")
gene_ids = c(
"Phypoly-transcript-00381",
"Phypoly-transcript-00261",
"Phypoly-transcript-01132",
"Phypoly-transcript-14390",
"Phypoly-transcript-04660",
"Phypoly-transcript-13888",
"Phypoly-transcript-00407",
"Phypoly-transcript-07151",
"Phypoly-transcript-04465",
"Phypoly-transcript-23095",
"Phypoly-transcript-02192",
"Phypoly-transcript-06376",
"Phypoly-transcript-06666",
"Phypoly-transcript-05693",
"Phypoly-transcript-08256",
"Phypoly-transcript-06678",
"Phypoly-transcript-00963",
"Phypoly-transcript-00502")
		


gg_Fig <- FeaturePlot(Flower, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("Flower" = 15)) })
	

pdf("Flower_Grid_feature_BSS1Marker.pdf",width=20.5,height=14.5)
CombinePlots( gg_Fig )
dev.off()
	

gene_names = c("NGAP_DICDI",
"CYP11_RHIO9",
"ACSM3_RAT",
"SMBP2_HUMAN",
"MFSD8_HUMAN",
"RACH_DICDI",
"NAF1_SCHPO",
"YIQ3_SCHPO",
"CYAA_PODAS",
"DGKH_HUMAN",
"GRLE_DICDI",
"PTP2_DICDI")
gene_ids = c("Phypoly-transcript-11991",
"Phypoly-transcript-05260",
"Phypoly-transcript-03524",
"Phypoly-transcript-00202",
"Phypoly-transcript-07556",
"Phypoly-transcript-16297",
"Phypoly-transcript-09403",
"Phypoly-transcript-18264",
"Phypoly-transcript-05907",
"Phypoly-transcript-05492",
"Phypoly-transcript-11890",
"Phypoly-transcript-06659")


gg_Fig <- FeaturePlot(Flower, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("Flower" = 15)) })
	

pdf("Flower_Grid_feature_Top4ClustMarkerANDFanANDSclerotiumMarker.pdf",width=20.5,height=14.5)
CombinePlots( gg_Fig )
dev.off()
	

```

##Combine 96 well data 

For finding cross plasmodia patterns

```{r}
All96 = merge(x = Flower, y = c(StarDeath,Number2))

#scale
#ScaleData(  object,  features = NULL,  assay = NULL,  vars.to.regress = NULL,  split.by = NULL,  model.use = "linear",  use.umi = FALSE,  do.scale = TRUE,  do.center = TRUE,  scale.max = 10,  block.size = 1000,  min.cells.to.block = 3000,  verbose = TRUE)

all.genes <- rownames(All96)
All96 = ScaleData(  All96, features = all.genes ,vars.to.regress = c("reads","nFeature_RNA"))

#run PCA on all genes
#RunPCA(object, assay = NULL, features = NULL,  npcs = 50, rev.pca = FALSE, weight.by.var = TRUE, verbose = TRUE,  ndims.print = 1:5, nfeatures.print = 30, reduction.name = "pca",  reduction.key = "PC_", seed.use = 42)

All96 = RunPCA(All96,features = all.genes,npcs = 50)

#plot PCA heatmaps
#DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

library(RColorBrewer)

pdf("All96_PCAheatmaps.pdf",width=30,height=20)
DimHeatmap(All96, dims = 1:16, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(All96, dims = 17:32, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(All96, dims = 33:48, cells = 100, balanced = TRUE,  ncol = 4, fast = F) 
dev.off()


#PCA plot
pdf("All96_PCA.pdf",width=6,height=5)
DimPlot(object = All96, reduction = 'pca', dims = 1:2, pt.size = 2,group.by = "orig.ident",cols = c('#88CCEE', '#44AA99', '#999933','#CC6677','#117733', '#332288', '#DDCC77',  '#882255', '#AA4499', '#DDDDDD'))
DimPlot(object = All96, reduction = 'pca', dims = c(1,3), pt.size = 2,group.by = "orig.ident",cols = c('#88CCEE', '#44AA99', '#999933','#CC6677','#117733', '#332288', '#DDCC77',  '#882255', '#AA4499', '#DDDDDD'))
DimPlot(object = All96, reduction = 'pca', dims = 2:3, pt.size = 2,group.by = "orig.ident",cols = c('#88CCEE', '#44AA99', '#999933','#CC6677','#117733', '#332288', '#DDCC77',  '#882255', '#AA4499', '#DDDDDD'))
dev.off()


#elbow plot
pdf("All96_PCelbow.pdf",width=10,height=10)
ElbowPlot(All96, ndims = 50)
dev.off()



#run UMAP
#min.dist deafult is 0.3. Can go down to 0.001. Play around to modify clustering
All96 = RunUMAP(All96,dims = 1:10, min.dist = 0.3)

#Cluster cells
All96<- FindNeighbors(All96, dims = 1:10)
All96 <- FindClusters(All96, resolution = 0.4)


# Perform Hierarchical Clustering
library(tidyverse)
Data.hc <- All96@reductions$pca@cell.embeddings[,c(1:10)] %>% get_dist(method = "pearson") %>% hclust(method = 'ward.D') ## PCs
Hcls <- data.frame(Clusters = as.factor(cutree(Data.hc, k = 3)))
Hcls <- Hcls %>% mutate(Wells = colnames(All96))

All96@meta.data$Hclust = Hcls$Clusters

#plot result
#DimPlot(object, dims = c(1, 2), cells = NULL, cols = NULL,  pt.size = NULL, reduction = NULL, group.by = NULL,  split.by = NULL, shape.by = NULL, order = NULL, label = FALSE,  label.size = 4, repel = FALSE, cells.highlight = NULL,  cols.highlight = "red", sizes.highlight = 1, na.value = "grey50",  combine = TRUE)

pdf("All96_Embedding_PC10_res0.4.pdf",width=10,height=10)
DimPlot(object = All96, reduction = 'umap', pt.size = 2)
DimPlot(object = All96, reduction = 'umap', pt.size = 2,group.by = "orig.ident",cols = c('#88CCEE', '#44AA99', '#999933','#CC6677','#117733', '#332288', '#DDCC77',  '#882255', '#AA4499', '#DDDDDD'))
DimPlot(object = All96, reduction = 'umap', pt.size = 2,group.by = "Hclust")
dev.off()

#Cluster 1 is flowerand 0 and 2 are a mixture of star death and number 2

#some QC features

pdf("All96_feature_QC.pdf",width=10,height=10)
FeaturePlot(All96, reduction = 'umap', pt.size = 0.5, features = c("nFeature_RNA","nCount_RNA","reads"),order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]))
dev.off()


plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(All96, only.pos = TRUE,  logfc.threshold = 0.2)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(cluster , desc(avg_logFC))

write.csv(markers.anno,"All96_allMarker.csv")

saveRDS(All96, file = "All96_SeuratObj.RDS")

gene_names = c("FB293_ARATH","ELAV1_MOUSE","ARUS_PSEAE","CHIL3_MOUSE","WDR43_MOUSE","PX24B_DICDI","RGAA_DICDI","MSL2_DROME","UBCD2_DROME","PHSC_ARATH","FXL20_HUMAN","CALM_SCHPO","KI67_HUMAN","TOP2_DICDI","KIF22_DANRE","TRA1_DICDI","POLR_DROME","POL_BAEVM","NHLC2_CHICK","CADH3_ARATH","ADH6_YEAST","UBIQP_AGLNE","UBIQP_AGLNE","CHDH_HUMAN","ASPM_PONPY","FB293_ARATH","CDC20_DICDI","RACH_DICDI","M2K1_ORYSJ","MFSD8_DANRE","YKJ7_SCHPO","GCA_DICDI","Y723_CHLP8")

gene_ids = c("Phypoly-transcript-24913","Phypoly-transcript-19167","Phypoly-transcript-04786","Phypoly-transcript-08608","Phypoly-transcript-04781","Phypoly-transcript-10622","Phypoly-transcript-02695","Phypoly-transcript-08526","Phypoly-transcript-22056","Phypoly-transcript-09063","Phypoly-transcript-04596","Phypoly-transcript-18971","Phypoly-transcript-00261","Phypoly-transcript-00381","Phypoly-transcript-03486","Phypoly-transcript-25566","Phypoly-transcript-22123","Phypoly-transcript-06623","Phypoly-transcript-07740","Phypoly-transcript-10944","Phypoly-transcript-15025","Phypoly-transcript-17966","Phypoly-transcript-25680","Phypoly-transcript-05350","Phypoly-transcript-01017","Phypoly-transcript-18441","Phypoly-transcript-17752","Phypoly-transcript-16297","Phypoly-transcript-10910","Phypoly-transcript-04492","Phypoly-transcript-14329","Phypoly-transcript-01321","Phypoly-transcript-21401")


gg_Fig <- FeaturePlot(All96, pt.size = 1, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("All96_feature_ClustMarkerAND384AllMarker.pdf",width=22,height=20)
CombinePlots( gg_Fig )
dev.off()


gg_Fig <-VlnPlot(All96, features = gene_ids,pt.size = -1,group.by = "orig.ident")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("All96_Marker_Vln_Dot.pdf",width=10,height=10)
CombinePlots( gg_Fig )
DotPlot(All96, features = gene_ids,group.by = "orig.ident") +scale_colour_gradientn(colors =  c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9])) + theme(axis.text.x = element_text(angle = 90))
dev.off()




```




##384 well format


```{r}
library(Seurat)

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/384_format/")

# sequencing library GER063

input.dataRaw.GER063 <- read.delim("/mnt/SingleCellGenomics/scg_projects/SlimeMold/SpatialTranscriptomics384Experiments/SlimeMold_SmartSeq2_SpatialTranscriptomics384_BSS2_tpm_final", header=T)

#log transform
tmp<-input.dataRaw.GER063[,c("TPM")]
tmp[tmp<1]<-NA
tmp<-log2(tmp)
tmp[is.na(tmp)]<-0
rm(input.dataRaw_log2)
input.dataRaw_log2<-input.dataRaw.GER063
input.dataRaw_log2$log2<-tmp
rm(tmp)

#cast
library(reshape)

data.reshape<-input.dataRaw_log2[,c("cellID","gene","log2")]
data.GER063 <- as.data.frame(cast(data.reshape,cellID ~ gene,mean))
rm(data.reshape)

#add rownames
rownames(data.GER063) = data.GER063[,"cellID"]
data.GER063[,"cellID"] = NULL





# sequencing library GER064

input.dataRaw.GER064 <- read.delim("/mnt/SingleCellGenomics/scg_projects/SlimeMold/SpatialTranscriptomics384Experiments/SlimeMold_SmartSeq2_SpatialTranscriptomics384_BSS3_tpm_final", header=T)


#log transform
tmp<-input.dataRaw.GER064[,c("TPM")]
tmp[tmp<1]<-NA
tmp<-log2(tmp)
tmp[is.na(tmp)]<-0
rm(input.dataRaw_log2)
input.dataRaw_log2<-input.dataRaw.GER064
input.dataRaw_log2$log2<-tmp
rm(tmp)

#cast
library(reshape)

data.reshape<-input.dataRaw_log2[,c("cellID","gene","log2")]
data.GER064 <- as.data.frame(cast(data.reshape,cellID ~ gene,mean))
rm(data.reshape)

#add rownames
rownames(data.GER064) = data.GER064[,"cellID"]
data.GER064[,"cellID"] = NULL

# sequencing library GER065

input.dataRaw.GER065 <- read.delim("/mnt/SingleCellGenomics/scg_projects/SlimeMold/SpatialTranscriptomics384Experiments/SlimeMold_SmartSeq2_SpatialTranscriptomics384_BSS1_tpm_final", header=T)

#log transform
tmp<-input.dataRaw.GER065[,c("TPM")]
tmp[tmp<1]<-NA
tmp<-log2(tmp)
tmp[is.na(tmp)]<-0
rm(input.dataRaw_log2)
input.dataRaw_log2<-input.dataRaw.GER065
input.dataRaw_log2$log2<-tmp
rm(tmp)

#cast
library(reshape)

data.reshape<-input.dataRaw_log2[,c("cellID","gene","log2")]
data.GER065 <- as.data.frame(cast(data.reshape,cellID ~ gene,mean))
rm(data.reshape)

#add rownames
rownames(data.GER065) = data.GER065[,"cellID"]
data.GER065[,"cellID"] = NULL

# sequencing library GER066

input.dataRaw.GER066 <- read.delim("/mnt/SingleCellGenomics/scg_projects/SlimeMold/SpatialTranscriptomics384Experiments/SlimeMold_SmartSeq2_SpatialTranscriptomics384_BSS4_tpm_final", header=T)

#log transform
tmp<-input.dataRaw.GER066[,c("TPM")]
tmp[tmp<1]<-NA
tmp<-log2(tmp)
tmp[is.na(tmp)]<-0
rm(input.dataRaw_log2)
input.dataRaw_log2<-input.dataRaw.GER066
input.dataRaw_log2$log2<-tmp
rm(tmp)

#cast
library(reshape)

data.reshape<-input.dataRaw_log2[,c("cellID","gene","log2")]
data.GER066 <- as.data.frame(cast(data.reshape,cellID ~ gene,mean))
rm(data.reshape)

#add rownames
rownames(data.GER066) = data.GER066[,"cellID"]
data.GER066[,"cellID"] = NULL


##############finalize plate schemes

####BSS2####
#libB#

BSS2 = data.GER063

rownames(BSS2) = gsub("./BSS2/","",rownames(BSS2))

#get and add meta data

meta.GER063 <- read.delim("/mnt/SingleCellGenomics/scg_projects/SlimeMold/SpatialTranscriptomics384Experiments/BSS2/deMux_QC.csv",header = T,sep = ",")
rownames(meta.GER063) = meta.GER063$reads
meta.GER063$reads = NULL

BSS2 = cbind(meta.GER063,BSS2,stringsAsFactors=F)


rownames(BSS2) = gsub(pattern="(.*\\d)A", replacement="\\1_plate2",rownames(BSS2),perl = T)
rownames(BSS2) = gsub(pattern="(.*\\d)B", replacement="\\1_plate1",rownames(BSS2),perl = T)

rownames(BSS2) = paste("BSS2",rownames(BSS2),sep = "_")
  
#remove empty wells

BSS2 = BSS2[!rownames(BSS2) %in% rownames(BSS2)[grep("\\w12_plate2",rownames(BSS2))],]
BSS2 = BSS2[!rownames(BSS2) %in% c("BSS2_D11_plate2","BSS2_E11_plate2","BSS2_F11_plate2","BSS2_G11_plate2","BSS2_H11_plate2"),]

#save
saveRDS(BSS2,"data_BSS2_SS2_SpatialTranscriptomics384.RDS")


####BSS3####
#libC#

BSS3 = data.GER064

rownames(BSS3) = gsub("./BSS3/","",rownames(BSS3))

#get and add meta data

meta.GER064 <- read.delim("/mnt/SingleCellGenomics/scg_projects/SlimeMold/SpatialTranscriptomics384Experiments/BSS3/deMux_QC.csv",header = T,sep = ",")
rownames(meta.GER064) = meta.GER064$reads
meta.GER064$reads = NULL

BSS3 = cbind(meta.GER064,BSS3,stringsAsFactors=F)

rownames(BSS3) = gsub(pattern="(.*\\d)A", replacement="\\1_plate2",rownames(BSS3),perl = T)
rownames(BSS3) = gsub(pattern="(.*\\d)B", replacement="\\1_plate1",rownames(BSS3),perl = T)

#extract BSS1 cells from plate
tmp.BSS1 = rbind(BSS3[rownames(BSS3) %in% rownames(BSS3)[grep("\\w9_plate2",rownames(BSS3))],] , BSS3[rownames(BSS3) %in% rownames(BSS3)[grep("\\w10_plate2",rownames(BSS3))],] , BSS3[rownames(BSS3) %in% rownames(BSS3)[grep("\\w11_plate2",rownames(BSS3))],] , BSS3[rownames(BSS3) %in% rownames(BSS3)[grep("\\w12_plate2",rownames(BSS3))],] , stringsAsFactors = F)
rownames(tmp.BSS1) = paste("BSS1",rownames(tmp.BSS1),sep = "_")

rownames(tmp.BSS1) = gsub("plate2","plate3",rownames(tmp.BSS1))

rownames(tmp.BSS1) = gsub("9","1",rownames(tmp.BSS1))
rownames(tmp.BSS1) = gsub("10","2",rownames(tmp.BSS1))
rownames(tmp.BSS1) = gsub("11","3",rownames(tmp.BSS1))
rownames(tmp.BSS1) = gsub("12","4",rownames(tmp.BSS1))
#

BSS3 = BSS3[!rownames(BSS3) %in% c( rownames(BSS3)[grep("\\w12_plate2",rownames(BSS3))] , rownames(BSS3)[grep("\\w11_plate2",rownames(BSS3))] , rownames(BSS3)[grep("\\w10_plate2",rownames(BSS3))] , rownames(BSS3)[grep("\\w9_plate2",rownames(BSS3))]  ),]


rownames(BSS3) = paste("BSS3",rownames(BSS3),sep = "_")
  
#save
saveRDS(BSS3,"data_BSS3_SS2_SpatialTranscriptomics384.RDS")




####BSS4####
#libD#

BSS4 = data.GER066

rownames(BSS4) = gsub("./BSS4/","",rownames(BSS4))

#get and add meta data

meta.GER066 <- read.delim("/mnt/SingleCellGenomics/scg_projects/SlimeMold/SpatialTranscriptomics384Experiments/BSS4/deMux_QC.csv",header = T,sep = ",")
rownames(meta.GER066) = meta.GER066$reads
meta.GER066$reads = NULL

BSS4 = cbind(meta.GER066,BSS4,stringsAsFactors=F)

rownames(BSS4) = gsub(pattern="(.*\\d)A", replacement="\\1_plate1",rownames(BSS4),perl = T)
rownames(BSS4) = gsub(pattern="(.*\\d)B", replacement="\\1_plate2",rownames(BSS4),perl = T)

#extract BSS1 cells from plate
tmp2.BSS1 = rbind(BSS4[rownames(BSS4) %in% rownames(BSS4)[grep("\\w9_plate2",rownames(BSS4))],] , BSS4[rownames(BSS4) %in% rownames(BSS4)[grep("\\w10_plate2",rownames(BSS4))],] , BSS4[rownames(BSS4) %in% rownames(BSS4)[grep("\\w11_plate2",rownames(BSS4))],] , BSS4[rownames(BSS4) %in% rownames(BSS4)[grep("\\w12_plate2",rownames(BSS4))],] , stringsAsFactors = F)
rownames(tmp2.BSS1) = paste("BSS1",rownames(tmp2.BSS1),sep = "_")

rownames(tmp2.BSS1) = gsub("plate2","plate3",rownames(tmp2.BSS1))

rownames(tmp2.BSS1) = gsub("9","5",rownames(tmp2.BSS1))
rownames(tmp2.BSS1) = gsub("10","6",rownames(tmp2.BSS1))
rownames(tmp2.BSS1) = gsub("11","7",rownames(tmp2.BSS1))
rownames(tmp2.BSS1) = gsub("12","8",rownames(tmp2.BSS1))
#

BSS4 = BSS4[!rownames(BSS4) %in% c( rownames(BSS4)[grep("\\w12_plate2",rownames(BSS4))] , rownames(BSS4)[grep("\\w11_plate2",rownames(BSS4))] , rownames(BSS4)[grep("\\w10_plate2",rownames(BSS4))] , rownames(BSS4)[grep("\\w9_plate2",rownames(BSS4))]  ),]


rownames(BSS4) = paste("BSS4",rownames(BSS4),sep = "_")
  
#save
saveRDS(BSS4,"data_BSS4_SS2_SpatialTranscriptomics384.RDS")


####BSS1####
#libA#

BSS1 = data.GER065

rownames(BSS1) = gsub("./BSS1/","",rownames(BSS1))

#get and add meta data

meta.GER065 <- read.delim("/mnt/SingleCellGenomics/scg_projects/SlimeMold/SpatialTranscriptomics384Experiments/BSS1/deMux_QC.csv",header = T,sep = ",")
rownames(meta.GER065) = meta.GER065$reads
meta.GER065$reads = NULL

BSS1 = cbind(meta.GER065,BSS1,stringsAsFactors=F)


rownames(BSS1) = gsub(pattern="(.*\\d)A", replacement="\\1_plate1",rownames(BSS1),perl = T)
rownames(BSS1) = gsub(pattern="(.*\\d)B", replacement="\\1_plate2",rownames(BSS1),perl = T)

rownames(BSS1) = paste("BSS1",rownames(BSS1),sep = "_")
  
BSS1 = rbind(BSS1, tmp.BSS1, tmp2.BSS1, stringsAsFactors = F)

#save
saveRDS(BSS1,"data_BSS1_SS2_SpatialTranscriptomics384.RDS")





```


############# Analysis 384 data ###############

BSS1

```{r}
library(Seurat) #version 3
library(Matrix)
library(tidyverse)
library(RColorBrewer)
library(ggplot2)
library(RANN)
library(stats)
library(BiocManager)
library(cowplot)
library(readxl)
library(viridis)
library(ComplexHeatmap)
library(circlize)
library(statmod)
library(Rtsne)
library(cluster)
library(factoextra)
library(dendextend)
library(harmony)
library(gtools)
library(umap)
library(sctransform)

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/")
source("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/r-Analysis-TG/r-Analysis-TG/CustomFunctions.R")

# Set the color-codes. Colors obtained from: https://www.color-hex.com/color-palettes/?page=3
cols_row <- c("#ff9b35", "#ffd05c", "#8aa77e", "#ecd6b4", "#84ade7", "#a079c8", "#aaaaaa", "#bd3434")
cols_column <- c("#ff9b35", "#ffd05c", "#8aa77e", "#736e69","#84ade7", "#064887", 
                 "#aaaaaa", "#bd3434", "#006753", "#a079c8", "#3b0d6a", "#ecd6b4")
cols_seqBatch <- c("#6e4632", "#af7d50", "#cdaf78")
cols_clusters <- c("#067254", "#72c3ac", "#eebe62", "#abb8b4", "#19617e", "#0aa57a", "#cccc66", "#97449c", 
                  "#ae334c", "#194772", "#b27b37")

# Load the data matrix and the experimental information (well coordinates)
data = readRDS("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/384_format/data_BSS1_SS2_SpatialTranscriptomics384.RDS")
coordinates <- read_excel("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/r-Analysis-TG/r-Analysis-TG/DATA/BSS1_coordinates.xlsx", sheet = "BSS1_Grid", col_names = TRUE)

# Extract the QC data from the data matrix and relabel all the wells adding the 384-well-plate coordinates
data_m <- data[,c(4:dim(data)[2])]
QC_reads <- data[,c(1:3)]
data_m <- relabelWells(data_m, coordinates)
QC_reads <- relabelWells(QC_reads, coordinates)

# Create a data frame with all the metadata information 
infoData <- data.frame(
    ID = rownames(QC_reads),
    exp = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 1),
    plate.row = str_extract(apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo,
                                  pos = 2), "[A-Z]"),
    plate.col = str_extract(apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo,
                                  pos = 2), "[0-9]{1,2}"),
    plate.num = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 3),
    well = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 4),
    totalReads = QC_reads[ ,1],
    pairedReads = QC_reads[ ,3],
    nnzeroEntries = rowSums(data_m != 0),
    zeroEntries = rowSums(data_m == 0),
    normCounts =  rowSums(data_m)
)
grid <- expand.grid( grid.x = c("A","B","C","D","E","F","G","H","I","J","K","L","M","N","O","P","Q", "R", "S"), 
                     grid.y = c(1:17) ) %>% mutate( grid.xy = paste0(grid.x, grid.y) ) 
s_metaData <- grid %>% left_join( infoData, by = c("grid.xy"="well") )

rm(QC_reads, infoData, grid, coordinates)

# Plot histogram of normalize Counts to check the distributions an set the cut-off for removing low quality wells
# Plot the data on the spatial grid to observe the Count distributions
ggplot( s_metaData, aes(pairedReads) ) + geom_histogram( binwidth = 15000 ) +
        scale_x_continuous( breaks = seq(0, 5000000, 50000) ) + 
        theme( axis.text.x = element_text(size = 8, angle = 90) )
ggplot( s_metaData, aes(x = grid.x, y = grid.y) ) +
        geom_point( aes(size = pairedReads, color = nnzeroEntries), shape = 15)  +
        scale_color_viridis( discrete = FALSE, option = "D", na.value = "transparent" ) +
        scale_x_discrete( expand = c(0.55, 0) ) + 
        scale_y_discrete( breaks = seq(1, 17, 1), expand = c(0.2, 0) ) + 
        theme (panel.background = element_blank() ) +
        theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
               panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
               legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
  
# Remove wells with low-quality counts from metadata and from the data matrix
emptyWells_ID <- s_metaData[s_metaData$pairedReads<200000,]$grid.xy
#emptyWells_ID <- rbind( emptyWells_ID, s_metaData[s_metaData$pairedReads>3000000,]$grid.xy )
emptyWells_ID <- append( emptyWells_ID, c("A3", "B3", "B1", "B2", "C1","C2", "O10", "L10", "M9","D1","D2","S10","H15","H16","I15","I16","B7","J15","J16","R10","S9","M9","L7","L8","L3","L4","L2","M2","M3","B14","B15","B16","B12","C12") )

emptyWells_df <- lapply(X = emptyWells_ID, FUN = function(x){ s_metaData[grep(paste0("^", x, "$"),
                                                                              s_metaData$grid.xy),] })
emptyWells_df <- as.data.frame(do.call(rbind, emptyWells_df))
emptyWells_df <- emptyWells_df[!duplicated(emptyWells_df$ID), ] %>% drop_na()

metaData <- s_metaData %>% anti_join(emptyWells_df, by = c("ID"="ID")) %>% 
            mutate(plate.col=as.factor(plate.col)) %>% 
            mutate(plate.row=as.factor(plate.row)) %>% 
            mutate(plate.num=as.factor(plate.num))
data_m <-  data_m[-which(rownames(data_m) %in% emptyWells_df$ID), ]

rm(emptyWells_ID, emptyWells_df)

# Add the Sequencing Batch label to the metadata file 
df1 <- metaData %>% filter(metaData$plate.num == 1) %>% mutate(SeqBatch = 1)
df2 <- metaData %>% filter(metaData$plate.num == 2) %>% mutate(SeqBatch = 1)
df3 <- metaData %>% filter(metaData$plate.num == 3)
df3 <- df3 %>% filter(df3$plate.col %in% c(1, 2, 3, 4 )) %>% mutate(SeqBatch = 2)
df4 <- metaData %>% filter(metaData$plate.num == 3)
df4 <- df4 %>% filter(df4$plate.col %in% c(5, 6, 7, 8 )) %>% mutate(SeqBatch = 3)
metaData <- rbind(df1, df2, df3, df4)
rm(df1, df2, df3, df4)

saveRDS(data_m,"BSS1_final_data.RDS")
saveRDS(metaData,"BSS1_final_meta.RDS")

ggplot( metaData, aes(x = grid.x, y = grid.y) ) +
        geom_point( aes(size = pairedReads, color = nnzeroEntries), shape = 15)  +
        scale_color_viridis( discrete = FALSE, option = "D", na.value = "transparent" ) +
        scale_x_discrete( expand = c(0.55, 0) ) + 
        scale_y_discrete( breaks = seq(1, 17, 1), expand = c(0.2, 0) ) + 
        theme (panel.background = element_blank() ) +
        theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
               panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
               legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
  
```

BSS2

```{r}

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/")
source("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/r-Analysis-TG/r-Analysis-TG/CustomFunctions.R")

# Set the color-codes. Colors obtained from: https://www.color-hex.com/color-palettes/?page=3
cols_row <- c("#ff9b35", "#ffd05c", "#8aa77e", "#ecd6b4", "#84ade7", "#a079c8", "#aaaaaa", "#bd3434")
cols_column <- c("#ff9b35", "#ffd05c", "#8aa77e", "#736e69","#84ade7", "#064887", 
                 "#aaaaaa", "#bd3434", "#006753", "#a079c8", "#3b0d6a", "#ecd6b4")
cols_seqBatch <- c("#6e4632", "#af7d50", "#cdaf78")
cols_clusters <- c("#067254", "#72c3ac", "#eebe62", "#abb8b4", "#19617e", "#0aa57a", "#cccc66", "#97449c", 
                  "#ae334c", "#194772", "#b27b37")

# Load the data matrix and the experimental information (well coordinates)
data = readRDS("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/384_format/data_BSS2_SS2_SpatialTranscriptomics384.RDS")
coordinates <- read_excel("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/r-Analysis-TG/r-Analysis-TG/DATA/BSS2_coordinates.xlsx", sheet = "BSS2_Grid", col_names = TRUE)

# Extract the QC data from the data matrix and relabel all the wells adding the 384-well-plate coordinates
data_m <- data[,c(4:dim(data)[2])]
QC_reads <- data[,c(1:3)]
data_m <- relabelWells(data_m, coordinates)
QC_reads <- relabelWells(QC_reads, coordinates)

# Create a data frame with all the metadata information 
infoData <- data.frame(
  ID = rownames(QC_reads),
  exp = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 1),
  plate.row = str_extract(apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo,
                                pos = 2), "[A-Z]"),
  plate.col = str_extract(apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo,
                                pos = 2), "[0-9]{1,2}"),
  plate.num = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 3),
  well = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 4),
  totalReads = QC_reads[,1],
  pairedReads = QC_reads[,3],
  nnzeroEntries = rowSums(data_m!=0),
  zeroEntries = rowSums(data_m==0),
  normCounts =  rowSums(data_m)
)
grid <- expand.grid( grid.x = c("A","B","C","D","E","F","G","H","I","J","K","L","M","N","O","P","Q"), 
                     grid.y = c(1:15) ) %>% mutate( grid.xy = paste0(grid.x, grid.y) ) 
s_metaData <- grid %>% left_join(infoData, by = c("grid.xy"="well"))

rm(QC_reads, infoData, grid, coordinates)

# Plot histogram of normalize Counts to check the distributions an set the cut-off for removing low quality wells
# Plot the data on the spatial grid to observe the Count distributions
ggplot( s_metaData, aes(pairedReads) ) + geom_histogram( binwidth = 10000 ) +
        scale_x_continuous( breaks = seq(0, 5000000, 50000) ) + 
        theme( axis.text.x = element_text(size=8, angle=90) )
ggplot( s_metaData, aes(x = grid.x, y = grid.y) ) +
        geom_point( aes(size = normCounts, color = nnzeroEntries), shape = 15)  +
        scale_color_viridis( discrete = FALSE, option = "D", na.value = "transparent" ) +
        scale_x_discrete( expand = c(0.75, 0) ) + 
        scale_y_discrete( breaks = seq(1, 17, 1), expand = c(0.35, 0) ) + 
        theme (panel.background = element_blank() ) +
        theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
               panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
               legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )

# Remove wells with low-quality counts from metadata and from the data matrix
emptyWells_ID <- s_metaData[s_metaData$pairedReads<100000,]$grid.xy
emptyWells_ID <- append(emptyWells_ID, c("H7", "H8", "I7", "I8", "J7", "L7","N6","J8","K10","L10","B13","B14"))
emptyWells_df <- lapply( X = emptyWells_ID, 
                         FUN = function(x) { s_metaData[grep(paste0("^", x, "$"), s_metaData$grid.xy),] } )
emptyWells_df <- as.data.frame( do.call(rbind, emptyWells_df) )
emptyWells_df <- emptyWells_df[!duplicated(emptyWells_df$ID),] %>% drop_na()

metaData <- s_metaData %>% anti_join( emptyWells_df, by = c("ID" = "ID") ) %>% 
            mutate( plate.col = as.factor(plate.col) ) %>% 
            mutate( plate.row = as.factor(plate.row) ) %>% 
            mutate( plate.num = as.factor(plate.num) )
data_m <-  data_m[-which(rownames(data_m) %in% emptyWells_df$ID), ]

rm(emptyWells_ID, emptyWells_df)

# Add the Plate Batch label to the metadata file 
df1 <- metaData %>% filter(metaData$plate.num == 1)
df2 <- metaData %>% filter(metaData$plate.num == 2)
metaData <- rbind(df1, df2)
rm(df1, df2)

saveRDS(data_m,"BSS2_final_data.RDS")
saveRDS(metaData,"BSS2_final_meta.RDS")

ggplot( metaData, aes(x = grid.x, y = grid.y) ) +
        geom_point( aes(size = pairedReads, color = nnzeroEntries), shape = 15)  +
        scale_color_viridis( discrete = FALSE, option = "D", na.value = "transparent" ) +
        scale_x_discrete( expand = c(0.55, 0) ) + 
        scale_y_discrete( breaks = seq(1, 17, 1), expand = c(0.2, 0) ) + 
        theme (panel.background = element_blank() ) +
        theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
               panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
               legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
  

```

BSS3

```{r}

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/")
source("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/r-Analysis-TG/r-Analysis-TG/CustomFunctions.R")

# Set the color-codes. Colors obtained from: https://www.color-hex.com/color-palettes/?page=3
cols_row <- c("#ff9b35", "#ffd05c", "#8aa77e", "#ecd6b4", "#84ade7", "#a079c8", "#aaaaaa", "#bd3434")
cols_column <- c("#ff9b35", "#ffd05c", "#8aa77e", "#736e69","#84ade7", "#064887", 
                 "#aaaaaa", "#bd3434", "#006753", "#a079c8", "#3b0d6a", "#ecd6b4")
cols_seqBatch <- c("#6e4632", "#af7d50", "#cdaf78")
cols_clusters <- c("#067254", "#72c3ac", "#eebe62", "#abb8b4", "#19617e", "#0aa57a", "#cccc66", "#97449c", 
                  "#ae334c", "#194772", "#b27b37")

# Load the data matrix and the experimental information (well coordinates)
data = readRDS("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/384_format/data_BSS3_SS2_SpatialTranscriptomics384.RDS")
coordinates <- read_excel("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/r-Analysis-TG/r-Analysis-TG/DATA/BSS3_coordinates.xlsx", sheet = "BSS3_Grid", col_names = TRUE)

# Extract the QC data from the data matrix and relabel all the wells adding the 384-well-plate coordinates
data_m <- data[,c(4:dim(data)[2])]
QC_reads <- data[,c(1:3)]
data_m <- relabelWells(data_m, coordinates)
QC_reads <- relabelWells(QC_reads, coordinates)

# Create a data frame with all the metadata information 
infoData <- data.frame(
      ID = rownames(QC_reads),
      exp = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 1),
      plate.row = str_extract(apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo,
                                    pos = 2), "[A-Z]"),
      plate.col = str_extract(apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo,
                                    pos = 2), "[0-9]{1,2}"),
      plate.num = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 3),
      well = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 4),
      totalReads = QC_reads[,1],
      pairedReads = QC_reads[,3],
      nnzeroEntries = rowSums(data_m!=0),
      zeroEntries = rowSums(data_m==0),
      normCounts =  rowSums(data_m)
)
grid <- expand.grid( grid.x = c("A","B","C","D","E","F","G","H","I","J"), 
                     grid.y = c(1:16) ) %>% mutate( grid.xy = paste0(grid.x, grid.y) ) 
s_metaData <- grid %>% left_join(infoData, by = c("grid.xy"="well"))

rm(QC_reads, infoData, grid, coordinates)

# Plot histogram of normalize Counts to check the distributions an set the cut-off for removing low quality wells
# Plot the data on the spatial grid to observe the Count distributions
ggplot( s_metaData, aes(pairedReads) ) + geom_histogram( binwidth = 20000 ) + 
        scale_x_continuous( breaks = seq(0, 2500000, 50000) ) + 
        theme( axis.text.x = element_text(size=8, angle=90) )
ggplot( s_metaData, aes(x = grid.x, y = grid.y) ) +
        geom_point( aes(size = normCounts, color = nnzeroEntries), shape = 15)  +
        scale_color_viridis( discrete = FALSE, option = "D", na.value = "transparent" ) +
        scale_x_discrete( expand = c(1.8, 0) ) + 
        scale_y_discrete( breaks = seq(1, 17, 1), expand = c(0.3, 0) ) + 
        theme (panel.background = element_blank() ) +
        theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
               panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
               legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )

# Remove wells with low-quality counts from metadata and from the data matrix
emptyWells_ID <- s_metaData[s_metaData$pairedReads<100000,]$grid.xy
emptyWells_ID <- append(emptyWells_ID, c("A4","A7","H4","H5","I4","J9","J10","E16","G16","H16","I15","B3","C2","C3","D1"))

emptyWells_df <- lapply(X = emptyWells_ID, FUN = function(x){ s_metaData[grep(paste0("^", x, "$"),
                                                                              s_metaData$grid.xy),] })
emptyWells_df <- as.data.frame(do.call(rbind, emptyWells_df))
emptyWells_df <- emptyWells_df[!duplicated(emptyWells_df$ID), ] %>% drop_na()

metaData <- s_metaData %>% anti_join(emptyWells_df, by = c("ID"="ID")) %>% 
            mutate(plate.col = as.factor(plate.col)) %>% 
            mutate(plate.row = as.factor(plate.row)) %>% 
            mutate(plate.num = as.factor(plate.num))
data_m <-  data_m[-which(rownames(data_m) %in% emptyWells_df$ID), ]

rm(emptyWells_ID, emptyWells_df)

# Add the Plate Batch label to the metadata file 
df1 <- metaData %>% filter(metaData$plate.num == 1)
df2 <- metaData %>% filter(metaData$plate.num == 2)
metaData <- rbind(df1, df2)
rm(df1, df2)

saveRDS(data_m,"BSS3_final_data.RDS")
saveRDS(metaData,"BSS3_final_meta.RDS")

ggplot( metaData, aes(x = grid.x, y = grid.y) ) +
        geom_point( aes(size = pairedReads, color = nnzeroEntries), shape = 15)  +
        scale_color_viridis( discrete = FALSE, option = "D", na.value = "transparent" ) +
        scale_x_discrete( expand = c(0.55, 0) ) + 
        scale_y_discrete( breaks = seq(1, 17, 1), expand = c(0.2, 0) ) + 
        theme (panel.background = element_blank() ) +
        theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
               panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
               legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
  

```

BSS4

```{r}

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/")
source("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/r-Analysis-TG/r-Analysis-TG/CustomFunctions.R")

# Set the color-codes. Colors obtained from: https://www.color-hex.com/color-palettes/?page=3
cols_row <- c("#ff9b35", "#ffd05c", "#8aa77e", "#ecd6b4", "#84ade7", "#a079c8", "#aaaaaa", "#bd3434")
cols_column <- c("#ff9b35", "#ffd05c", "#8aa77e", "#736e69","#84ade7", "#064887", 
                 "#aaaaaa", "#bd3434", "#006753", "#a079c8", "#3b0d6a", "#ecd6b4")
cols_seqBatch <- c("#6e4632", "#af7d50", "#cdaf78")
cols_clusters <- c("#067254", "#72c3ac", "#eebe62", "#abb8b4", "#19617e", "#0aa57a", "#cccc66", "#97449c", 
                  "#ae334c", "#194772", "#b27b37")

# Load the data matrix and the experimental information (well coordinates)
data = readRDS("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/384_format/data_BSS4_SS2_SpatialTranscriptomics384.RDS")
coordinates <- read_excel("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/r-Analysis-TG/r-Analysis-TG/DATA/BSS4_coordinates.xlsx", sheet = "BSS4_Grid", col_names = TRUE)

# Extract the QC data from the data matrix and relabel all the wells adding the 384-well-plate coordinates
data_m <- data[,c(4:dim(data)[2])]
QC_reads <- data[,c(1:3)]
data_m <- relabelWells(data_m, coordinates)
QC_reads <- relabelWells(QC_reads, coordinates)

# Create a data frame with all the metadata information 
infoData <- data.frame(
      ID = rownames(QC_reads),
      exp = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 1),
      plate.row = str_extract(apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo,
                                    pos = 2), "[A-Z]"),
      plate.col = str_extract(apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo,
                                    pos = 2), "[0-9]{1,2}"),
      plate.num = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 3),
      well = apply(X = as.data.frame(rownames(QC_reads)), MARGIN = 1, FUN = extractInfo, pos = 4),
      totalReads = QC_reads[,1],
      pairedReads = QC_reads[,3],
      nnzeroEntries = rowSums(data_m!=0),
      zeroEntries = rowSums(data_m==0),
      normCounts =  rowSums(data_m)
)
grid <- expand.grid( grid.x = c("A","B","C","D","E","F","G","H","I","J","K","L","M","N"), 
                     grid.y = c(1:14) ) %>% mutate( grid.xy = paste0(grid.x, grid.y) ) 
s_metaData <- grid %>% left_join(infoData, by = c("grid.xy"="well"))

rm(QC_reads, infoData, grid)

# Plot histogram of normalize Counts to check the distributions an set the cut-off for removing low quality wells
# Plot the data on the spatial grid to observe the Count distributions
ggplot( s_metaData, aes(pairedReads) ) + geom_histogram( binwidth = 20000 ) + 
        scale_x_continuous(breaks = seq(0, 2500000, 50000)) +
        theme(axis.text.x = element_text(size=8, angle=90))
ggplot( s_metaData, aes(x = grid.x, y = grid.y) ) +
        geom_point( aes(size = pairedReads, color = nnzeroEntries), shape = 15)  +
        scale_color_viridis( discrete = FALSE, option = "D", na.value = "transparent" ) +
        scale_x_discrete( expand = c(1.2, 0) ) + 
        scale_y_discrete( breaks = seq(1, 17, 1), expand = c(0.42, 0) ) + 
        theme (panel.background = element_blank() ) +
        theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
               panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
               legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )

# Remove wells with low-quality counts from metadata and from the data matrix
emptyWells_ID <- s_metaData[s_metaData$pairedReads<100000,]$grid.xy
emptyWells_ID <- append(emptyWells_ID, c("D7","D10","D11","D12","D13", "D14","E10","E11","E12","E13", "E14", "F13", "F14","F12","F10","F9","F8","F7","G8","G9","G10", "L4", "L5", "L6","M4", "M5","M6","N4","N5","N6","N7","N8","K4","K5","K6","M13","M14","N13","N14","L13","L14","J14","E1","F1","G1","F2","J1","J2","J3","J4","I2","I3","I4"))
emptyWells_df <- lapply(X = emptyWells_ID, FUN = function(x){ s_metaData[grep(paste0("^", x, "$"),
                                                                              s_metaData$grid.xy),] })
emptyWells_df <- as.data.frame(do.call(rbind, emptyWells_df))
emptyWells_df <- emptyWells_df[!duplicated(emptyWells_df$ID), ] %>% drop_na()

metaData <- s_metaData %>% anti_join( emptyWells_df, by = c("ID"="ID") ) %>% 
            mutate(plate.col=as.factor(plate.col)) %>% 
            mutate(plate.row=as.factor(plate.row)) %>% 
            mutate(plate.num=as.factor(plate.num))
data_m <-  data_m[-which(rownames(data_m) %in% emptyWells_df$ID), ]

rm(emptyWells_ID, emptyWells_df)

# Add the Plate Batch label to the metadata file 
df1 <- metaData %>% filter(metaData$plate.num == 1)
df2 <- metaData %>% filter(metaData$plate.num == 2)
metaData <- rbind(df1, df2)
rm(df1, df2)


saveRDS(data_m,"BSS4_final_data.RDS")
saveRDS(metaData,"BSS4_final_meta.RDS")

ggplot( metaData, aes(x = grid.x, y = grid.y) ) +
        geom_point( aes(size = pairedReads, color = nnzeroEntries), shape = 15)  +
        scale_color_viridis( discrete = FALSE, option = "D", na.value = "transparent" ) +
        scale_x_discrete( expand = c(0.55, 0) ) + 
        scale_y_discrete( breaks = seq(1, 17, 1), expand = c(0.2, 0) ) + 
        theme (panel.background = element_blank() ) +
        theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
               panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
               legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
  

```


############################ Add location information to individual slime molds ########################

```{r}
BSS1 = readRDS("BSS1_final_data.RDS")
BSS1 = as.data.frame(t(BSS1),stringsAsFactors = F)
BSS1.meta = readRDS("BSS1_final_meta.RDS")

BSS2 = readRDS("BSS2_final_data.RDS")
BSS2 = as.data.frame(t(BSS2),stringsAsFactors = F)
BSS2.meta = readRDS("BSS2_final_meta.RDS")

BSS3 = readRDS("BSS3_final_data.RDS")
BSS3 = as.data.frame(t(BSS3),stringsAsFactors = F)
BSS3.meta = readRDS("BSS3_final_meta.RDS")

BSS4 = readRDS("BSS4_final_data.RDS")
BSS4 = as.data.frame(t(BSS4),stringsAsFactors = F)
BSS4.meta = readRDS("BSS4_final_meta.RDS")


#Get annotated transcripts

anno = read.delim("/mnt/SingleCellGenomics/genome_data/10xSlimeMoldTranscriptome/41598_2017_12250_MOESM2_ESM.txt.csv",sep = "\t",header = T)
anno = anno[1:49403,]
colnames(anno) = c("transcript","ORF_length","database","ID","description","from","to","IPR_number","Domain_description","GOterm","Log2_FoldChange","pvalue","baseMean","IfcSE","stat","padj","UniProt_Name","UniProt_ID","UniProt_Acc","Regulation")

anno.gene.id = unique(as.character(anno$transcript))

anno$transcript = gsub(x = anno$transcript,pattern = "_",replacement = "-")

anno.gene = anno[,c("transcript","UniProt_Acc")]
anno.gene = unique(anno.gene)
anno.gene$transcript = as.character(anno.gene$transcript)
anno.gene$UniProt_Acc = as.character(anno.gene$UniProt_Acc)
rownames(anno.gene) = anno.gene$transcript
anno.gene$UniProt_Acc[is.na(anno.gene$UniProt_Acc)] = "noGene"

#Get GO information
anno.GO = anno[,c("transcript","GOterm","ORF_length","description")]

#remove rows without GO assignment
emptyfields <- rowSums(anno.GO == "") == 1
anno.GO <- anno.GO[!emptyfields,]
anno.GO = unique(anno.GO)

anno.GO$transcript = as.character(anno.GO$transcript)
anno.GO$GOterm = as.character(anno.GO$GOterm)
anno.GO$ORF_length = as.numeric(anno.GO$ORF_length)

id_list_GO = vector("list")
id_list_Domain = vector("list")
ORF_list = vector("list")

for (i in 1:nrow(anno.GO)) {
  k = anno.GO[i,1]
  if (k %in% names(id_list_GO)) {
    if (anno.GO[i,3] == ORF_list[[k]]) {
      ORF_list[[k]] = anno.GO[i,3]
      id_list_GO[[k]] = paste(id_list_GO[[k]],anno.GO[i,2],sep = "|")
      if (id_list_Domain[[k]] == anno.GO[i,4]) {
        next
      } else {
        id_list_Domain[[k]] = paste(id_list_Domain[[k]],anno.GO[i,4],sep = "|")
      }
    } else {
      if (anno.GO[i,3] > ORF_list[[k]]) {
        ORF_list[[k]] = anno.GO[i,3]
        id_list_GO[[k]] = anno.GO[i,2]
        id_list_Domain[[k]] = anno.GO[i,4]
      } else {
        next
      }
    }
  } else {
    id_list_GO[[k]] = anno.GO[i,2]
    ORF_list[[k]] = anno.GO[i,3]
    id_list_Domain[[k]] = anno.GO[i,4]
  }
}

anno.GO = data.frame(unlist(id_list_GO,use.names = TRUE),stringsAsFactors=F)
anno.GO$transcript = rownames(anno.GO)
colnames(anno.GO) = c("GOterm","transcript")

anno.Domain = data.frame(unlist(id_list_Domain,use.names = TRUE),stringsAsFactors=F)
anno.Domain$transcript = rownames(anno.Domain)
colnames(anno.Domain) = c("Domain","transcript")

anno.gene = merge(anno.gene,anno.GO, by = "transcript")
anno.gene = merge(anno.gene,anno.Domain, by = "transcript")

############# get data and remove genes

BSS1 = BSS1[rownames(BSS1) %in% anno.gene.id,]
BSS2 = BSS2[rownames(BSS2) %in% anno.gene.id,]
BSS3 = BSS3[rownames(BSS3) %in% anno.gene.id,]
BSS4 = BSS4[rownames(BSS4) %in% anno.gene.id,]

############# load in Seurat

BSS1 <- CreateSeuratObject(counts = BSS1, project = "BSS1", min.cells = 3, min.features = 10)
BSS1
BSS2 <- CreateSeuratObject(counts = BSS2, project = "BSS2", min.cells = 3, min.features = 10)
BSS2
BSS3 <- CreateSeuratObject(counts = BSS3, project = "BSS3", min.cells = 3, min.features = 10)
BSS3
BSS4 <- CreateSeuratObject(counts = BSS4, project = "BSS4", min.cells = 3, min.features = 10)
BSS4
```


##BSS1

```{r}
############# load meta data into Seurat

BSS1.meta = BSS1.meta[match(colnames(BSS1),BSS1.meta$ID),]
BSS1.meta$Spatial_1 = BSS1.meta$grid.x
BSS1.meta$Spatial_2 = BSS1.meta$grid.y
dim = as.data.frame(BSS1.meta[,c("Spatial_1","Spatial_2")],stringsAsFactors = F)
rownames(dim) = colnames(BSS1)

myLetters <- toupper(letters[1:26])
dim$Spatial_1 = as.numeric(match(dim$Spatial_1, myLetters))
dim$Spatial_2 = as.numeric(dim$Spatial_2)
dim = as.matrix(dim)

BSS1@meta.data$reads = BSS1.meta$pairedReads

##############

#parallize scaling
library(future)
plan("multiprocess", workers = 8)
plan()

#set maximum to 10GB for each worker
options(future.globals.maxSize = 10000 * 1024^2)

all.genes <- rownames(BSS1)
BSS1 = ScaleData(  BSS1, features = all.genes ,vars.to.regress = c("reads","nFeature_RNA"))

#run PCA on all genes
#RunPCA(object, assay = NULL, features = NULL,  npcs = 50, rev.pca = FALSE, weight.by.var = TRUE, verbose = TRUE,  ndims.print = 1:5, nfeatures.print = 30, reduction.name = "pca",  reduction.key = "PC_", seed.use = 42)

BSS1 = RunPCA(BSS1,features = all.genes,npcs = 50)

#plot PCA heatmaps
#DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

library(RColorBrewer)
library(ggplot2)

pdf("BSS1_PCAheatmaps.pdf",width=30,height=20)
DimHeatmap(BSS1, dims = 1:16, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(BSS1, dims = 17:32, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(BSS1, dims = 33:48, cells = 100, balanced = TRUE,  ncol = 4, fast = F) 
dev.off()

#elbow plot
pdf("BSS1_PCelbow.pdf",width=20,height=10)
ElbowPlot(BSS1, ndims = 100)
dev.off()

BSS1[["spatial"]] <- CreateDimReducObject(embeddings = dim,  key = "Spatial_", assay = "RNA")

#Cluster cells
BSS1<- FindNeighbors(BSS1, dims = 1:15)
BSS1 <- FindClusters(BSS1, resolution = 0.4)


# Perform Hierarchical Clustering
library(tidyverse)
Data.hc <- BSS1@reductions$pca@cell.embeddings[,c(1:15)] %>% get_dist(method = "pearson") %>% hclust(method = 'ward.D') ## PCs
Hcls <- data.frame(Clusters = as.factor(cutree(Data.hc, k = 3)))
Hcls <- Hcls %>% mutate(Wells = colnames(BSS1))

BSS1@meta.data$Hclust = Hcls$Clusters

BSS1 = RunUMAP(BSS1, dims = 1:15,min.dist = 0.03)

#plot result
#DimPlot(object, dims = c(1, 2), cells = NULL, cols = NULL,  pt.size = NULL, reduction = NULL, group.by = NULL,  split.by = NULL, shape.by = NULL, order = NULL, label = FALSE,  label.size = 4, repel = FALSE, cells.highlight = NULL,  cols.highlight = "red", sizes.highlight = 1, na.value = "grey50",  combine = TRUE)

pdf("BSS1_Embedding_Grid_PC15_res0.4.pdf",width=6.8,height=5)
DimPlot(object = BSS1, reduction = 'spatial', pt.size = 5, shape.by = "orig.ident") + scale_shape_manual(values = c("BSS1" = 15))
DimPlot(object = BSS1, reduction = 'spatial', pt.size = 5, shape.by = "orig.ident",group.by = "Hclust") + scale_shape_manual(values = c("BSS1" = 15))
DimPlot(object = BSS1, reduction = 'umap', pt.size = 5, shape.by = "orig.ident") + scale_shape_manual(values = c("BSS1" = 15))
dev.off()

#some QC features

pdf("BSS1_feature_QC.pdf",width=10,height=10)
FeaturePlot(BSS1, reduction = 'spatial', shape.by = "orig.ident", pt.size = 2, features = c("nFeature_RNA","nCount_RNA","reads"),order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9])) + scale_shape_manual(values = c("BSS1" = 15))
dev.off()


pdf("BSS1_Embedding_Grid_PC15_res0.4.pdf",width=6.8,height=5)
DimPlot(object = BSS1, reduction = 'spatial', pt.size = 5, shape.by = "orig.ident") + scale_shape_manual(values = c("BSS1" = 15))
DimPlot(object = BSS1, reduction = 'spatial', pt.size = 5, shape.by = "orig.ident",group.by = "Hclust") + scale_shape_manual(values = c("BSS1" = 15))
dev.off()




#Idents(object = BSS1) <- BSS1@meta.data$Hclust

plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(BSS1, only.pos = TRUE,  logfc.threshold = 0.2)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(cluster , desc(avg_logFC))

write.csv(markers.anno,"BSS1_Grid_allMarker.csv")

saveRDS(BSS1, file = "BSS1_Grid_SeuratObj.RDS")

						


gene_names = c(
"TOP2_DICDI",
"KI67_HUMAN",
"DPOLN_MOUSE",
"DNLI_HAEIN",
"TOLB_PSEA6",
"WDR54_MOUSE",
"ACTN4_RAT",
"CDC20_DICDI",
"DPOLB_RAT",
"RHOQ_HUMAN",
"FORD_DICDI",
"RASA2_MOUSE",
"CDKA1_ORYSJ",
"CCA11_ORYSJ",
"CCNB_DICDI",
"TBB2_PHYPO",
"MCM9_XENTR",
"PATS1_DICDI")
gene_ids = c(
"Phypoly-transcript-00381",
"Phypoly-transcript-00261",
"Phypoly-transcript-01132",
"Phypoly-transcript-14390",
"Phypoly-transcript-04660",
"Phypoly-transcript-13888",
"Phypoly-transcript-00407",
"Phypoly-transcript-07151",
"Phypoly-transcript-04465",
"Phypoly-transcript-23095",
"Phypoly-transcript-02192",
"Phypoly-transcript-06376",
"Phypoly-transcript-06666",
"Phypoly-transcript-05693",
"Phypoly-transcript-08256",
"Phypoly-transcript-06678",
"Phypoly-transcript-00963",
"Phypoly-transcript-00502")
		


gg_Fig <- FeaturePlot(BSS1, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS1" = 15)) })

pdf("BSS1_Grid_feature_Top6ClustMarker.pdf",width=15,height=11)
CombinePlots( gg_Fig )
dev.off()
	
#on PCA
gg_Fig <- FeaturePlot(BSS1, reduction = 'pca',dims = c(2,4),pt.size = 3, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("BSS1_PCAscatter_feature_Top6ClustMarker.pdf",width=15,height=11)
CombinePlots( gg_Fig )
dev.off()

#on heatmap
my_levels <- c(1,0,2)
levels(BSS1) <- my_levels

pdf("BSS1_Heatmap_feature_Top6ClustMarker.pdf.pdf",width=5,height=5)
DoHeatmap(BSS1, features = gene_ids, raster = F, draw.lines = F) + scale_fill_gradientn(colors =  c(rep("black",9),brewer.pal(9,"Blues")[9:1])) + scale_y_discrete(labels = rev(gene_names))
dev.off()


		

gene_names = c("MFSD8_DANRE",
"LIMK1_HUMAN",
"MDMC_STRMY",
"G3BP_SCHPO",
"IP6K3_MOUSE",
"SART3_PONAB",
"Y3934_DICDI",
"RWD2B_MOUSE",
"TSN_PONAB",
"APO1_AGRAE",
"NCS1_SCHPO",
"ARL6_BOVIN",
"SYT7_HUMAN",
"TM86B_RAT",
"RB11B_TOBAC",
"Y093_RHIME",
"CDK20_DANRE",
"YD156_PHANO",
"PPSA_MYCTU",
"ROCO5_DICDI",
"TF29_SCHPO",
"NEG1_NEUCR",
"IPLA_DICDI",
"MUS81_XENTR",
"SMHD1_HUMAN",
"ANK1_HUMAN",
"ROCO6_DICDI",
"PATS1_DICDI",
"FBN1_BOVIN",
"NOTCH_DROME",
"AQP2_HORSE",
"TOP2_DICDI",		
"KI67_HUMAN",
"CDC20_DICDI",
"RAD51_CHICK",
"MCM9_XENTR")	

gene_ids = c("Phypoly-transcript-04492",
"Phypoly-transcript-13062",
"Phypoly-transcript-06472",
"Phypoly-transcript-20941",
"Phypoly-transcript-08709",
"Phypoly-transcript-02705",
"Phypoly-transcript-12045",
"Phypoly-transcript-08273",
"Phypoly-transcript-20272",
"Phypoly-transcript-15552",
"Phypoly-transcript-16350",
"Phypoly-transcript-10238",
"Phypoly-transcript-10933",
"Phypoly-transcript-14708",
"Phypoly-transcript-14205",
"Phypoly-transcript-05606",
"Phypoly-transcript-14602",
"Phypoly-transcript-05766",
"Phypoly-transcript-00211",
"Phypoly-transcript-12239",
"Phypoly-transcript-09973",
"Phypoly-transcript-10029",
"Phypoly-transcript-00996",
"Phypoly-transcript-00664",
"Phypoly-transcript-13710",
"Phypoly-transcript-00953",
"Phypoly-transcript-06536",
"Phypoly-transcript-02110",
"Phypoly-transcript-00834",
"Phypoly-transcript-00580",
"Phypoly-transcript-01487",
"Phypoly-transcript-00381",
"Phypoly-transcript-00261",
"Phypoly-transcript-07151",
"Phypoly-transcript-11446",
"Phypoly-transcript-00963")


gg_Fig <- FeaturePlot(BSS1, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 5.5, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS1" = 15)) })

pdf("BSS1_Grid_feature_ClustMarker.pdf",width=37,height=30)
CombinePlots( gg_Fig )
dev.off()
	
gene_names = c("TOP2_DICDI",		
"KI67_HUMAN",
"CDC20_DICDI",
"DPOLA_XENLA",
"CDC45_DICDI",	
"MCM4_ARATH",
"MCM7_ARATH",
"BUB1_ARATH",
"MCM6_DICDI",
"VGA_BPS13",
"TBAN_PHYPO",
"UBIQP_AGLNE",
"AGD6_ARATH",
"GTAI_DICDI",
"UBC32_ARATH",
"H2B_COCIM",
"FBXW7_MOUSE",
"CCA11_ORYSJ",
"CCNB_DICDI",
"CDKA1_ORYSJ",
"TBB2_PHYPO",
"ARL6_BOVIN",
"PO21_NASVI",
"FBXW7_HUMAN",
"TBB1_PHYPO",
"CDK1_DICDI",
"AURAA_XENLA",
"RAD51_CHICK",
"PATS1_DICDI",
"MUS81_XENTR")
gene_ids = c("Phypoly-transcript-00381",
"Phypoly-transcript-00261",
"Phypoly-transcript-07151",
"Phypoly-transcript-00494",
"Phypoly-transcript-06023",
"Phypoly-transcript-03575",
"Phypoly-transcript-03498",
"Phypoly-transcript-00614",
"Phypoly-transcript-00395",
"Phypoly-transcript-00051",
"Phypoly-transcript-07720",
"Phypoly-transcript-17966",
"Phypoly-transcript-15759",
"Phypoly-transcript-08000",
"Phypoly-transcript-05787",
"Phypoly-transcript-20418",
"Phypoly-transcript-10476",
"Phypoly-transcript-05693",
"Phypoly-transcript-08256",
"Phypoly-transcript-06666",
"Phypoly-transcript-06678",
"Phypoly-transcript-10238",
"Phypoly-transcript-01160",
"Phypoly-transcript-04948",
"Phypoly-transcript-07525",
"Phypoly-transcript-11405",
"Phypoly-transcript-07863",
"Phypoly-transcript-11446",
"Phypoly-transcript-02110",
"Phypoly-transcript-00664")

gg_Fig <- FeaturePlot(BSS1, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 6, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS1" = 15)) })

pdf("BSS1_Grid_feature_CycleMarker.pdf",width=35,height=23)
CombinePlots( gg_Fig )
dev.off()
	
			

#Late Cycle Marker
gene_names = c("TF29_SCHPO","NEG1_NEUCR","IPLA_DICDI","MUS81_XENTR","SMHD1_HUMAN","ANK1_HUMAN","ROCO6_DICDI","PATS1_DICDI","FBN1_BOVIN","AQP2_HORSE","MCM9_XENTR","PMA1_DICDI","TALA1_DICDI","NOTCH_DROME","PHC2_DANRE","LBR_HUMAN","CAP_DICDI","CALM_SCHPO")

gene_ids = c("Phypoly-transcript-09973","Phypoly-transcript-10029","Phypoly-transcript-00996","Phypoly-transcript-00664","Phypoly-transcript-13710","Phypoly-transcript-00953","Phypoly-transcript-06536","Phypoly-transcript-02110","Phypoly-transcript-00834","Phypoly-transcript-01487","Phypoly-transcript-00963","Phypoly-transcript-01492","Phypoly-transcript-03235","Phypoly-transcript-00580","Phypoly-transcript-15399","Phypoly-transcript-09169","Phypoly-transcript-07637","Phypoly-transcript-04224")

gg_Fig <- FeaturePlot(BSS1, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 7.5, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS1" = 15)) })

pdf("BSS1_Grid_feature_LateCycleMarker.pdf",width=39,height=25)
CombinePlots( gg_Fig )
dev.off()
	
gene_names = c("VGA_BPS13",
"TBAN_PHYPO",
"UBIQP_AGLNE",
"AGD6_ARATH",
"GTAI_DICDI",
"UBC32_ARATH",
"H2B_COCIM",
"FBXW7_MOUSE",
"CCA11_ORYSJ",
"CCNB_DICDI",
"CDKA1_ORYSJ",
"TBB2_PHYPO",
"ARL6_BOVIN",
"PO21_NASVI",
"FBXW7_HUMAN",
"TBB1_PHYPO",
"E2FB_ARATH")
gene_ids = c("Phypoly-transcript-00051",
"Phypoly-transcript-07720",
"Phypoly-transcript-17966",
"Phypoly-transcript-15759",
"Phypoly-transcript-08000",
"Phypoly-transcript-05787",
"Phypoly-transcript-20418",
"Phypoly-transcript-10476",
"Phypoly-transcript-05693",
"Phypoly-transcript-08256",
"Phypoly-transcript-06666",
"Phypoly-transcript-06678",
"Phypoly-transcript-10238",
"Phypoly-transcript-01160",
"Phypoly-transcript-04948",
"Phypoly-transcript-07525",
"Phypoly-transcript-07599")							
				
gg_Fig <- FeaturePlot(BSS1, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 7.5, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS1" = 15)) })

pdf("BSS1_Grid_feature_LateCycleMarkerAvgFC.pdf",width=37,height=25)
CombinePlots( gg_Fig )
dev.off()
	

#Selected marker

gene_names = c("MYOD_DICDI",
"MYOM_DICDI",
"MYOJ_DICDI",
"MYOC_DICDI",
"MYO9A_HUMAN",
"MYS2_DICDI",
"MYOH_DICDI",
"MYOF_HUMAN",
"MYOE_DICDI",
"MYOI_DICDI",
"ACTNA_DICDI",
"ACTNA_DICDI",
"ACTN3_MOUSE",
"ACT6_DIPDE",
"ACTN4_CHICK",
"ACTA_PHYPO",
"ACL6B_HUMAN",
"ACTP_ACACA",
"RB11C_DICDI",
"RB11C_DICDI",
"CYP3_CAEEL",
"MSMOB_DICDI",
"TBCD1_HUMAN",
"V246_FOWPN")

gene_ids = c("Phypoly-transcript-21403",
"Phypoly-transcript-01397",
"Phypoly-transcript-04150",
"Phypoly-transcript-00804",
"Phypoly-transcript-29127",
"Phypoly-transcript-12130",
"Phypoly-transcript-00390",
"Phypoly-transcript-00462",
"Phypoly-transcript-01523",
"Phypoly-transcript-00070",
"Phypoly-transcript-00905",
"Phypoly-transcript-02032",
"Phypoly-transcript-02172",
"Phypoly-transcript-02606",
"Phypoly-transcript-04140",
"Phypoly-transcript-10401",
"Phypoly-transcript-10433",
"Phypoly-transcript-20142",
"Phypoly-transcript-15687",
"Phypoly-transcript-29889",
"Phypoly-transcript-20830",
"Phypoly-transcript-15382",
"Phypoly-transcript-13212",
"Phypoly-transcript-20059")

gg_Fig <- FeaturePlot(BSS1, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 3, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS1" = 15)) })

pdf("BSS1_Grid_feature_SelectedMarker.pdf",width=20,height=14)
CombinePlots( gg_Fig )
dev.off()
	


gene_names = c("SR1B_PHYPO",
"SCP1_ASTLP",
"SYYC_CHICK",
"RHP26_SCHPO",
"ADH6_YEAST",
"ELAV2_XENTR",
"CPAS1_DICDI",
"NPHP3_MOUSE",
"RS30_PLAF7",
"YFMJ_BACSU",
"RACH_DICDI",
"MFSD8_DANRE",
"PASDomain",
"YVDB_BACSU",
"MFS10_BOVIN",
"ASNS_SCHPO",
"SPXS4_DICDI")

gene_ids = c("Phypoly-transcript-15654",
"Phypoly-transcript-17763",
"Phypoly-transcript-05209",
"Phypoly-transcript-00778",
"Phypoly-transcript-15025",
"Phypoly-transcript-13086",
"Phypoly-transcript-17606",
"Phypoly-transcript-04537",
"Phypoly-transcript-21996",
"Phypoly-transcript-22446",
"Phypoly-transcript-16297",
"Phypoly-transcript-04492",
"Phypoly-transcript-26096",
"Phypoly-transcript-04627",
"Phypoly-transcript-09824",
"Phypoly-transcript-05971",
"Phypoly-transcript-02158")



gg_Fig <- FeaturePlot(BSS1, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 3, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS1" = 15)) })

pdf("BSS1_Grid_feature_BSS4Marker.pdf",width=20,height=12)
CombinePlots( gg_Fig )
dev.off()
	



gene_names = c("RAB1A_LYMST",
"SYN2_MOUSE",
"CHMP4_DICDI",
"RABD1_ARATH",
"ISP7_SCHPO",
"TPX_PSEAE",
"MAT2B_BOVIN",
"DIMB_DICDI",
"GCH1_DICDI",
"PATL6_ARATH",
"EB1C_ARATH",
"UBE2C_DICDI",
"AURAA_XENLA",
"KIF15_RAT",
"TRB5_ARATH",
"DEK_MOUSE",
"TBAD_PHYPO",
"RAB1_DIPOM",
"KI67_HUMAN",
"SMC3_ARATH",
"SMC1A_MOUSE",
"RTF1_PONAB",
"H2AV1_ORYSJ",
"DYR_HAEIN",
"UCKB_DICDI",
"HBX10_DICDI",
"CC135_CHLRE",
"PLSC_PHYPO",
"IF2B_WHEAT",
"TNIK_MOUSE",
"FEM1B_CHICK",
"IF5A_SPOFR",
"HST_ARATH",
"NOLC1_HUMAN",
"MYS2_DICDI",
"TCTP_USTMA",
"CCNB_DICDI",
"ACTNA_DICDI")
gene_ids = c("Phypoly-transcript-15390",
"Phypoly-transcript-09696",
"Phypoly-transcript-18237",
"Phypoly-transcript-02031",
"Phypoly-transcript-14030",
"Phypoly-transcript-20022",
"Phypoly-transcript-13909",
"Phypoly-transcript-07397",
"Phypoly-transcript-14216",
"Phypoly-transcript-08023",
"Phypoly-transcript-13407",
"Phypoly-transcript-20628",
"Phypoly-transcript-07863",
"Phypoly-transcript-00408",
"Phypoly-transcript-04676",
"Phypoly-transcript-06998",
"Phypoly-transcript-08562",
"Phypoly-transcript-17410",
"Phypoly-transcript-00261",
"Phypoly-transcript-00797",
"Phypoly-transcript-00833",
"Phypoly-transcript-06858",
"Phypoly-transcript-18931",
"Phypoly-transcript-15374",
"Phypoly-transcript-15271",
"Phypoly-transcript-11361",
"Phypoly-transcript-13229",
"Phypoly-transcript-02353",
"Phypoly-transcript-17742",
"Phypoly-transcript-06398",
"Phypoly-transcript-11935",
"Phypoly-transcript-18652",
"Phypoly-transcript-09131",
"Phypoly-transcript-13375",
"Phypoly-transcript-06241",
"Phypoly-transcript-25465",
"Phypoly-transcript-08256",
"Phypoly-transcript-02032")


gg_Fig <- FeaturePlot(BSS1, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 3, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS1" = 15)) })

pdf("BSS1_Grid_feature_AmoebaMarker.pdf",width=30,height=20)
CombinePlots( gg_Fig )
dev.off()
	
#check DE genes between amoeba and BSS1 cell cycle marker
gene_names = c("TBAD_PHYPO",
"TBAN_PHYPO",
"TBB1_PHYPO",
"TBB2_PHYPO",
"TBB4_CAEEL",
"TBE_MOUSE",
"TBG_PHYPA",
"FORA_DICDI",
"PAKA_DICDI",
"KIF1_DICDI")

	
gene_ids = c("Phypoly-transcript-08562",
"Phypoly-transcript-07720",
"Phypoly-transcript-07525",	
"Phypoly-transcript-06678",	
"Phypoly-transcript-04535",
"Phypoly-transcript-07705",
"Phypoly-transcript-08559",
"Phypoly-transcript-07527",
"Phypoly-transcript-00387",	
"Phypoly-transcript-00221")


gg_Fig <- FeaturePlot(BSS1, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 3, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS1" = 15)) })

pdf("BSS1_Grid_feature_DEAmoebaBSS1Marker.pdf",width=18,height=10.5)
CombinePlots( gg_Fig )
dev.off()
	
########get similarity score in correlation to distance

markers = read.csv("BSS1_Grid_allMarker.csv")

BSS1 <- FindVariableFeatures(BSS1, selection.method = "vst", nfeatures = 200)

counts = as.matrix(x = GetAssayData(object = BSS1, assay = "RNA", slot = "scale.data"))
#make scaled values positive
for (i in 1:nrow(counts)) {
  counts[i,] = counts[i,] + abs(min(counts[i,]))
}
counts = counts[rownames(counts) %in% VariableFeatures(object = BSS1),]

cor.matrix = cor(counts)

grid = Embeddings(BSS1, reduction = "spatial")
grid = cbind(grid,Embeddings(BSS1, reduction = "spatial"))
colnames(grid) = c("x1","y1","x2","y2")

library(raster)
grid.dist = pointDistance(grid[,c("x1","y1")], grid[,c("x2","y2")], lonlat = F, allpairs=T)

colnames(grid.dist) = colnames(counts)
rownames(grid.dist) = colnames(counts)

library(reshape2)
plot.dist = melt(grid.dist)
plot.cor = melt(cor.matrix)

plot.all = cbind(plot.dist,plot.cor$value)
colnames(plot.all) = c("x","y","dist","cor")

plot.all = plot.all[plot.all$dist > 0,]

library(ggplot2)

pdf("BSS1_Similarity_VarMarker_CorDist_Scatter.pdf",width=15,height=15)
ggplot( plot.all, aes(x = dist, y = cor) ) +
        geom_point( size = 0.2)  + facet_wrap(.~y) + geom_smooth(method = "lm",size = 0.5) +
        theme (panel.background = element_blank() ) #+
       # theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
       #        panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
        #       legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
dev.off()


pdf("BSS1_Similarity_VarMarker_CorDist_ScatterAll.pdf",width=40,height=15)
ggplot( plot.all, aes(x = dist, y = cor) ) + geom_line(aes(col = y), stat="smooth",method = "lm", formula = y ~ x, size = 1,alpha = 0.5) + scale_color_manual(values = rep("grey40",length(unique(plot.all$y))))  + geom_line(stat="smooth",method = "lm", formula = y ~ x, size = 2, col = "green") + geom_ribbon(stat='smooth', method = "lm", se=TRUE, alpha=0.4,  fill = "green") + theme (panel.background = element_blank() ) #+
       # theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
       #        panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
        #       legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
dev.off()



pdf("BSS1_Similarity_VarMarker_CorDist_DensityAll.pdf",width=10,height=10)
smoothScatter(plot.all$dist, plot.all$cor , nbin = 20,nrpoints = 0, colramp = colorRampPalette(c("white",brewer.pal(9,"Greys")[1:9])))
abline(lm(plot.all[,"cor"] ~ plot.all[,"dist"]), col="green", lty=1, lwd=3)
dev.off()

#calculate r of regression line
rsq <- function (x, y) cor(x, y) ^ 2

rsq(plot.all$dist,plot.all$cor)

summary(lm(plot.all[,"cor"] ~ plot.all[,"dist"]))


#Get a pesudotime

BSS1 = RunUMAP(BSS1,dims = 1:10, min.dist = 0.001, seed.use = 24)

pdf("BSS1_Embedding_UMAP_PC10_res0.5.pdf",width=6.5,height=5)
DimPlot(object = BSS1, reduction = 'umap', pt.size = 5) 
dev.off()


saveRDS(BSS1, file = "BSS1_Grid_SeuratObj.RDS") 


library("destiny")

diff.df = as.data.frame(Embeddings(BSS1, "pca" ),stringsAsFactor = F)
diff.df$pt = 0
diff.df$rank = 0
diff.df = cbind(diff.df,as.data.frame(Embeddings(BSS1, "spatial" ),stringsAsFactor = F))
diff.df$cluster = BSS1@active.ident

DatasetDM <- DiffusionMap(diff.df[diff.df$cluster %in% 0,1:15])
dpt_c0 = DPT(DatasetDM)
DatasetDM <- DiffusionMap(diff.df[diff.df$cluster %in% 1,1:15])
dpt_c1 = DPT(DatasetDM)
DatasetDM <- DiffusionMap(diff.df[diff.df$cluster %in% 2,1:15])
dpt_c2 = DPT(DatasetDM)

pseudotime_BSS_c1 = rev(dpt_c1[random_root(dpt_c1), ])
pseudotime_BSS_c0 = dpt_c0[random_root(dpt_c0), ] + max(pseudotime_BSS_c1)
pseudotime_BSS_c2 = dpt_c2[random_root(dpt_c2), ] + max(pseudotime_BSS_c0) 


diff.df$pt[diff.df$cluster %in% 0] = pseudotime_BSS_c0
diff.df$pt[diff.df$cluster %in% 1] = pseudotime_BSS_c1
diff.df$pt[diff.df$cluster %in% 2] = pseudotime_BSS_c2

#DatasetDM <- DiffusionMap(diff.df[diff.df$Spatial_1 > 10,1:5])
#dpt_g2 = DPT(DatasetDM)
#DatasetDM <- DiffusionMap(diff.df[diff.df$Spatial_1 < 11,1:5])
#dpt_g1 = DPT(DatasetDM)

#pseudotime_BSS_g1 = dpt_g1[random_root(dpt_g1), ]
#pseudotime_BSS_g2 = dpt_g2[random_root(dpt_g2), ] + max(pseudotime_BSS_g1)

#diff.df$pt[diff.df$Spatial_1 > 10] = pseudotime_BSS_g1
#diff.df$pt[diff.df$Spatial_1 < 11] = pseudotime_BSS_g2

#DatasetDM <- DiffusionMap(diff.df[,1:8])
#dpt = DPT(DatasetDM)
#pseudotime_BSS = (dpt[random_root(dpt), ] )

#diff.df$pt = as.numeric(pseudotime_BSS)
diff.df = diff.df[order(diff.df$pt),]
diff.df$rank = (1:nrow(diff.df))

library(ggplot2)
library(RColorBrewer)

pdf("BSS1_Grid_PT.pdf",width=4,height=3)
ggplot(diff.df, aes(Spatial_1,Spatial_2, color=rank)) + geom_point( size=3.5 ,shape = 15)+ xlab("Spatial 1") + ylab("Spatial 2")+scale_colour_gradientn(colors = c(rev(brewer.pal(11,"RdYlBu")[1:11])), space = "rgb", na.value = "grey50", guide = "colourbar") + xlim(min(diff.df$Spatial_1),max(diff.df$Spatial_1))+ylim(min(diff.df$Spatial_2),max(diff.df$Spatial_2)) +theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))
dev.off()


gene_names = c("DUT_SOLLC",
"CCNB_DICDI",
"CDK1_DICDI",
"AURAA_XENLA",
"TOP2_DICDI",
"KI67_HUMAN",
"DPOLN_MOUSE",
"DNLI_HAEIN",
"TOLB_PSEA6",
"WDR54_MOUSE",
"ACTN4_RAT",
"CDC20_DICDI",
"DPOLB_RAT",
"RHOQ_HUMAN",
"FORD_DICDI",
"RASA2_MOUSE",
"CDKA1_ORYSJ",
"CCA11_ORYSJ",
"TBB2_PHYPO",
"MCM9_XENTR",
"PATS1_DICDI")
gene_ids = c("Phypoly-transcript-20038",
"Phypoly-transcript-08256",
"Phypoly-transcript-11405",
"Phypoly-transcript-07863",
"Phypoly-transcript-00381",
"Phypoly-transcript-00261",
"Phypoly-transcript-01132",
"Phypoly-transcript-14390",
"Phypoly-transcript-04660",
"Phypoly-transcript-13888",
"Phypoly-transcript-00407",
"Phypoly-transcript-07151",
"Phypoly-transcript-04465",
"Phypoly-transcript-23095",
"Phypoly-transcript-02192",
"Phypoly-transcript-06376",
"Phypoly-transcript-06666",
"Phypoly-transcript-05693",
"Phypoly-transcript-06678",
"Phypoly-transcript-00963",
"Phypoly-transcript-00502")
		


plot.df = FetchData(BSS1, c("Phypoly-transcript-20038",
"Phypoly-transcript-08256",
"Phypoly-transcript-11405",
"Phypoly-transcript-07863",
"Phypoly-transcript-00381",
"Phypoly-transcript-00261",
"Phypoly-transcript-01132",
"Phypoly-transcript-14390",
"Phypoly-transcript-04660",
"Phypoly-transcript-13888",
"Phypoly-transcript-00407",
"Phypoly-transcript-07151",
"Phypoly-transcript-04465",
"Phypoly-transcript-23095",
"Phypoly-transcript-02192",
"Phypoly-transcript-06376",
"Phypoly-transcript-06666",
"Phypoly-transcript-05693",
"Phypoly-transcript-06678",
"Phypoly-transcript-00963",
"Phypoly-transcript-00502"))


plot.df = cbind(plot.df,diff.df[match(rownames(plot.df),rownames(diff.df)),],stringsAsFactors = F)

plot_list=list()
count = 0
for (i in colnames(plot.df[,gene_ids])) {
  count = count + 1
  plot_list[[i]]=ggplot(plot.df, aes(rank,plot.df[,i])) + geom_smooth(se = FALSE, method='loess',span = 10 ) + scale_color_manual(values = c('#88CCEE', '#44AA99', '#117733', '#332288', '#DDCC77', '#999933','#CC6677', '#882255', '#AA4499', '#DDDDDD')) + theme_bw() + theme(panel.border =  element_blank(), panel.grid.major = element_blank(),panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) + labs(title=gene_names[count])
#+ ylim(min(0),max(plot.df[i]))

}

pdf("./BSS1_PT_CycleExpression.pdf",width=4.5,height=4)

for (i in colnames(plot.df[,gene_ids])) {
    print(plot_list[[i]])
}
dev.off()







```

##BSS2

```{r}
############# load meta data into Seurat

BSS2.meta = BSS2.meta[match(colnames(BSS2),BSS2.meta$ID),]
BSS2.meta$Spatial_1 = BSS2.meta$grid.x
BSS2.meta$Spatial_2 = BSS2.meta$grid.y
dim = as.data.frame(BSS2.meta[,c("Spatial_1","Spatial_2")],stringsAsFactors = F)
rownames(dim) = colnames(BSS2)

myLetters <- toupper(letters[1:26])
dim$Spatial_1 = as.numeric(match(dim$Spatial_1, myLetters))
dim$Spatial_2 = as.numeric(dim$Spatial_2)
dim = as.matrix(dim)

BSS2@meta.data$reads = BSS2.meta$pairedReads

##############

#parallize scaling
library(future)
plan("multiprocess", workers = 8)
plan()

#set maximum to 10GB for each worker
options(future.globals.maxSize = 10000 * 1024^2)

all.genes <- rownames(BSS2)
BSS2 = ScaleData(  BSS2, features = all.genes ,vars.to.regress = c("reads","nFeature_RNA"))

#run PCA on all genes
#RunPCA(object, assay = NULL, features = NULL,  npcs = 50, rev.pca = FALSE, weight.by.var = TRUE, verbose = TRUE,  ndims.print = 1:5, nfeatures.print = 30, reduction.name = "pca",  reduction.key = "PC_", seed.use = 42)

BSS2 = RunPCA(BSS2,features = all.genes,npcs = 50)

#plot PCA heatmaps
#DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

library(RColorBrewer)

pdf("BSS2_PCAheatmaps.pdf",width=30,height=20)
DimHeatmap(BSS2, dims = 1:16, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(BSS2, dims = 17:32, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(BSS2, dims = 33:48, cells = 100, balanced = TRUE,  ncol = 4, fast = F) 
dev.off()

#elbow plot
pdf("BSS2_PCelbow.pdf",width=20,height=10)
ElbowPlot(BSS2, ndims = 100)
dev.off()

BSS2[["spatial"]] <- CreateDimReducObject(embeddings = dim,  key = "Spatial_", assay = "RNA")

#Cluster cells
BSS2<- FindNeighbors(BSS2, reduction = 'pca',dims = 5)
BSS2 <- FindClusters(BSS2, resolution = 0.1)


# Perform Hierarchical Clustering
library(tidyverse)
Data.hc <- BSS2@reductions$pca@cell.embeddings[,c(1:5)] %>% get_dist(method = "pearson") %>% hclust(method = 'ward.D') ## PCs
Hcls <- data.frame(Clusters = as.factor(cutree(Data.hc, k = 2)))
Hcls <- Hcls %>% mutate(Wells = colnames(BSS2))

BSS2@meta.data$Hclust = Hcls$Clusters


BSS2 = RunUMAP(BSS2, dims = 1:5,min.dist = 0.3)

#plot result
#DimPlot(object, dims = c(1, 2), cells = NULL, cols = NULL,  pt.size = NULL, reduction = NULL, group.by = NULL,  split.by = NULL, shape.by = NULL, order = NULL, label = FALSE,  label.size = 4, repel = FALSE, cells.highlight = NULL,  cols.highlight = "red", sizes.highlight = 1, na.value = "grey50",  combine = TRUE)

pdf("BSS2_Embedding_Grid_PC5_res0.1.pdf",width=5.5,height=4)
DimPlot(object = BSS2, reduction = 'spatial', pt.size = 4, shape.by = "orig.ident") + scale_shape_manual(values = c("BSS2" = 15))
DimPlot(object = BSS2, reduction = 'spatial', pt.size = 4, shape.by = "orig.ident",group.by = "Hclust") + scale_shape_manual(values = c("BSS2" = 15))
DimPlot(object = BSS2, reduction = 'umap', pt.size = 4, shape.by = "orig.ident") + scale_shape_manual(values = c("BSS2" = 15))
dev.off()

#some QC features

pdf("BSS2_feature_QC.pdf",width=10,height=5)
FeaturePlot(BSS2, reduction = 'spatial', shape.by = "orig.ident", pt.size = 2, features = c("nFeature_RNA","nCount_RNA","reads"),order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9])) + scale_shape_manual(values = c("BSS2" = 15))
dev.off()


#Idents(object = BSS2) <- BSS2@meta.data$Hclust

plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(BSS2, only.pos = TRUE,  logfc.threshold = 0.2)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(cluster , desc(avg_logFC))

write.csv(markers.anno,"BSS2_Grid_allMarker.csv")

saveRDS(BSS2, file = "BSS2_Grid_SeuratObj.RDS")

gene_names = c("PPI2_SYNY3",
"ADCA_DICDI",
"GAR1_ASHGO",
"GCA_DICDI",
"RAC2_HUMAN",
"SR43C_ARATH",
"AGD6_ARATH",
"TUP1_DICDI",
"PCNA_BRANA",
"VPS15_DICDI",
"TARB1_HUMAN",
"VIS_VIBSP",
"PAN_SULTO",
"GALC_DANRE",
"TENX_HUMAN",
"SPKUL_DICDI",
"GXCJJ_DICDI",
"GACS_PSESY")
gene_ids = c("Phypoly-transcript-14630",
"Phypoly-transcript-06475",
"Phypoly-transcript-16602",
"Phypoly-transcript-01321",
"Phypoly-transcript-28123",
"Phypoly-transcript-10708",
"Phypoly-transcript-15759",
"Phypoly-transcript-16293",
"Phypoly-transcript-13303",
"Phypoly-transcript-00384",
"Phypoly-transcript-06892",
"Phypoly-transcript-00382",
"Phypoly-transcript-00944",
"Phypoly-transcript-03489",
"Phypoly-transcript-00880",
"Phypoly-transcript-00611",
"Phypoly-transcript-03427",
"Phypoly-transcript-01910")



gg_Fig <- FeaturePlot(BSS2, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS2" = 15)) })

pdf("BSS2_Grid_feature_Top9ClustMarker.pdf",width=15,height=11)
CombinePlots( gg_Fig )
dev.off()
	
#on PCA
gg_Fig <- FeaturePlot(BSS2, reduction = 'pca',dims = c(2,4),pt.size = 3, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("BSS2_PCAscatter_feature_Top9ClustMarker.pdf",width=15,height=11)
CombinePlots( gg_Fig )
dev.off()

#on heatmap
#my_levels <- c(1,0,2)
#levels(BSS2) <- my_levels

pdf("BSS2_Heatmap_feature_Top9ClustMarker.pdf.pdf",width=5,height=5)
DoHeatmap(BSS2, features = gene_ids, raster = F, draw.lines = F)  + scale_fill_gradientn(colors =  c(rep("black",9),brewer.pal(9,"Blues")[9:1])) + scale_y_discrete(labels = rev(gene_names))
dev.off()


gene_names = c("TOP2_DICDI","KI67_HUMAN")
gene_ids = c("Phypoly-transcript-00381","Phypoly-transcript-00261")

gg_Fig <- FeaturePlot(BSS2, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 5, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS2" = 15)) })

pdf("BSS2_Grid_feature_MKI67_TOP2A.pdf",width=10,height=4)
CombinePlots( gg_Fig )
dev.off()
	





gene_names = c("ABCG2_DICDI",
"AGLU_BETVU",
"SPG16_HUMAN",
"DNLI_HAEIN",
"KLHL6_MOUSE",
"NMNAT_ORYSJ",
"ABCCA_DICDI",
"AT10B_HUMAN",
"AB14C_ARATH",
"MLT7_CAEEL",
"FBXL2_BOVIN",
"MPKC_NEOFI",
"OGFD2_HUMAN",
"4CLL7_ARATH",
"MPKC_NEOFI",
"GXCDD_DICDI",
"YR831_MIMIV",
"CHDH_HUMAN",
"CLU_MAGO7",
"CDK1_DICDI",
"MEG10_HUMAN",
"TENX_HUMAN",
"TOP2_DICDI",		
"KI67_HUMAN",
"CDC20_DICDI",
"RAD51_CHICK",
"MCM9_XENTR")	

gene_ids = c("Phypoly-transcript-00680",
"Phypoly-transcript-14178",
"Phypoly-transcript-05280",
"Phypoly-transcript-02153",
"Phypoly-transcript-11649",
"Phypoly-transcript-17126",
"Phypoly-transcript-00483",
"Phypoly-transcript-00832",
"Phypoly-transcript-25300",
"Phypoly-transcript-29032",
"Phypoly-transcript-08112",
"Phypoly-transcript-02835",
"Phypoly-transcript-05381",
"Phypoly-transcript-01772",
"Phypoly-transcript-04279",
"Phypoly-transcript-02473",
"Phypoly-transcript-02806",
"Phypoly-transcript-04498",
"Phypoly-transcript-01846",
"Phypoly-transcript-11405",
"Phypoly-transcript-12926",
"Phypoly-transcript-04408",
"Phypoly-transcript-00381",
"Phypoly-transcript-00261",
"Phypoly-transcript-07151",
"Phypoly-transcript-11446",
"Phypoly-transcript-00963")


gg_Fig <- FeaturePlot(BSS2, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 4, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS2" = 15)) })

pdf("BSS2_Grid_feature_ClustMarker.pdf",width=25,height=17)
CombinePlots( gg_Fig )
dev.off()
	

gene_names = c("MYOD_DICDI",
"MYOM_DICDI",
"MYOJ_DICDI",
"MYOC_DICDI",
"MYO9A_HUMAN",
"MYS2_DICDI",
"MYOH_DICDI",
"MYOF_HUMAN",
"MYOE_DICDI",
"MYOI_DICDI",
"ACTNA_DICDI",
"ACTNA_DICDI",
"ACTN3_MOUSE",
"ACT6_DIPDE",
"ACTN4_CHICK",
"ACTA_PHYPO",
"ACL6B_HUMAN",
"ACTP_ACACA",
"RB11C_DICDI",
"RB11C_DICDI",
"CYP3_CAEEL",
"MSMOB_DICDI",
"TBCD1_HUMAN",
"V246_FOWPN")

gene_ids = c("Phypoly-transcript-21403",
"Phypoly-transcript-01397",
"Phypoly-transcript-04150",
"Phypoly-transcript-00804",
"Phypoly-transcript-29127",
"Phypoly-transcript-12130",
"Phypoly-transcript-00390",
"Phypoly-transcript-00462",
"Phypoly-transcript-01523",
"Phypoly-transcript-00070",
"Phypoly-transcript-00905",
"Phypoly-transcript-02032",
"Phypoly-transcript-02172",
"Phypoly-transcript-02606",
"Phypoly-transcript-04140",
"Phypoly-transcript-10401",
"Phypoly-transcript-10433",
"Phypoly-transcript-20142",
"Phypoly-transcript-15687",
"Phypoly-transcript-29889",
"Phypoly-transcript-20830",
"Phypoly-transcript-15382",
"Phypoly-transcript-13212",
"Phypoly-transcript-20059")

gg_Fig <- FeaturePlot(BSS2, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 3, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS2" = 15)) })

pdf("BSS2_Grid_feature_SelectedMarker.pdf",width=19,height=14)
CombinePlots( gg_Fig )
dev.off()
	

gene_names = c("SR1B_PHYPO",
"SCP1_ASTLP",
"SYYC_CHICK",
"RHP26_SCHPO",
"ADH6_YEAST",
"ELAV2_XENTR",
"CPAS1_DICDI",
"NPHP3_MOUSE",
"RS30_PLAF7",
"YFMJ_BACSU",
"RACH_DICDI",
"MFSD8_DANRE",
"PASDomain",
"YVDB_BACSU",
"MFS10_BOVIN",
"ASNS_SCHPO",
"SPXS4_DICDI")

gene_ids = c("Phypoly-transcript-15654",
"Phypoly-transcript-17763",
"Phypoly-transcript-05209",
"Phypoly-transcript-00778",
"Phypoly-transcript-15025",
"Phypoly-transcript-13086",
"Phypoly-transcript-17606",
"Phypoly-transcript-04537",
"Phypoly-transcript-21996",
"Phypoly-transcript-22446",
"Phypoly-transcript-16297",
"Phypoly-transcript-04492",
"Phypoly-transcript-26096",
"Phypoly-transcript-04627",
"Phypoly-transcript-09824",
"Phypoly-transcript-05971",
"Phypoly-transcript-02158")



gg_Fig <- FeaturePlot(BSS2, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 3, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS2" = 15)) })

pdf("BSS2_Grid_feature_BSS4Marker.pdf",width=20,height=12)
CombinePlots( gg_Fig )
dev.off()
	

########get similarity score in correlation to distance

counts = as.matrix(x = GetAssayData(object = BSS2, assay = "RNA", slot = "scale.data"))
#make scaled values positive
for (i in 1:nrow(counts)) {
  counts[i,] = counts[i,] + abs(min(counts[i,]))
}

cor.matrix = cor(counts)

grid = Embeddings(BSS2, reduction = "spatial")
grid = cbind(grid,Embeddings(BSS2, reduction = "spatial"))
colnames(grid) = c("x1","y1","x2","y2")

library(raster)
grid.dist = pointDistance(grid[,c("x1","y1")], grid[,c("x2","y2")], lonlat = F, allpairs=T)

colnames(grid.dist) = colnames(counts)
rownames(grid.dist) = colnames(counts)

library(reshape2)
plot.dist = melt(grid.dist)
plot.cor = melt(cor.matrix)

plot.all = cbind(plot.dist,plot.cor$value)
colnames(plot.all) = c("x","y","dist","cor")

plot.all = plot.all[plot.all$dist > 0,]

library(ggplot2)

pdf("BSS2_Similarity_CorDist_Scatter.pdf",width=15,height=15)
ggplot( plot.all, aes(x = dist, y = cor) ) +
        geom_point( size = 0.2)  + facet_wrap(.~y) + geom_smooth(method = "lm",size = 0.5) +
        theme (panel.background = element_blank() ) #+
       # theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
       #        panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
        #       legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
dev.off()



pdf("BSS2_Similarity_VarMarker_CorDist_ScatterAll.pdf",width=40,height=15)
ggplot( plot.all, aes(x = dist, y = cor) ) + geom_line(aes(col = y), stat="smooth",method = "lm", formula = y ~ x, size = 1,alpha = 0.5) + scale_color_manual(values = rep("grey40",length(unique(plot.all$y))))  + geom_line(stat="smooth",method = "lm", formula = y ~ x, size = 2, col = "green") + geom_ribbon(stat='smooth', method = "lm", se=TRUE, alpha=0.4,  fill = "green") + theme (panel.background = element_blank() ) #+
       # theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
       #        panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
        #       legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
dev.off()



pdf("BSS2_Similarity_VarMarker_CorDist_DensityAll.pdf",width=10,height=10)
smoothScatter(plot.all$dist, plot.all$cor , nbin = 20,nrpoints = 0, colramp = colorRampPalette(c("white",brewer.pal(9,"Greys")[1:9])))
abline(lm(plot.all[,"cor"] ~ plot.all[,"dist"]), col="green", lty=1, lwd=3)
dev.off()

```


##BSS3

```{r}
############# load meta data into Seurat

BSS3.meta = BSS3.meta[match(colnames(BSS3),BSS3.meta$ID),]
BSS3.meta$Spatial_1 = BSS3.meta$grid.x
BSS3.meta$Spatial_2 = BSS3.meta$grid.y
dim = as.data.frame(BSS3.meta[,c("Spatial_1","Spatial_2")],stringsAsFactors = F)
rownames(dim) = colnames(BSS3)

myLetters <- toupper(letters[1:26])
dim$Spatial_1 = as.numeric(match(dim$Spatial_1, myLetters))
dim$Spatial_2 = as.numeric(dim$Spatial_2)
dim = as.matrix(dim)

BSS3@meta.data$reads = BSS3.meta$pairedReads

##############

#parallize scaling
library(future)
plan("multiprocess", workers = 8)
plan()

#set maximum to 10GB for each worker
options(future.globals.maxSize = 10000 * 1024^2)

all.genes <- rownames(BSS3)
BSS3 = ScaleData(  BSS3, features = all.genes ,vars.to.regress = c("reads","nFeature_RNA"))

#run PCA on all genes
#RunPCA(object, assay = NULL, features = NULL,  npcs = 50, rev.pca = FALSE, weight.by.var = TRUE, verbose = TRUE,  ndims.print = 1:5, nfeatures.print = 30, reduction.name = "pca",  reduction.key = "PC_", seed.use = 42)

BSS3 = RunPCA(BSS3,features = all.genes,npcs = 50)

#plot PCA heatmaps
#DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

library(RColorBrewer)

pdf("BSS3_PCAheatmaps.pdf",width=30,height=20)
DimHeatmap(BSS3, dims = 1:16, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(BSS3, dims = 17:32, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(BSS3, dims = 33:48, cells = 100, balanced = TRUE,  ncol = 4, fast = F) 
dev.off()

#elbow plot
pdf("BSS3_PCelbow.pdf",width=20,height=10)
ElbowPlot(BSS3, ndims = 100)
dev.off()

BSS3[["spatial"]] <- CreateDimReducObject(embeddings = dim,  key = "Spatial_", assay = "RNA")

#Cluster cells
BSS3<- FindNeighbors(BSS3, reduction = 'pca',dims = 5)
BSS3 <- FindClusters(BSS3, resolution = 0.3)


# Perform Hierarchical Clustering
library(tidyverse)
Data.hc <- BSS3@reductions$pca@cell.embeddings[,c(1:5)] %>% get_dist(method = "pearson") %>% hclust(method = 'ward.D') ## PCs
Hcls <- data.frame(Clusters = as.factor(cutree(Data.hc, k = 3)))
Hcls <- Hcls %>% mutate(Wells = colnames(BSS3))

BSS3@meta.data$Hclust = Hcls$Clusters


BSS3 = RunUMAP(BSS3, dims = 1:5,min.dist = 0.3)

#plot result
#DimPlot(object, dims = c(1, 2), cells = NULL, cols = NULL,  pt.size = NULL, reduction = NULL, group.by = NULL,  split.by = NULL, shape.by = NULL, order = NULL, label = FALSE,  label.size = 4, repel = FALSE, cells.highlight = NULL,  cols.highlight = "red", sizes.highlight = 1, na.value = "grey50",  combine = TRUE)


Idents(object = BSS3) <- BSS3@meta.data$Hclust

pdf("BSS3_Embedding_Grid_PC5_res0.3.pdf",width=4.5,height=5)
DimPlot(object = BSS3, reduction = 'spatial', pt.size = 5, shape.by = "orig.ident",group.by = "seurat_clusters") + scale_shape_manual(values = c("BSS3" = 15))
DimPlot(object = BSS3, reduction = 'spatial', pt.size = 5, shape.by = "orig.ident",group.by = "Hclust") + scale_shape_manual(values = c("BSS3" = 15))
DimPlot(object = BSS3, reduction = 'umap', pt.size = 5, shape.by = "orig.ident") + scale_shape_manual(values = c("BSS3" = 15))
dev.off()


#some QC features

pdf("BSS3_feature_QC.pdf",width=10,height=10)
FeaturePlot(BSS3, reduction = 'spatial', shape.by = "orig.ident", pt.size = 5, features = c("nFeature_RNA","nCount_RNA","reads"),order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9])) + scale_shape_manual(values = c("BSS3" = 15))
dev.off()


plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(BSS3, only.pos = TRUE,  logfc.threshold = 0.2)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(cluster , desc(avg_logFC))

write.csv(markers.anno,"BSS3_Grid_allMarker.csv")

saveRDS(BSS3, file = "BSS3_Grid_SeuratObj.RDS")

gene_names = c("TRYP_FUSOX",
"MA2B1_BOVIN",
"HIW_DROME",
"PRKAG_DICDI",
"CHFR_DANRE",
"RAC1A_DICDI",
"CALR_DICDI",
"SYVN1_MOUSE",
"IF4G_ORYSJ",
"PPC1B_MOUSE",
"DCAF7_DICDI",
"RAB32_HUMAN",
"EIF3C_ARATH",
"OMS1_YEAST",
"SAP1_ARATH",
"COQ7_DICDI",
"CDIPT_HUMAN",
"IMB1_HUMAN")
gene_ids = c("Phypoly-transcript-14029",
"Phypoly-transcript-14331",
"Phypoly-transcript-05481",
"Phypoly-transcript-11458",
"Phypoly-transcript-13853",
"Phypoly-transcript-16780",
"Phypoly-transcript-08943",
"Phypoly-transcript-03818",
"Phypoly-transcript-01737",
"Phypoly-transcript-25541",
"Phypoly-transcript-07919",
"Phypoly-transcript-24609",
"Phypoly-transcript-01873",
"Phypoly-transcript-16002",
"Phypoly-transcript-12348",
"Phypoly-transcript-14840",
"Phypoly-transcript-17527",
"Phypoly-transcript-02378")



gg_Fig <- FeaturePlot(BSS3, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS3" = 15)) })

pdf("BSS3_Grid_feature_Top6ClustMarker.pdf",width=15,height=11)
CombinePlots( gg_Fig )
dev.off()
	
#on PCA
gg_Fig <- FeaturePlot(BSS3, reduction = 'pca',dims = c(2,4),pt.size = 3, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("BSS3_PCAscatter_feature_Top6ClustMarker.pdf",width=15,height=11)
CombinePlots( gg_Fig )
dev.off()

#on heatmap
#my_levels <- c(1,0,2)
#levels(BSS2) <- my_levels

pdf("BSS3_Heatmap_feature_Top6ClustMarker.pdf.pdf",width=5,height=5)
DoHeatmap(BSS3, features = gene_ids, raster = F, draw.lines = F)  + scale_fill_gradientn(colors =  c(rep("black",9),brewer.pal(9,"Blues")[9:1])) + scale_y_discrete(labels = rev(gene_names))
dev.off()


gene_names = c("TOP2_DICDI","KI67_HUMAN")
gene_ids = c("Phypoly-transcript-00381","Phypoly-transcript-00261")

gg_Fig <- FeaturePlot(BSS3, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 5, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS3" = 15)) })

pdf("BSS3_Grid_feature_MKI67_TOP2A.pdf",width=8,height=5)
CombinePlots( gg_Fig )
dev.off()
	





gene_names = c("PEM2_DICDI",
"SC61G_BRABE",
"DGK1_ASHGO",
"RAB18_DICDI",
"RBP10_DANRE",
"Y381_RICFE",
"LRC40_MOUSE",
"UAF30_SCHPO",
"RAB7A_DICDI",
"ZFPL1_DICDI",
"WASL_RAT",
"KAD2_ORYSJ",
"IMP2_SOLLC",
"GLYR2_ARATH",
"YKK1_SCHPO",
"PARPT_MOUSE",
"CYSP1_DICDI",
"RAB1D_DICDI",
"CYP3_CAEEL",
"SYN3_HUMAN",
"IMA7_BOVIN",
"CYTA_MOUSE",
"TSN_PONAB",
"MAEA_XENLA",
"ICI1_PHAAN",
"UBIE_BRUME",
"COFI_SCHPO",
"CALM_SCHPO",
"TV23A_DICDI",
"DBNL_DICDI",
"PKHF2_CHICK")

gene_ids = c("Phypoly-transcript-20599",
"Phypoly-transcript-29254",
"Phypoly-transcript-13547",
"Phypoly-transcript-14080",
"Phypoly-transcript-13460",
"Phypoly-transcript-17783",
"Phypoly-transcript-14744",
"Phypoly-transcript-12580",
"Phypoly-transcript-17339",
"Phypoly-transcript-14058",
"Phypoly-transcript-12400",
"Phypoly-transcript-06334",
"Phypoly-transcript-15936",
"Phypoly-transcript-12839",
"Phypoly-transcript-15660",
"Phypoly-transcript-23283",
"Phypoly-transcript-19722",
"Phypoly-transcript-21641",
"Phypoly-transcript-20830",
"Phypoly-transcript-08721",
"Phypoly-transcript-15614",
"Phypoly-transcript-24777",
"Phypoly-transcript-20272",
"Phypoly-transcript-12234",
"Phypoly-transcript-17637",
"Phypoly-transcript-11680",
"Phypoly-transcript-19524",
"Phypoly-transcript-29807",
"Phypoly-transcript-16463",
"Phypoly-transcript-20553",
"Phypoly-transcript-23540")




gg_Fig <- FeaturePlot(BSS3, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 4, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS3" = 15)) })

pdf("BSS3_Grid_feature_ClustMarker.pdf",width=22,height=17)
CombinePlots( gg_Fig )
dev.off()
	
gene_names = c("MIOX_DICDI",
"DHPR_DICDI",
"TGDS_DICDI",
"MOXD2_DANRE",
"FRAY2_DICDI",
"CP4B1_RABIT",
"RB11A_DICDI",
"DUPD1_GASAC",
"DUS1B_ARATH",
"GCSH_PHEZH",
"TMEDA_DICDI",
"Y4MH_RHISN",
"ARL6_MOUSE",
"RB11C_DICDI",
"Y8013_DICDI",
"MAEL_XENTR",
"GUNE_CLOTM",
"Y1297_ARCFU",
"RL362_ARATH",
"MYOK_DICDI",
"YP20_BACLI",
"TM14C_HUMAN")

gene_ids = c("Phypoly-transcript-28982",
"Phypoly-transcript-17368",
"Phypoly-transcript-13956",
"Phypoly-transcript-06714",
"Phypoly-transcript-05880",
"Phypoly-transcript-04578",
"Phypoly-transcript-16147",
"Phypoly-transcript-16844",
"Phypoly-transcript-20684",
"Phypoly-transcript-24802",
"Phypoly-transcript-20123",
"Phypoly-transcript-17378",
"Phypoly-transcript-22051",
"Phypoly-transcript-15687",
"Phypoly-transcript-24937",
"Phypoly-transcript-06878",
"Phypoly-transcript-24778",
"Phypoly-transcript-05349",
"Phypoly-transcript-31701",
"Phypoly-transcript-25606",
"Phypoly-transcript-23166",
"Phypoly-transcript-30952")


gg_Fig <- FeaturePlot(BSS3, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 4, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS3" = 15)) })

pdf("BSS3_Grid_feature_ManualClusterMarker.pdf",width=22,height=17)
CombinePlots( gg_Fig )
dev.off()
	

#Selected marker

gene_names = c("MYOD_DICDI",
"MYOM_DICDI",
"MYOJ_DICDI",
"MYOC_DICDI",
"MYO9A_HUMAN",
"MYS2_DICDI",
"MYOH_DICDI",
"MYOF_HUMAN",
"MYOE_DICDI",
"MYOI_DICDI",
"ACTNA_DICDI",
"ACTNA_DICDI",
"ACTN3_MOUSE",
"ACT6_DIPDE",
"ACTN4_CHICK",
"ACTA_PHYPO",
"ACL6B_HUMAN",
"ACTP_ACACA",
"RB11C_DICDI",
"RB11C_DICDI",
"CYP3_CAEEL",
"MSMOB_DICDI",
"TBCD1_HUMAN",
"V246_FOWPN")

gene_ids = c("Phypoly-transcript-21403",
"Phypoly-transcript-01397",
"Phypoly-transcript-04150",
"Phypoly-transcript-00804",
"Phypoly-transcript-29127",
"Phypoly-transcript-12130",
"Phypoly-transcript-00390",
"Phypoly-transcript-00462",
"Phypoly-transcript-01523",
"Phypoly-transcript-00070",
"Phypoly-transcript-00905",
"Phypoly-transcript-02032",
"Phypoly-transcript-02172",
"Phypoly-transcript-02606",
"Phypoly-transcript-04140",
"Phypoly-transcript-10401",
"Phypoly-transcript-10433",
"Phypoly-transcript-20142",
"Phypoly-transcript-15687",
"Phypoly-transcript-29889",
"Phypoly-transcript-20830",
"Phypoly-transcript-15382",
"Phypoly-transcript-13212",
"Phypoly-transcript-20059")

gg_Fig <- FeaturePlot(BSS3, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 5.5, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS3" = 15)) })

pdf("BSS3_Grid_feature_SelectedMarker.pdf",width=20,height=24)
CombinePlots( gg_Fig )
dev.off()
	

gene_names = c("SR1B_PHYPO",
"SCP1_ASTLP",
"SYYC_CHICK",
"RHP26_SCHPO",
"ADH6_YEAST",
"ELAV2_XENTR",
"CPAS1_DICDI",
"NPHP3_MOUSE",
"RS30_PLAF7",
"YFMJ_BACSU",
"RACH_DICDI",
"MFSD8_DANRE",
"PASDomain",
"YVDB_BACSU",
"MFS10_BOVIN",
"ASNS_SCHPO",
"SPXS4_DICDI")

gene_ids = c("Phypoly-transcript-15654",
"Phypoly-transcript-17763",
"Phypoly-transcript-05209",
"Phypoly-transcript-00778",
"Phypoly-transcript-15025",
"Phypoly-transcript-13086",
"Phypoly-transcript-17606",
"Phypoly-transcript-04537",
"Phypoly-transcript-21996",
"Phypoly-transcript-22446",
"Phypoly-transcript-16297",
"Phypoly-transcript-04492",
"Phypoly-transcript-26096",
"Phypoly-transcript-04627",
"Phypoly-transcript-09824",
"Phypoly-transcript-05971",
"Phypoly-transcript-02158")




gg_Fig <- FeaturePlot(BSS3, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 3, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS3" = 15)) })

pdf("BSS3_Grid_feature_BSS4Marker.pdf",width=15,height=12)
CombinePlots( gg_Fig )
dev.off()
	

########get similarity score in correlation to distance



########get similarity score in correlation to distance

counts = as.matrix(x = GetAssayData(object = BSS3, assay = "RNA", slot = "scale.data"))
#make scaled values positive
for (i in 1:nrow(counts)) {
  counts[i,] = counts[i,] + abs(min(counts[i,]))
}

cor.matrix = cor(counts)

grid = Embeddings(BSS3, reduction = "spatial")
grid = cbind(grid,Embeddings(BSS3, reduction = "spatial"))
colnames(grid) = c("x1","y1","x2","y2")

library(raster)
grid.dist = pointDistance(grid[,c("x1","y1")], grid[,c("x2","y2")], lonlat = F, allpairs=T)

colnames(grid.dist) = colnames(counts)
rownames(grid.dist) = colnames(counts)

library(reshape2)
plot.dist = melt(grid.dist)
plot.cor = melt(cor.matrix)

plot.all = cbind(plot.dist,plot.cor$value)
colnames(plot.all) = c("x","y","dist","cor")

plot.all = plot.all[plot.all$dist > 0,]

library(ggplot2)

pdf("BSS3_Similarity_CorDist_Scatter.pdf",width=15,height=15)
ggplot( plot.all, aes(x = dist, y = cor) ) +
        geom_point( size = 0.2)  + facet_wrap(.~y) + geom_smooth(method = "lm",size = 0.5) +
        theme (panel.background = element_blank() ) #+
       # theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
       #        panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
        #       legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
dev.off()



pdf("BSS3_Similarity_VarMarker_CorDist_ScatterAll.pdf",width=40,height=15)
ggplot( plot.all, aes(x = dist, y = cor) ) + geom_line(aes(col = y), stat="smooth",method = "lm", formula = y ~ x, size = 1,alpha = 0.5) + scale_color_manual(values = rep("grey40",length(unique(plot.all$y))))  + geom_line(stat="smooth",method = "lm", formula = y ~ x, size = 2, col = "green") + geom_ribbon(stat='smooth', method = "lm", se=TRUE, alpha=0.4,  fill = "green") + theme (panel.background = element_blank() ) #+
       # theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
       #        panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
        #       legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
dev.off()



pdf("BSS3_Similarity_VarMarker_CorDist_DensityAll.pdf",width=10,height=10)
smoothScatter(plot.all$dist, plot.all$cor , nbin = 20,nrpoints = 0, colramp = colorRampPalette(c("white",brewer.pal(9,"Greys")[1:9])))
abline(lm(plot.all[,"cor"] ~ plot.all[,"dist"]), col="green", lty=1, lwd=3)
dev.off()

```

##BSS4

```{r}
############# load meta data into Seurat

BSS4.meta = BSS4.meta[match(colnames(BSS4),BSS4.meta$ID),]
BSS4.meta$Spatial_1 = BSS4.meta$grid.x
BSS4.meta$Spatial_2 = BSS4.meta$grid.y
dim = as.data.frame(BSS4.meta[,c("Spatial_1","Spatial_2")],stringsAsFactors = F)
rownames(dim) = colnames(BSS4)

myLetters <- toupper(letters[1:26])
dim$Spatial_1 = as.numeric(match(dim$Spatial_1, myLetters))
dim$Spatial_2 = as.numeric(dim$Spatial_2)
dim = as.matrix(dim)

BSS4@meta.data$reads = BSS4.meta$pairedReads

##############

#parallize scaling
library(future)
plan("multiprocess", workers = 8)
plan()

#set maximum to 10GB for each worker
options(future.globals.maxSize = 10000 * 1024^2)

all.genes <- rownames(BSS4)
BSS4 = ScaleData(  BSS4, features = all.genes ,vars.to.regress = c("reads","nFeature_RNA"))

#run PCA on all genes
#RunPCA(object, assay = NULL, features = NULL,  npcs = 50, rev.pca = FALSE, weight.by.var = TRUE, verbose = TRUE,  ndims.print = 1:5, nfeatures.print = 30, reduction.name = "pca",  reduction.key = "PC_", seed.use = 42)

BSS4 = RunPCA(BSS4,features = all.genes,npcs = 50)

#plot PCA heatmaps
#DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

library(RColorBrewer)

pdf("BSS4_PCAheatmaps.pdf",width=30,height=20)
DimHeatmap(BSS4, dims = 1:16, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(BSS4, dims = 17:32, cells = 100, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(BSS4, dims = 33:48, cells = 100, balanced = TRUE,  ncol = 4, fast = F) 
dev.off()

#elbow plot
pdf("BSS4_PCelbow.pdf",width=20,height=10)
ElbowPlot(BSS4, ndims = 100)
dev.off()

BSS4[["spatial"]] <- CreateDimReducObject(embeddings = dim,  key = "Spatial_", assay = "RNA")

#Cluster cells
BSS4<- FindNeighbors(BSS4, reduction = 'pca',dims = 15)
BSS4 <- FindClusters(BSS4, resolution = 0.2)


# Perform Hierarchical Clustering
library(tidyverse)
Data.hc <- BSS4@reductions$pca@cell.embeddings[,c(1:15)] %>% get_dist(method = "pearson") %>% hclust(method = 'ward.D') ## PCs
Hcls <- data.frame(Clusters = as.factor(cutree(Data.hc, k = 3)))
Hcls <- Hcls %>% mutate(Wells = colnames(BSS4))

BSS4@meta.data$Hclust = Hcls$Clusters

BSS4 = RunUMAP(BSS4, dims = 1:15,min.dist = 0.03)

#plot result
#DimPlot(object, dims = c(1, 2), cells = NULL, cols = NULL,  pt.size = NULL, reduction = NULL, group.by = NULL,  split.by = NULL, shape.by = NULL, order = NULL, label = FALSE,  label.size = 4, repel = FALSE, cells.highlight = NULL,  cols.highlight = "red", sizes.highlight = 1, na.value = "grey50",  combine = TRUE)


#plot result
#DimPlot(object, dims = c(1, 2), cells = NULL, cols = NULL,  pt.size = NULL, reduction = NULL, group.by = NULL,  split.by = NULL, shape.by = NULL, order = NULL, label = FALSE,  label.size = 4, repel = FALSE, cells.highlight = NULL,  cols.highlight = "red", sizes.highlight = 1, na.value = "grey50",  combine = TRUE)

pdf("BSS4_Embedding_Grid_PC15_res0.2.pdf",width=5.1,height=4)
DimPlot(object = BSS4, reduction = 'spatial', pt.size = 5, shape.by = "orig.ident", group.by = "seurat_clusters") + scale_shape_manual(values = c("BSS4" = 15))
DimPlot(object = BSS4, reduction = 'spatial', pt.size = 5, shape.by = "orig.ident", group.by = "Hclust") + scale_shape_manual(values = c("BSS4" = 15))
DimPlot(object = BSS4, reduction = 'umap', pt.size = 5, shape.by = "orig.ident", group.by = "Hclust") + scale_shape_manual(values = c("BSS4" = 15))
dev.off()

#some QC features

pdf("BSS4_feature_QC.pdf",width=10,height=10)
FeaturePlot(BSS4, reduction = 'spatial', shape.by = "orig.ident", pt.size = 5, features = c("nFeature_RNA","nCount_RNA","reads"),order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9])) + scale_shape_manual(values = c("BSS4" = 15))
dev.off()

#change active assays

Idents(object = BSS4) <- BSS4@meta.data$Hclust

plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(BSS4, only.pos = TRUE,  logfc.threshold = 0.2)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(cluster , desc(avg_logFC))

write.csv(markers.anno,"BSS4_Grid_allMarker.csv")

####do a biased DE for wells that are around the oat flake

tmp = FetchData(BSS4,c("Spatial_1","Spatial_2"))

oat.wells = rownames(tmp)[tmp$Spatial_1 %in% c(4,5,6,7) & tmp$Spatial_2 %in% c(3,4,5,6)]

markers <- FindMarkers(BSS4, oat.wells, colnames(BSS4)[!colnames(BSS4) %in% oat.wells], only.pos = F, logfc.threshold = 0.2)
markers.anno = markers
markers.anno$gene = rownames(markers.anno)
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene
markers.anno$diff = 0
markers.anno$diff[markers.anno$avg_logFC > 0 ] = markers.anno$pct.1[markers.anno$avg_logFC > 0 ] - markers.anno$pct.2[markers.anno$avg_logFC > 0 ]
markers.anno$diff[markers.anno$avg_logFC < 0 ] = markers.anno$pct.2[markers.anno$avg_logFC < 0 ] - markers.anno$pct.1[markers.anno$avg_logFC < 0 ]

markers.anno = markers.anno  %>% arrange(desc(avg_logFC))

write.csv(markers.anno,"BSS4_Grid_OatMarker_16G.csv")




saveRDS(BSS4, file = "BSS4_Grid_SeuratObj.RDS")

gene_names = c("HSE1_MAGO7",
"CBPB_DICDI",
"AVIL_HUMAN",
"SCP1_ASTLP",
"LIMD2_BOVIN",
"CYAA_ANACY",
"ALKB_ECOLI",
"GTAG_DICDI",
"XPR1_MUSMC",
"ABCGL_DICDI",
"RAB32_MOUSE",
"UFD1_DROME",
"MFSD8_DANRE",
"SPXS4_DICDI",
"CP3A6_RABIT",
"ARUS_PSEAE",
"PLAS_SYNY3",
"RACH_DICDI")
gene_ids = c("Phypoly-transcript-06951",
"Phypoly-transcript-16629",
"Phypoly-transcript-13968",
"Phypoly-transcript-17763",
"Phypoly-transcript-02119",
"Phypoly-transcript-01311",
"Phypoly-transcript-17043",
"Phypoly-transcript-09221",
"Phypoly-transcript-22816",
"Phypoly-transcript-16211",
"Phypoly-transcript-13603",
"Phypoly-transcript-23641",
"Phypoly-transcript-04492",
"Phypoly-transcript-02158",
"Phypoly-transcript-11480",
"Phypoly-transcript-04786",
"Phypoly-transcript-17808",
"Phypoly-transcript-16297")



gg_Fig <- FeaturePlot(BSS4, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS4" = 15)) })

pdf("BSS4_Grid_feature_Top6ClustMarker.pdf",width=15,height=11)
CombinePlots( gg_Fig )
dev.off()
	
#on PCA
gg_Fig <- FeaturePlot(BSS4, reduction = 'pca',dims = c(2,4),pt.size = 3, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("BSS4_PCAscatter_feature_Top6ClustMarker.pdf",width=15,height=11)
CombinePlots( gg_Fig )
dev.off()

#on heatmap
#my_levels <- c(1,0,2)
#levels(BSS2) <- my_levels

pdf("BSS4_Heatmap_feature_Top6ClustMarker.pdf.pdf",width=5,height=5)
DoHeatmap(BSS4, features = gene_ids, raster = F, draw.lines = F)  + scale_fill_gradientn(colors =  c(rep("black",9),brewer.pal(9,"Blues")[9:1])) + scale_y_discrete(labels = rev(gene_names))
dev.off()


gene_names = c("TOP2_DICDI","KI67_HUMAN")
gene_ids = c("Phypoly-transcript-00381","Phypoly-transcript-00261")

gg_Fig <- FeaturePlot(BSS4, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 5, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS4" = 15)) })

pdf("BSS4_Grid_feature_MKI67_TOP2A.pdf",width=9.5,height=4)
CombinePlots( gg_Fig )
dev.off()
	
gene_names = c("SPXS4_DICDI",
"CHX17_ARATH",
"ARUS_PSEAE",
"APRX_BACSU",
"TRI23_CAEEL",
"SPKA_DICDI",
"P2XD_DICDI",
"RAS_BOTFU",
"AFC2_ARATH",
"PLAS_SYNY3",
"AGAL_COFAR",
"UDB10_HUMAN",
"DHKI_DICDI",
"RAB5B_DICDI",
"TOP1_DICDI",
"H33_VITVI",
"RHG39_HUMAN",
"GCY1_YEAST",
"ELAV2_XENTR",
"Y2030_MYCTU",
"Y7588_DICDI",
"MCH1_HANAN",
"NCLDA_DANRE",
"SCP1_ASTLP")
gene_ids = c("Phypoly-transcript-02158",
"Phypoly-transcript-13802",
"Phypoly-transcript-04786",
"Phypoly-transcript-04762",
"Phypoly-transcript-04528",
"Phypoly-transcript-26316",
"Phypoly-transcript-21294",
"Phypoly-transcript-13734",
"Phypoly-transcript-10195",
"Phypoly-transcript-17808",
"Phypoly-transcript-07886",
"Phypoly-transcript-07344",
"Phypoly-transcript-14164",
"Phypoly-transcript-18571",
"Phypoly-transcript-22037",
"Phypoly-transcript-26166",
"Phypoly-transcript-08110",
"Phypoly-transcript-15518",
"Phypoly-transcript-13086",
"Phypoly-transcript-07503",
"Phypoly-transcript-25065",
"Phypoly-transcript-15080",
"Phypoly-transcript-17862",
"Phypoly-transcript-17763")


gg_Fig <- FeaturePlot(BSS4, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS4" = 15)) })

pdf("BSS4_Grid_feature_OatMarker_16G.pdf",width=15,height=11)
CombinePlots( gg_Fig )
dev.off()
	
gene_names = c("SPXS4_DICDI",
"APRX_BACSU",
"ENGB_GLUOX",
"VILI_CHICK",
"PLAS_SYNY3",
"AFC2_ARATH",
"XYNZ_CLOTH",
"MFSD8_DANRE",
"TRI23_CAEEL",
"FBLN1_CHICK",
"CP3A6_RABIT",
"ARMC9_BOVIN",
"MYBD_DICDI",
"POC1A_DANRE",
"VILD_DICDI",
"RAC1_RAT",
"SCP1_ASTLP",
"GRRA_EMENI",
"LIP_RHIOR",
"PRPC_ECOLI",
"ELAV2_XENTR",
"KRP95_STRPU",
"SCAR_DICDI",
"DBNL_DICDI")
gene_ids = c("Phypoly-transcript-02158",
"Phypoly-transcript-04762",
"Phypoly-transcript-09296",
"Phypoly-transcript-18710",
"Phypoly-transcript-17808",
"Phypoly-transcript-07824",
"Phypoly-transcript-12093",
"Phypoly-transcript-04492",
"Phypoly-transcript-04528",
"Phypoly-transcript-01236",
"Phypoly-transcript-11480",
"Phypoly-transcript-17948",
"Phypoly-transcript-07739",
"Phypoly-transcript-04674",
"Phypoly-transcript-08399",
"Phypoly-transcript-17792",
"Phypoly-transcript-17763",
"Phypoly-transcript-04661",
"Phypoly-transcript-13234",
"Phypoly-transcript-11006",
"Phypoly-transcript-13086",
"Phypoly-transcript-01362",
"Phypoly-transcript-10617",
"Phypoly-transcript-20553")

gg_Fig <- FeaturePlot(BSS4, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 2, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS4" = 15)) })

pdf("BSS4_Grid_feature_OatMarker_9G.pdf",width=15,height=11)
CombinePlots( gg_Fig )
dev.off()
	

#volcano plot
library(EnhancedVolcano)

pdf("BSS4_Grid_Volcano_OatMarker_16G.pdf",width=10,height=10)
EnhancedVolcano(markers.anno,
    lab = markers.anno[,"UniProt_Acc"],
    x = 'avg_logFC',
    y = 'p_val',
    xlim = c(-4.5,4.5),
    pCutoff = 0.001,
    FCcutoff = 1.1,
    drawConnectors = T,
    widthConnectors = 0.1,
    xlab = "avg_logFC",
    ylab = "p_val",
    gridlines.major = F,
    gridlines.minor = F,    
    col=c('black', 'black', 'black', 'red3'),
    colAlpha = 0.6,
    transcriptPointSize = 2)

EnhancedVolcano(markers.anno,
    lab = markers.anno[,"UniProt_Acc"],
    x = 'avg_logFC',
    y = 'p_val',
    xlim = c(-4.5,4.5),
    pCutoff = 0.001,
    FCcutoff = 1.1,
    #drawConnectors = T,
    #widthConnectors = 0.5,
    xlab = "avg_logFC",
    ylab = "p_val",
    gridlines.major = F,
    gridlines.minor = F,    
    col=c('black', 'black', 'black', 'red3'),
    colAlpha = 0.6,
    transcriptPointSize = 2)
dev.off()

  

gene_names = c("SR1B_PHYPO",
"SCP1_ASTLP",
"SYYC_CHICK",
"RHP26_SCHPO",
"ADH6_YEAST",
"ELAV2_XENTR",
"CPAS1_DICDI",
"NPHP3_MOUSE",
"RS30_PLAF7",
"YFMJ_BACSU",
"RACH_DICDI",
"MFSD8_DANRE",
"PASDomain",
"YVDB_BACSU",
"MFS10_BOVIN",
"ASNS_SCHPO",
"SPXS4_DICDI")

gene_ids = c("Phypoly-transcript-15654",
"Phypoly-transcript-17763",
"Phypoly-transcript-05209",
"Phypoly-transcript-00778",
"Phypoly-transcript-15025",
"Phypoly-transcript-13086",
"Phypoly-transcript-17606",
"Phypoly-transcript-04537",
"Phypoly-transcript-21996",
"Phypoly-transcript-22446",
"Phypoly-transcript-16297",
"Phypoly-transcript-04492",
"Phypoly-transcript-26096",
"Phypoly-transcript-04627",
"Phypoly-transcript-09824",
"Phypoly-transcript-05971",
"Phypoly-transcript-02158")




gg_Fig <- FeaturePlot(BSS4, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 3, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS4" = 15)) })

pdf("BSS4_Grid_feature_ManualClusterMarker.pdf",width=18,height=11)
CombinePlots( gg_Fig )
dev.off()
	

#Selected marker

gene_names = c("MYOD_DICDI",
"MYOM_DICDI",
"MYOJ_DICDI",
"MYOC_DICDI",
"MYO9A_HUMAN",
"MYS2_DICDI",
"MYOH_DICDI",
"MYOF_HUMAN",
"MYOE_DICDI",
"MYOI_DICDI",
"ACTNA_DICDI",
"ACTNA_DICDI",
"ACTN3_MOUSE",
"ACT6_DIPDE",
"ACTN4_CHICK",
"ACTA_PHYPO",
"ACL6B_HUMAN",
"ACTP_ACACA",
"RB11C_DICDI",
"RB11C_DICDI",
"CYP3_CAEEL",
"MSMOB_DICDI",
"TBCD1_HUMAN",
"V246_FOWPN")

gene_ids = c("Phypoly-transcript-21403",
"Phypoly-transcript-01397",
"Phypoly-transcript-04150",
"Phypoly-transcript-00804",
"Phypoly-transcript-29127",
"Phypoly-transcript-12130",
"Phypoly-transcript-00390",
"Phypoly-transcript-00462",
"Phypoly-transcript-01523",
"Phypoly-transcript-00070",
"Phypoly-transcript-00905",
"Phypoly-transcript-02032",
"Phypoly-transcript-02172",
"Phypoly-transcript-02606",
"Phypoly-transcript-04140",
"Phypoly-transcript-10401",
"Phypoly-transcript-10433",
"Phypoly-transcript-20142",
"Phypoly-transcript-15687",
"Phypoly-transcript-29889",
"Phypoly-transcript-20830",
"Phypoly-transcript-15382",
"Phypoly-transcript-13212",
"Phypoly-transcript-20059")

gg_Fig <- FeaturePlot(BSS4, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 3, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) + scale_shape_manual(values = c("BSS4" = 15)) })

pdf("BSS4_Grid_feature_SelectedMarker.pdf",width=18,height=14)
CombinePlots( gg_Fig )
dev.off()
	

########get similarity score in correlation to distance

markers = read.csv("BSS4_Grid_allMarker.csv")

BSS4 <- FindVariableFeatures(BSS4, selection.method = "vst", nfeatures = 200)

counts = as.matrix(x = GetAssayData(object = BSS4, assay = "RNA", slot = "scale.data"))
#make scaled values positive
for (i in 1:nrow(counts)) {
  counts[i,] = counts[i,] + abs(min(counts[i,]))
}
counts = counts[rownames(counts) %in% VariableFeatures(object = BSS4),]

cor.matrix = cor(counts)

grid = Embeddings(BSS4, reduction = "spatial")
grid = cbind(grid,Embeddings(BSS4, reduction = "spatial"))
colnames(grid) = c("x1","y1","x2","y2")

library(raster)
grid.dist = pointDistance(grid[,c("x1","y1")], grid[,c("x2","y2")], lonlat = F, allpairs=T)

colnames(grid.dist) = colnames(counts)
rownames(grid.dist) = colnames(counts)

library(reshape2)
plot.dist = melt(grid.dist)
plot.cor = melt(cor.matrix)

plot.all = cbind(plot.dist,plot.cor$value)
colnames(plot.all) = c("x","y","dist","cor")

plot.all = plot.all[plot.all$dist > 0,]

library(ggplot2)

pdf("BSS4_Similarity_VarMarker_CorDist_Scatter.pdf",width=15,height=15)
ggplot( plot.all, aes(x = dist, y = cor) ) +
        geom_point( size = 0.2)  + facet_wrap(.~y) + geom_smooth(method = "lm",size = 0.5) +
        theme (panel.background = element_blank() ) #+
       # theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
       #        panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
        #       legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
dev.off()


pdf("BSS4_Similarity_VarMarker_CorDist_ScatterAll.pdf",width=40,height=15)
ggplot( plot.all, aes(x = dist, y = cor) ) + geom_line(aes(col = y), stat="smooth",method = "lm", formula = y ~ x, size = 1,alpha = 0.5) + scale_color_manual(values = rep("grey40",length(unique(plot.all$y))))  + geom_line(stat="smooth",method = "lm", formula = y ~ x, size = 2, col = "green") + geom_ribbon(stat='smooth', method = "lm", se=TRUE, alpha=0.4,  fill = "green") + theme (panel.background = element_blank() ) #+
       # theme( legend.position = "bottom", axis.title = element_blank(), legend.direction = "horizontal",
       #        panel.border = element_blank(), axis.ticks = element_blank(), axis.text = element_blank(),
        #       legend.title = element_blank(), plot.title = element_text(hjust = 0.5, vjust = -0.5) )
dev.off()



pdf("BSS4_Similarity_VarMarker_CorDist_DensityAll.pdf",width=10,height=10)
smoothScatter(plot.all$dist, plot.all$cor , nbin = 20,nrpoints = 0, colramp = colorRampPalette(c("white",brewer.pal(9,"Greys")[1:9])))
abline(lm(plot.all[,"cor"] ~ plot.all[,"dist"]), col="green", lty=1, lwd=3)
dev.off()


#calculate r of regression line
rsq <- function (x, y) cor(x, y) ^ 2

rsq(plot.all$dist,plot.all$cor)

summary(lm(plot.all[,"cor"] ~ plot.all[,"dist"]))


```


##Integrate all plates to identify common patterns

```{r}

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/")

BSS1 = readRDS(file = "BSS1_Grid_SeuratObj.RDS")
BSS2 = readRDS(file = "BSS2_Grid_SeuratObj.RDS")
BSS3 = readRDS(file = "BSS3_Grid_SeuratObj.RDS")
BSS4 = readRDS(file = "BSS4_Grid_SeuratObj.RDS")

#merge ojects
#merge(x = NULL, y = NULL, add.cell.ids = NULL, merge.data = TRUE, project = "SeuratProject", ...)

BSS = merge(BSS1, y = c(BSS2,BSS3,BSS4), add.cell.ids = c("BSS1","BSS2","BSS3","BSS4"), project = "SpatialTranscriptomicsAllBSS")

```

Run Seurat

```{r}
#scale
#ScaleData(  object,  features = NULL,  assay = NULL,  vars.to.regress = NULL,  split.by = NULL,  model.use = "linear",  use.umi = FALSE,  do.scale = TRUE,  do.center = TRUE,  scale.max = 10,  block.size = 1000,  min.cells.to.block = 3000,  verbose = TRUE)

all.genes <- rownames(BSS)
BSS = ScaleData(  BSS, features = all.genes ,vars.to.regress = c("reads","nFeature_RNA"))

#run PCA on all genes
#RunPCA(object, assay = NULL, features = NULL,  npcs = 50, rev.pca = FALSE, weight.by.var = TRUE, verbose = TRUE,  ndims.print = 1:5, nfeatures.print = 30, reduction.name = "pca",  reduction.key = "PC_", seed.use = 42)

BSS = RunPCA(BSS,features = all.genes,npcs = 100)

#plot PCA heatmaps
#DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

library(RColorBrewer)

pdf("BSS_PCAheatmaps.pdf",width=30,height=20)
DimHeatmap(BSS, dims = 1:16, cells = 1000, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(BSS, dims = 17:32, cells = 1000, balanced = TRUE, ncol = 4,  fast = F) 
DimHeatmap(BSS, dims = 33:48, cells = 1000, balanced = TRUE,  ncol = 4, fast = F) 
DimHeatmap(BSS, dims = 49:64, cells = 1000, balanced = TRUE,  ncol = 4, fast = F) 
DimHeatmap(BSS, dims = 65:80, cells = 1000, balanced = TRUE,  ncol = 4, fast = F) 
DimHeatmap(BSS, dims = 81:96, cells = 1000, balanced = TRUE,  ncol = 4, fast = F) 
dev.off()


#PCA plot
pdf("BSS_PCA.pdf",width=6,height=5)
DimPlot(object = BSS, reduction = 'pca', dims = 1:2, pt.size = 2,group.by = "orig.ident",cols = c('#88CCEE', '#44AA99', '#999933','#CC6677','#117733', '#332288', '#DDCC77',  '#882255', '#AA4499', '#DDDDDD'))
DimPlot(object = BSS, reduction = 'pca', dims = c(1,3), pt.size = 2,group.by = "orig.ident",cols = c('#88CCEE', '#44AA99', '#999933','#CC6677','#117733', '#332288', '#DDCC77',  '#882255', '#AA4499', '#DDDDDD'))
DimPlot(object = BSS, reduction = 'pca', dims = 2:3, pt.size = 2,group.by = "orig.ident",cols = c('#88CCEE', '#44AA99', '#999933','#CC6677','#117733', '#332288', '#DDCC77',  '#882255', '#AA4499', '#DDDDDD'))
dev.off()


#elbow plot
pdf("BSS_PCelbow.pdf",width=20,height=10)
ElbowPlot(BSS, ndims = 100)
dev.off()



#run UMAP
#min.dist deafult is 0.3. Can go down to 0.001. Play around to modify clustering
BSS = RunUMAP(BSS,dims = 1:5, min.dist = 0.3)

#Cluster cells
BSS<- FindNeighbors(BSS, dims = 1:5)
BSS <- FindClusters(BSS, resolution = 0.2)


# Perform Hierarchical Clustering
library(tidyverse)
Data.hc <- BSS@reductions$pca@cell.embeddings[,c(1:5)] %>% get_dist(method = "pearson") %>% hclust(method = 'ward.D') ## PCs
Hcls <- data.frame(Clusters = as.factor(cutree(Data.hc, k = 4)))
Hcls <- Hcls %>% mutate(Wells = colnames(BSS))

BSS@meta.data$Hclust = Hcls$Clusters

#plot result
#DimPlot(object, dims = c(1, 2), cells = NULL, cols = NULL,  pt.size = NULL, reduction = NULL, group.by = NULL,  split.by = NULL, shape.by = NULL, order = NULL, label = FALSE,  label.size = 4, repel = FALSE, cells.highlight = NULL,  cols.highlight = "red", sizes.highlight = 1, na.value = "grey50",  combine = TRUE)

pdf("BSS_Embedding_PC5_res0.2.pdf",width=10,height=10)
DimPlot(object = BSS, reduction = 'umap', pt.size = 2)
DimPlot(object = BSS, reduction = 'umap', pt.size = 2,group.by = "orig.ident",cols = c('#88CCEE', '#44AA99', '#999933','#CC6677','#117733', '#332288', '#DDCC77',  '#882255', '#AA4499', '#DDDDDD'))
DimPlot(object = BSS, reduction = 'umap', pt.size = 2,group.by = "Hclust")
dev.off()

#some QC features

pdf("BSS_feature_QC.pdf",width=10,height=10)
FeaturePlot(BSS, reduction = 'umap', pt.size = 0.5, features = c("nFeature_RNA","nCount_RNA","reads"),order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]))
dev.off()


plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(BSS, only.pos = TRUE,  logfc.threshold = 0.2)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(cluster , desc(avg_logFC))

write.csv(markers.anno,"BSS_allMarker.csv")

saveRDS(BSS, file = "BSS_SeuratObj.RDS")

gene_names = c("KTHY_HUMAN","KI67_HUMAN","TOP2_DICDI","KIF22_DANRE","TRA1_DICDI","POLR_DROME","POL_BAEVM","NHLC2_CHICK","CADH3_ARATH","ADH6_YEAST","UBIQP_AGLNE","UBIQP_AGLNE","CHDH_HUMAN","ASPM_PONPY","FB293_ARATH","CDC20_DICDI","RACH_DICDI","M2K1_ORYSJ","MFSD8_DANRE","YKJ7_SCHPO","GCA_DICDI","Y723_CHLP8")

gene_ids = c("Phypoly-transcript-16291","Phypoly-transcript-00261","Phypoly-transcript-00381","Phypoly-transcript-03486","Phypoly-transcript-25566","Phypoly-transcript-22123","Phypoly-transcript-06623","Phypoly-transcript-07740","Phypoly-transcript-10944","Phypoly-transcript-15025","Phypoly-transcript-17966","Phypoly-transcript-25680","Phypoly-transcript-05350","Phypoly-transcript-01017","Phypoly-transcript-18441","Phypoly-transcript-17752","Phypoly-transcript-16297","Phypoly-transcript-10910","Phypoly-transcript-04492","Phypoly-transcript-14329","Phypoly-transcript-01321","Phypoly-transcript-21401")


gg_Fig <- FeaturePlot(BSS, pt.size = 1, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("BSS_feature_ClustMarker.pdf",width=22,height=20)
CombinePlots( gg_Fig )
dev.off()


plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindMarkers(BSS, c(1), c(2,3),   logfc.threshold = 0.5)
markers.anno = markers
markers.anno$gene = rownames(markers.anno)
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(desc(avg_logFC))

write.csv(markers.anno,"BSS_BSS2vsBSS3-4.csv")

gene_names = c("LBR_HUMAN","SHKD_DICDI","CALM_CHLRE","PSD3_SCHPO","CANB1_DICDI","PSD_PSYA2","DIS1D_DICDI","GRP_HORVU","AGD6_ARATH","NUOC_RHOCA","GEFA_DICDI","CYB_MARPO","NU2M_CHOCR","AT5G2_HUMAN","NU4M_MARPO")

gene_ids = c("Phypoly-transcript-09169","Phypoly-transcript-06719","Phypoly-transcript-05190","Phypoly-transcript-07627","Phypoly-transcript-06944","Phypoly-transcript-23820","Phypoly-transcript-07291","Phypoly-transcript-09768","Phypoly-transcript-15759","Phypoly-transcript-00269","Phypoly-transcript-21022","Phypoly-transcript-03845","Phypoly-transcript-02965","Phypoly-transcript-15172","Phypoly-transcript-00920")

gg_Fig <- FeaturePlot(BSS, pt.size = 1, features = gene_ids,order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T)
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("BSS_feature_BSS2vsBSS3-4_Marker.pdf",width=22,height=20)
CombinePlots( gg_Fig )
dev.off()


plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindMarkers(BSS, c(1), c(2),   logfc.threshold = 0.5)
markers.anno = markers
markers.anno$gene = rownames(markers.anno)
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(desc(avg_logFC))

write.csv(markers.anno,"BSS_BSS2vsBSS4.csv")

gene_names = c("LBR_HUMAN","SHKD_DICDI","CALM_CHLRE","PSD3_SCHPO","CANB1_DICDI","PSD_PSYA2","DIS1D_DICDI","GRP_HORVU","AGD6_ARATH","NUOC_RHOCA","GEFA_DICDI","CYB_MARPO","NU2M_CHOCR","AT5G2_HUMAN","NU4M_MARPO","GUB_CLOTM","CADH1_ARATH","RHP26_SCHPO","CADH3_ARATH","TDPZ4_MOUSE","ADH6_YEAST","RB11B_TOBAC","NHLC2_CHICK","ROC11_DICDI","DNLI_HAEIN","HDNO_ARTOX","TENX_HUMAN")

gene_ids = c("Phypoly-transcript-09169","Phypoly-transcript-06719","Phypoly-transcript-05190","Phypoly-transcript-07627","Phypoly-transcript-06944","Phypoly-transcript-23820","Phypoly-transcript-07291","Phypoly-transcript-09768","Phypoly-transcript-15759","Phypoly-transcript-00269","Phypoly-transcript-21022","Phypoly-transcript-03845","Phypoly-transcript-02965","Phypoly-transcript-15172","Phypoly-transcript-00920","Phypoly-transcript-03644","Phypoly-transcript-10102","Phypoly-transcript-00778","Phypoly-transcript-10944","Phypoly-transcript-09740","Phypoly-transcript-15025","Phypoly-transcript-18273","Phypoly-transcript-07740","Phypoly-transcript-07081","Phypoly-transcript-00504","Phypoly-transcript-06344","Phypoly-transcript-04408")

gg_Fig <- FeaturePlot(BSS, pt.size = 1, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T)
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("BSS_feature_BSS2vsBSS3_Marker.pdf",width=22,height=20)
CombinePlots( gg_Fig )
dev.off()

gene_names = c("YG31B_YEAST",
"DHKL_DICDI",
"SR43C_ARATH",
"CATA2_ARATH",
"GEFA_DICDI",
"DDI1_KLULA",
"POLX_TOBAC",
"NCPR_CAVPO",
"POL3_DROME",
"PKSJ_BACSU",
"TEA1_SCHPO")
gene_ids = c("Phypoly-transcript-14988",
"Phypoly-transcript-01520",
"Phypoly-transcript-10708",
"Phypoly-transcript-04361",
"Phypoly-transcript-21022",
"Phypoly-transcript-14523",
"Phypoly-transcript-13059",
"Phypoly-transcript-00473",
"Phypoly-transcript-29968",
"Phypoly-transcript-00079",
"Phypoly-transcript-00353")


gg_Fig <- FeaturePlot(BSS, pt.size = 1, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("BSS_feature_BSS3Marker.pdf",width=22,height=18)
CombinePlots( gg_Fig )
dev.off()

gene_names = c("CDC20_DICDI","KIF3B_MOUSE","BUB1_ARATH","KI67_HUMAN","CDK1_DICDI","TOP2_DICDI")
gene_ids = c("Phypoly-transcript-17752","Phypoly-transcript-02804","Phypoly-transcript-00614","Phypoly-transcript-00261","Phypoly-transcript-11405","Phypoly-transcript-29539")

gg_Fig <- FeaturePlot(BSS, pt.size = 1, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("BSS_feature_BSS1Marker.pdf",width=22,height=18)
CombinePlots( gg_Fig )
dev.off()


gene_names = c("CDC20_DICDI",
               "KIF3B_MOUSE",
               "BUB1_ARATH",
               "KI67_HUMAN",
               "CDK1_DICDI",
"MK67I_XENTR",
"CAND_DROME",
"NDH2_YARLI",
"GXCDD_DICDI",
"PSD_PSYA2",
"GCA_DICDI",
"TOP2_DICDI",

"YG31B_YEAST",
"DHKL_DICDI",
"SR43C_ARATH",
"CATA2_ARATH",
"GEFA_DICDI",
"DDI1_KLULA",
"POLX_TOBAC",
"NCPR_CAVPO",
"POL3_DROME",
"PKSJ_BACSU",
"TEA1_SCHPO",
"LAMA_DROME",
"CADH3_ARATH",
"ADH6_YEAST",
"RACH_DICDI",
"MFSD8_DANRE")
			

gene_ids = c("Phypoly-transcript-17752","Phypoly-transcript-02804","Phypoly-transcript-00614","Phypoly-transcript-00261","Phypoly-transcript-11405",
"Phypoly-transcript-07117",
"Phypoly-transcript-15303",
"Phypoly-transcript-07101",
"Phypoly-transcript-05895",
"Phypoly-transcript-23820",
"Phypoly-transcript-01321",
"Phypoly-transcript-29539","Phypoly-transcript-14988",
"Phypoly-transcript-01520",
"Phypoly-transcript-10708",
"Phypoly-transcript-04361",
"Phypoly-transcript-21022",
"Phypoly-transcript-14523",
"Phypoly-transcript-13059",
"Phypoly-transcript-00473",
"Phypoly-transcript-29968",
"Phypoly-transcript-00079",
"Phypoly-transcript-00353",
"Phypoly-transcript-00008",
"Phypoly-transcript-10944",
"Phypoly-transcript-15025",
"Phypoly-transcript-16297",
"Phypoly-transcript-04492")

 
gg_Fig <-VlnPlot(BSS, features = gene_ids,pt.size = -1,group.by = "orig.ident")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x]) })

pdf("BSS_Marker_Vln.pdf",width=10,height=10)
CombinePlots( gg_Fig )
DotPlot(BSS, features = gene_ids,group.by = "orig.ident") +scale_colour_gradientn(colors =  c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9])) + theme(axis.text.x = element_text(angle = 90))
dev.off()


#plot heatmap

PCgenes = as.data.frame(Loadings(BSS[['pca']])[,1:10])
PCgenes.list = unique(c(head(rownames(PCgenes)[order(PCgenes$PC_1)],30),tail(rownames(PCgenes)[order(PCgenes$PC_1)],10), head(rownames(PCgenes)[order(PCgenes$PC_2)],10),tail(rownames(PCgenes)[order(PCgenes$PC_2)],10), head(rownames(PCgenes)[order(PCgenes$PC_3)],10),tail(rownames(PCgenes)[order(PCgenes$PC_3)],10), head(rownames(PCgenes)[order(PCgenes$PC_4)],10),tail(rownames(PCgenes)[order(PCgenes$PC_4)],10), head(rownames(PCgenes)[order(PCgenes$PC_5)],10),tail(rownames(PCgenes)[order(PCgenes$PC_5)],10) , head(rownames(PCgenes)[order(PCgenes$PC_6)],10),tail(rownames(PCgenes)[order(PCgenes$PC_6)],10),
head(rownames(PCgenes)[order(PCgenes$PC_7)],10),tail(rownames(PCgenes)[order(PCgenes$PC_7)],10) , head(rownames(PCgenes)[order(PCgenes$PC_8)],10),tail(rownames(PCgenes)[order(PCgenes$PC_8)],10),
head(rownames(PCgenes)[order(PCgenes$PC_9)],10),tail(rownames(PCgenes)[order(PCgenes$PC_9)],10) , head(rownames(PCgenes)[order(PCgenes$PC_10)],10),tail(rownames(PCgenes)[order(PCgenes$PC_10)],10) ))

tmp = anno.gene[anno.gene$transcript %in% PCgenes.list,c("UniProt_Acc","transcript")]
PCgenes.list = intersect(PCgenes.list,tmp$transcript)
tmp = tmp[!tmp$UniProt_Acc %in% "noGene",]
PCgenes.list = intersect(PCgenes.list,tmp$transcript)
tmp = tmp[match(PCgenes.list,tmp$transcript),]

pdf("BSS_PCloadings_1_5.pdf",width=5,height=20)
DoHeatmap(BSS, features = PCgenes.list, group.by = "orig.ident",raster = F, draw.lines = F) + NoLegend() + scale_fill_gradientn(colors =  c("white",(brewer.pal(9,"Greys")[1:9]),"black")) + scale_y_discrete(labels = rev(tmp$UniProt_Acc))
dev.off()


#do nGene QC plot for paper

BSS = readRDS(file = "BSS_SeuratObj.RDS")

tmp = FetchData(BSS,c("nFeature_RNA","nCount_RNA","orig.ident"))

pdf("./BSS_Histogram_nGene.pdf",width = 5, height = 5)
ggplot(data = tmp, aes(x = nFeature_RNA)) + geom_histogram(binwidth=200) 
ggplot(data = tmp, aes(x = nCount_RNA)) + geom_histogram(binwidth=200) 
ggplot(data = tmp, aes(x = nFeature_RNA,group = orig.ident)) + geom_histogram(binwidth=200) + facet_wrap(.~orig.ident)
ggplot(data = tmp, aes(x = nCount_RNA,group = orig.ident)) + geom_histogram(binwidth=200) + facet_wrap(.~orig.ident)
dev.off()

```





## Integrate cycling BSS1 with amoeba data to find deferentially expressed genes

```{r}
#parallize scaling
library(future)
plan("multiprocess", workers = 8)
plan()

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/")

BSS1 = readRDS(file = "BSS1_Grid_SeuratObj.RDS")

Amoeba_all_anno = readRDS(file = "/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/Amoeba_Data/Amoeba_all_anno_clean_SeuratObj.RDS")
Amoeba_cc = subset(Amoeba_all_anno, idents = c(3))

Amoeba_BSS1 = merge(BSS1, Amoeba_cc, add.cell.ids = c("BSS1","Amoeba"), project = "Amoeba_BSS1_cc")
Amoeba_BSS1@meta.data$origin[Amoeba_BSS1@meta.data$orig.ident %in% c("Amoeba_A","Amoeba_B")] = "Amoeba"
Amoeba_BSS1@meta.data$origin[Amoeba_BSS1@meta.data$orig.ident %in% c("BSS1")] = "BSS1"

Amoeba_BSS1.list <- SplitObject(Amoeba_BSS1, split.by = "origin")
Amoeba_BSS1.list <- Amoeba_BSS1.list[c("Amoeba", "BSS1")]

markers.anno.BSS1 = read.csv("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/BSS1_Grid_allMarker.csv")
markers.anno.Amoeba = read.csv("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/Amoeba_Data/Amoeba_cc_BSSmarker_allMarker.csv")

reference.list <- Amoeba_BSS1.list[c("Amoeba", "BSS1")]
Amoeba_BSS1.anchors <- FindIntegrationAnchors(object.list = reference.list, scale = T, anchor.features = rownames(Amoeba_BSS1), k.filter = 100)

Amoeba_BSS1.integrated <- IntegrateData(anchorset = Amoeba_BSS1.anchors, dims = 1:20)


# switch to integrated assay. The variable features of this assay are automatically
# set during IntegrateData
DefaultAssay(Amoeba_BSS1.integrated) <- "integrated"

#set maximum to 50GB for each worker
options(future.globals.maxSize = 10000 * 1024^2)

# Run the standard workflow for visualization and clustering
Amoeba_BSS1.integrated <- ScaleData(Amoeba_BSS1.integrated, verbose = FALSE)
Amoeba_BSS1.integrated <- RunPCA(Amoeba_BSS1.integrated, npcs = 50,features = rownames(Amoeba_BSS1.integrated), verbose = FALSE)
#Amoeba_BSS1.integrated <- RunPCA(Amoeba_BSS1.integrated, npcs = 10,features = intersect(as.character(markers.anno.BSS1$ID),as.character(markers.anno.Amoeba$ID)), verbose = FALSE)
Amoeba_BSS1.integrated <- RunUMAP(Amoeba_BSS1.integrated, reduction = "pca", dims = 1:10, min.dist = 0.03,seed.use = 24)
Amoeba_BSS1.integrated <- RunTSNE(Amoeba_BSS1.integrated, reduction = "pca", dims = 1:10, seed.use = 24)

#Cluster cells
Amoeba_BSS1.integrated<- FindNeighbors(Amoeba_BSS1.integrated, dims = 1:10)
Amoeba_BSS1.integrated <- FindClusters(Amoeba_BSS1.integrated, resolution = 0.4)



pdf("Amoeba_BSS1_Embedding_PC10_res0.4.pdf",width=10,height=10)
DimPlot(object = Amoeba_BSS1.integrated, reduction = 'umap', pt.size = 2)
DimPlot(object = Amoeba_BSS1.integrated, reduction = 'umap', pt.size = 2,group.by = "origin")
DimPlot(object = Amoeba_BSS1.integrated, reduction = 'tsne', pt.size = 2)
DimPlot(object = Amoeba_BSS1.integrated, reduction = 'tsne', pt.size = 2,group.by = "origin")
dev.off()

#some QC features

pdf("Amoeba_BSS1_feature_QC.pdf",width=5,height=5)
FeaturePlot(Amoeba_BSS1.integrated, reduction = 'umap', pt.size = 0.5, features = c("Phypoly-transcript-08256","Phypoly-transcript-07720","nFeature_RNA","nCount_RNA"),order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]))
FeaturePlot(Amoeba_BSS1.integrated, reduction = 'tsne', pt.size = 0.5, features = c("Phypoly-transcript-08256","Phypoly-transcript-07720","nFeature_RNA","nCount_RNA"),order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]))
dev.off()

saveRDS(Amoeba_BSS1.integrated, file = "Amoeba_BSS1_SeuratObj.RDS")



gene_names = c("TOP2_DICDI",		
"KI67_HUMAN",
"CDC20_DICDI",
"DPOLA_XENLA",
"CDC45_DICDI",	
"MCM4_ARATH",
"MCM7_ARATH",
"BUB1_ARATH",
"RACH_DICDI")

gene_ids = c("Phypoly-transcript-00381",
"Phypoly-transcript-00261",
"Phypoly-transcript-07151",
"Phypoly-transcript-00494",
"Phypoly-transcript-06023",
"Phypoly-transcript-03575",
"Phypoly-transcript-03498",
"Phypoly-transcript-00614",
"Phypoly-transcript-16297")


gg_Fig <- FeaturePlot(Amoeba_BSS1.integrated, reduction = 'umap',pt.size = 1, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_feature_CycleMarker.pdf",width=15,height=12)
CombinePlots( gg_Fig )
dev.off()

gene_names = c("VGA_BPS13","TBAD_PHYPO",
"TBAN_PHYPO",
"UBIQP_AGLNE",
"AGD6_ARATH",
"GTAI_DICDI",
"UBC32_ARATH",
"H2B_COCIM",
"FBXW7_MOUSE",
"CCA11_ORYSJ",
"CCNB_DICDI",
"CDKA1_ORYSJ",
"TBB2_PHYPO",
"ARL6_BOVIN",
"PO21_NASVI",
"FBXW7_HUMAN",
"TBB1_PHYPO",
"AURAA_XENLA","PATS1_DICDI","MUS81_XENTR")
gene_ids = c("Phypoly-transcript-00051","Phypoly-transcript-08562",
"Phypoly-transcript-07720",
"Phypoly-transcript-17966",
"Phypoly-transcript-15759",
"Phypoly-transcript-08000",
"Phypoly-transcript-05787",
"Phypoly-transcript-20418",
"Phypoly-transcript-10476",
"Phypoly-transcript-05693",
"Phypoly-transcript-08256",
"Phypoly-transcript-06666",
"Phypoly-transcript-06678",
"Phypoly-transcript-10238",
"Phypoly-transcript-01160",
"Phypoly-transcript-04948",
"Phypoly-transcript-07525",							
"Phypoly-transcript-07863","Phypoly-transcript-02110","Phypoly-transcript-00664")							
			
gg_Fig <- FeaturePlot(Amoeba_BSS1.integrated, reduction = 'umap',pt.size = 2, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T,min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_feature_LateBSSCycleMarker.pdf",width=27,height=20)
CombinePlots( gg_Fig )
dev.off()
	
#Late Cycle Marker
gene_names = c("TF29_SCHPO","NEG1_NEUCR","IPLA_DICDI","MUS81_XENTR","SMHD1_HUMAN","ANK1_HUMAN","ROCO6_DICDI","PATS1_DICDI","FBN1_BOVIN","AQP2_HORSE","MCM9_XENTR","PMA1_DICDI","TALA1_DICDI","NOTCH_DROME","PHC2_DANRE","LBR_HUMAN","CAP_DICDI","RAD51_CHICK","CALM_SCHPO")

gene_ids = c("Phypoly-transcript-09973","Phypoly-transcript-10029","Phypoly-transcript-00996","Phypoly-transcript-00664","Phypoly-transcript-13710","Phypoly-transcript-00953","Phypoly-transcript-06536","Phypoly-transcript-02110","Phypoly-transcript-00834","Phypoly-transcript-01487","Phypoly-transcript-00963","Phypoly-transcript-01492","Phypoly-transcript-03235","Phypoly-transcript-00580","Phypoly-transcript-15399","Phypoly-transcript-09169","Phypoly-transcript-07637","Phypoly-transcript-11446","Phypoly-transcript-04224")



gg_Fig <- FeaturePlot(Amoeba_BSS1.integrated, reduction = 'umap',pt.size = 2, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_feature_LateCycleMarker.pdf",width=30,height=25)
CombinePlots( gg_Fig )
dev.off()
	

gene_names = c("RAB1A_LYMST",
"SYN2_MOUSE",
"CHMP4_DICDI",
"RABD1_ARATH",
"ISP7_SCHPO",
"TPX_PSEAE",
"MAT2B_BOVIN",
"DIMB_DICDI",
"GCH1_DICDI",
"PATL6_ARATH",
"EB1C_ARATH",
"UBE2C_DICDI",
"AURAA_XENLA",
"KIF15_RAT",
"TRB5_ARATH",
"DEK_MOUSE",
"TBAD_PHYPO",
"RAB1_DIPOM",
"KI67_HUMAN",
"SMC3_ARATH",
"SMC1A_MOUSE",
"RTF1_PONAB",
"H2AV1_ORYSJ",
"DYR_HAEIN",
"UCKB_DICDI",
"HBX10_DICDI",
"CC135_CHLRE",
"PLSC_PHYPO",
"IF2B_WHEAT",
"TNIK_MOUSE",
"FEM1B_CHICK",
"IF5A_SPOFR",
"HST_ARATH",
"NOLC1_HUMAN",
"MYS2_DICDI",
"TCTP_USTMA",
"CCNB_DICDI",
"ACTNA_DICDI",
"CDC20_DICDI")
gene_ids = c("Phypoly-transcript-15390",
"Phypoly-transcript-09696",
"Phypoly-transcript-18237",
"Phypoly-transcript-02031",
"Phypoly-transcript-14030",
"Phypoly-transcript-20022",
"Phypoly-transcript-13909",
"Phypoly-transcript-07397",
"Phypoly-transcript-14216",
"Phypoly-transcript-08023",
"Phypoly-transcript-13407",
"Phypoly-transcript-20628",
"Phypoly-transcript-07863",
"Phypoly-transcript-00408",
"Phypoly-transcript-04676",
"Phypoly-transcript-06998",
"Phypoly-transcript-08562",
"Phypoly-transcript-17410",
"Phypoly-transcript-00261",
"Phypoly-transcript-00797",
"Phypoly-transcript-00833",
"Phypoly-transcript-06858",
"Phypoly-transcript-18931",
"Phypoly-transcript-15374",
"Phypoly-transcript-15271",
"Phypoly-transcript-11361",
"Phypoly-transcript-13229",
"Phypoly-transcript-02353",
"Phypoly-transcript-17742",
"Phypoly-transcript-06398",
"Phypoly-transcript-11935",
"Phypoly-transcript-18652",
"Phypoly-transcript-09131",
"Phypoly-transcript-13375",
"Phypoly-transcript-06241",
"Phypoly-transcript-25465",
"Phypoly-transcript-08256",
"Phypoly-transcript-02032",
"Phypoly-transcript-07151")




gg_Fig <- FeaturePlot(Amoeba_BSS1.integrated, reduction = 'umap',pt.size = 1, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_feature_AmoebaClusterMarker.pdf",width=30,height=20)
CombinePlots( gg_Fig )
dev.off()
	
plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(Amoeba_BSS1.integrated, only.pos = TRUE,  logfc.threshold = 0.2)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(cluster , desc(avg_logFC))

write.csv(markers.anno,"Amoeba_BSS1_IntallMarker.csv")



gene_names = c("PARPT_MOUSE",
"DNLI_HAEIN",
"LVSC_DICDI",
"UBCD1_DROME",
"GTAT_DICDI",
"ELOA_DICDI",
"ELOV5_RAT",
"UCKB_DICDI",
"CH10_ARATH",
"PLSC_PHYPO",
"CHMP4_DICDI",
"RAB1A_LYMST",
"SYN2_MOUSE",
"KITH_DICDI",
"H2AV1_ORYSJ",
"TYW23_ORYSJ"
)
gene_ids = c("Phypoly-transcript-23283",
"Phypoly-transcript-25076",
"Phypoly-transcript-07121",
"Phypoly-transcript-22222",
"Phypoly-transcript-17219",
"Phypoly-transcript-11646",
"Phypoly-transcript-15329",
"Phypoly-transcript-15271",
"Phypoly-transcript-22743",
"Phypoly-transcript-02353",
"Phypoly-transcript-18237",
"Phypoly-transcript-15390",
"Phypoly-transcript-09696",
"Phypoly-transcript-17230",
"Phypoly-transcript-18931",
"Phypoly-transcript-07386")


gg_Fig <- FeaturePlot(Amoeba_BSS1.integrated, reduction = 'umap',pt.size = 1, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]))
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_feature_IntClusterMarker.pdf",width=30,height=20)
CombinePlots( gg_Fig )
dev.off()
	
#check within amoeba for gradient markers and if they agree in plasmodium

markers <- FindMarkers(Amoeba_BSS1,0,1,logfc.threshold = 0.5)
markers.anno = markers
markers.anno$gene = rownames(markers.anno)
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(desc(avg_logFC))

write.csv(markers.anno,"Amoeba_BSS1_C0vsC1Marker.csv")





DefaultAssay(Amoeba_BSS1.integrated) <- "integrated"

gene_names = c("FB293_ARATH",
"RB11B_TOBAC",
"AURAA_XENLA",
"TM86B_RAT",
"PCH2_MOUSE",
"CCNE_HEMPU",
"NCS1_SCHPO",
"CDC20_DICDI",
"CCNB_DICDI",
"SYT7_HUMAN",
"CDKA1_ORYSJ",
"CHX17_ARATH",
"PPSA_STAMF",
"ARL6_BOVIN",
"YD156_PHANO",
"CCA11_ORYSJ",
"VGA_BPS13",
"ESRP1_MOUSE",
"APO1_AGRAE",
"SR3A_PHYPO",
"B3GT6_MOUSE",
"LIMK1_HUMAN",
"SCP1_ASTLP",
"RAB1B_DICDI",
"OXDC_BACSU",
"CHID_BACCI",
"IF2B_WHEAT",
"HAPP_PHYPO",
"FEM1B_CHICK",
"CNN1_HUMAN",
"RS30_PLAF7",
"ANKHM_DROME",
"ROC1_SPIOL",
"LUT5_ARATH",
"TDPZ4_MOUSE",
"PTEN_DICDI")
gene_ids = c("Phypoly-transcript-18441",
"Phypoly-transcript-14205",
"Phypoly-transcript-07863",
"Phypoly-transcript-14708",
"Phypoly-transcript-08988",
"Phypoly-transcript-07696",
"Phypoly-transcript-16350",
"Phypoly-transcript-17752",
"Phypoly-transcript-08256",
"Phypoly-transcript-10933",
"Phypoly-transcript-06666",
"Phypoly-transcript-13802",
"Phypoly-transcript-22262",
"Phypoly-transcript-10238",
"Phypoly-transcript-05766",
"Phypoly-transcript-05693",
"Phypoly-transcript-00051",
"Phypoly-transcript-07316",
"Phypoly-transcript-15552",
"Phypoly-transcript-14650",
"Phypoly-transcript-08259",
"Phypoly-transcript-13062",
"Phypoly-transcript-17763",
"Phypoly-transcript-18448",
"Phypoly-transcript-22656",
"Phypoly-transcript-07972",
"Phypoly-transcript-17742",
"Phypoly-transcript-24525",
"Phypoly-transcript-11935",
"Phypoly-transcript-15300",
"Phypoly-transcript-21996",
"Phypoly-transcript-05594",
"Phypoly-transcript-13492",
"Phypoly-transcript-30414",
"Phypoly-transcript-09740",
"Phypoly-transcript-14363")


gg_Fig <- FeaturePlot(Amoeba_BSS1.integrated, reduction = 'umap',pt.size = 1, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_feature_C0vsC1Marker.pdf",width=30,height=20)
CombinePlots( gg_Fig )
dev.off()
	
#check within amoeba for gradient markers and if they agree in plasmodium

markers <- FindMarkers(Amoeba_BSS1,0,3,logfc.threshold = 0.2)
markers.anno = markers
markers.anno$gene = rownames(markers.anno)
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(desc(avg_logFC))

write.csv(markers.anno,"Amoeba_BSS1_C0vsC3Marker.csv")


DefaultAssay(Amoeba_BSS1.integrated) <- "integrated"

gene_names = c("PROF1_PHYPO",
"MLR_DICDI",
"GST8_CAEEL",
"PHYSA_PHYPO",
"MYS2_DICDI",
"GFPT2_HUMAN",
"CYSP_PEA",
"RAS2_PHYPO",
"LYSG2_DICDI",
"NPD2_PYRAE",
"MRP1_MACFA",
"CTXA_POLPA",
"RGS18_MOUSE",
"FKB12_VICFA",
"SDP1_YEAST",
"COAA_DICDI",
"ADH6_YEAST",
"NAT8L_RAT",
"Y0753_DICDI",
"FORA_DICDI",
"HD12_ENCCU",
"LRRA_DICDI",
"PAKA_DICDI",
"YBFA_BACSU",
"MYPH_ECHGR",
"MGAT3_HUMAN",
"CRNL1_RAT",
"ASPL2_ARATH",
"HBN1_YEAST")
gene_ids = c("Phypoly-transcript-24764",
"Phypoly-transcript-18714",
"Phypoly-transcript-20719",
"Phypoly-transcript-05607",
"Phypoly-transcript-06241",
"Phypoly-transcript-04086",
"Phypoly-transcript-13199",
"Phypoly-transcript-24858",
"Phypoly-transcript-18933",
"Phypoly-transcript-06752",
"Phypoly-transcript-00366",
"Phypoly-transcript-09764",
"Phypoly-transcript-22419",
"Phypoly-transcript-02178",
"Phypoly-transcript-16399",
"Phypoly-transcript-26492",
"Phypoly-transcript-15025",
"Phypoly-transcript-19869",
"Phypoly-transcript-24467",
"Phypoly-transcript-07527",
"Phypoly-transcript-24657",
"Phypoly-transcript-25982",
"Phypoly-transcript-00387",
"Phypoly-transcript-19160",
"Phypoly-transcript-14752",
"Phypoly-transcript-09127",
"Phypoly-transcript-03621",
"Phypoly-transcript-11204",
"Phypoly-transcript-14894")

gg_Fig <- FeaturePlot(Amoeba_BSS1.integrated, reduction = 'umap',pt.size = 1.5, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_feature_AmoebaSpecMarker.pdf",width=27,height=20)
CombinePlots( gg_Fig )
dev.off()



gene_names = c("TBAD_PHYPO",
"TBAN_PHYPO",
"TTL12_MOUSE",
"TBB4_CAEEL",
"TBE_MOUSE",
"TBB2_PHYPO",
"GBPC_DICDI",
"TBG_PHYPA",
"TBB1_PHYPO",
"TTL12_HUMAN",
"TTL_SCHPO",
"TBB4_CAEEL",
"TBE_MOUSE",
"DYL1_DROME",
"DYIN_DICDI",
"DYHC_DICDI",
"DYH1_RAT",
"KIF10_DICDI",
"KIF2_DICDI",
"KIF3B_MOUSE",
"KIF22_DANRE",
"KIF15_RAT",
"K125_TOBAC",
"KIF3A_MOUSE",
"KIF4_DICDI",
"KIF6_DICDI",
"KIFC3_MOUSE",
"KIF17_MOUSE",
"KIF5_DICDI",
"KIF11_HUMAN",
"KINH_BOTFU",
"KIF1_DICDI",
"KRP95_STRPU",
"PRC1_MOUSE",
"PRC1_HUMAN")
gene_ids = c("Phypoly-transcript-08562",
"Phypoly-transcript-07720",
"Phypoly-transcript-06761",
"Phypoly-transcript-04535",
"Phypoly-transcript-07705",
"Phypoly-transcript-06678",
"Phypoly-transcript-00072",
"Phypoly-transcript-08559",
"Phypoly-transcript-07525",
"Phypoly-transcript-01187",
"Phypoly-transcript-10161",
"Phypoly-transcript-05022",
"Phypoly-transcript-07705",
"Phypoly-transcript-26692",
"Phypoly-transcript-04010",
"Phypoly-transcript-12812",
"Phypoly-transcript-26083",
"Phypoly-transcript-02453",
"Phypoly-transcript-00675",
"Phypoly-transcript-02804",
"Phypoly-transcript-03486",
"Phypoly-transcript-00408",
"Phypoly-transcript-00639",
"Phypoly-transcript-04460",
"Phypoly-transcript-00195",
"Phypoly-transcript-02836",
"Phypoly-transcript-03402",
"Phypoly-transcript-04503",
"Phypoly-transcript-01528",
"Phypoly-transcript-02585",
"Phypoly-transcript-01593",
"Phypoly-transcript-00221",
"Phypoly-transcript-01362",
"Phypoly-transcript-01429",
"Phypoly-transcript-04973")


gg_Fig <- FeaturePlot(Amoeba_BSS1.integrated, reduction = 'umap',pt.size = 1, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_feature_SpindleMarker.pdf",width=22,height=20)
CombinePlots( gg_Fig )
dev.off()
	



#Boxplot for paper

gene_names = c("TBAD_PHYPO",
"TBAN_PHYPO",
"TBB1_PHYPO",
"TBB2_PHYPO",
"TBB4_CAEEL",
"TBE_MOUSE",
"TBG_PHYPA",
"FORA_DICDI",
"HD12_ENCCU",
"PAKA_DICDI",
"KIF1_DICDI")

	
gene_ids = c("Phypoly-transcript-08562",
"Phypoly-transcript-07720",
"Phypoly-transcript-07525",	
"Phypoly-transcript-06678",	
"Phypoly-transcript-04535",
"Phypoly-transcript-07705",
"Phypoly-transcript-08559",
"Phypoly-transcript-07527",
"Phypoly-transcript-24657",
"Phypoly-transcript-00387",	
"Phypoly-transcript-00221")


gg_Fig <- VlnPlot(Amoeba_BSS1_int, gene_ids,group.by = "origin", pt.size  = -1, slot = "counts", same.y.lims = F) 
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + geom_boxplot(width = 0.2,fill = "white") + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_HarmonyInt_BoxPlot_PaperMarker.pdf",width=15,height=15)
CombinePlots( gg_Fig )
dev.off()
	
#check for viral marker

markers <- FindMarkers(Amoeba_BSS1_int,3,c(0,1,2,4),logfc.threshold = 0.2)
markers.anno = markers
markers.anno$gene = rownames(markers.anno)
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(desc(avg_logFC))

write.csv(markers.anno,"Amoeba_BSS1_BSSvsAmoebaMarker.csv")

gene_ids = c("Phypoly-transcript-00051",
"Phypoly-transcript-23963",
"Phypoly-transcript-31473",
"Phypoly-transcript-30710",
"Phypoly-transcript-22262")

gene_names = c("VGA_BPS13",
"HEXO1_ARATH",
"TRM6_DICDI",
"ANK3_HUMAN",
"PPSA_STAMF")

gg_Fig <- FeaturePlot(Amoeba_BSS1_int, reduction = 'umap',pt.size = 2, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_HarmonyInt_feature_BSSvsAmoebaMarker.pdf",width=10,height=15)
CombinePlots( gg_Fig )
dev.off()
	
#
markers <- FindMarkers(Amoeba_BSS1_int,3,c(1),logfc.threshold = 0.2)
markers.anno = markers
markers.anno$gene = rownames(markers.anno)
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(desc(avg_log2FC))

write.csv(markers.anno,"Amoeba_BSS1_BSSvsAmoebaC1Marker.csv")


markers <- FindMarkers(Amoeba_BSS1_int,3,c(1),test.use = "bimod", logfc.threshold = 0.2)
markers.anno = markers
markers.anno$gene = rownames(markers.anno)
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno  %>% arrange(desc(avg_log2FC))

write.csv(markers.anno,"Amoeba_BSS1_BSSvsAmoebaC1Marker_bimod.csv")


gene_ids = c("Phypoly-transcript-25457",
"Phypoly-transcript-24112",
"Phypoly-transcript-13229",
"Phypoly-transcript-01260",
"Phypoly-transcript-02353",
"Phypoly-transcript-01260",
"Phypoly-transcript-25457",
"Phypoly-transcript-20231",
"Phypoly-transcript-31663",
"Phypoly-transcript-23282")
gene_names = c("RL29_DICDI",			
"RS22A_YEAST",			
"CC135_CHLRE",			
"RL3_ASPFU",			
"PLSC_PHYPO",
"RL3_ASPFU",
"RL29_DICDI",
"RL371_DROME",
"SH3L2_BOVIN",
"YGU5_SCHPO")

gg_Fig <- FeaturePlot(Amoeba_BSS1_int, reduction = 'umap',pt.size = 2, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T)
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_HarmonyInt_feature_BSSvsAmoebaC1Marker.pdf",width=10,height=15)
CombinePlots( gg_Fig )
dev.off()
	

amoeba_spec_genes = setdiff(rownames(Amoeba_all_anno),rownames(BSS1))

#get genes that are detected in the Seurat object
transcripts2plot = vector()

for (i in amoeba_spec_genes) {
  print(i)
  a = sum(GetAssayData(object = Amoeba_BSS1_int, slot = "data")[i,]>0)
  if (a > 0) {
    transcripts2plot = c(transcripts2plot,i)
  }
}

transcripts2plot = intersect(transcripts2plot,anno.gene$transcript)

gene_names = anno.gene$UniProt_Acc[anno.gene$transcript %in% transcripts2plot]


gg_Fig <- FeaturePlot(Amoeba_BSS1_int, reduction = 'umap',pt.size = 0.8, features = transcripts2plot,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T)
gg_Fig <- lapply( 1:length(gene_names), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_HarmonyInt_feature_AmoebaSpecMarker.pdf",width=22,height=20)
CombinePlots( gg_Fig )
dev.off()


gg_Fig <- FeaturePlot(Amoeba_all_anno, reduction = 'umap',pt.size = 0.8, features = transcripts2plot,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T)
gg_Fig <- lapply( 1:length(gene_names), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_feature_BSS1vsAmoebaSpecMarker.pdf",width=22,height=20)
CombinePlots( gg_Fig )
dev.off()


#check all known retrotransposon genes in Physarum

gene_ids = unique(anno$transcript[grep("retrov",anno$UniProt_Name)])
gene_names = anno.gene$UniProt_Acc[anno.gene$transcript %in% gene_ids]
gene_ids = anno.gene$transcript[anno.gene$UniProt_Acc %in% gene_names]

gg_Fig <- FeaturePlot(Amoeba_BSS1_int, reduction = 'umap',pt.size = 2, features = gene_ids,order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T, min.cutoff = "q10", max.cutoff = "q90")
gg_Fig <- lapply( 1:length(gene_ids), function(x) { gg_Fig[[x]] + labs(title=gene_names[x])  })

pdf("Amoeba_BSS1_HarmonyInt_feature_RetrotransposonMarker.pdf",width=10,height=15)
CombinePlots( gg_Fig )
dev.off()
	
```

##Correlate nuclei data to BSS1

```{r}

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/")


### for all nuclei

#load young slime mold

cycleNuc = readRDS("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/Nuc_Data/Nuc_young_anno_SeuratObj.RDS")

norm.data.Nuc = GetAssayData(object = cycleNuc, slot = "data")
norm.data.Nuc = as.matrix(norm.data.Nuc)
norm.data.Nuc = as.data.frame(t(norm.data.Nuc),stringsAsFactors=F)

meta.data.Nuc = FetchData(object = cycleNuc, vars = c('ident',"nFeature_RNA"))
meta.data.Nuc$ident = paste("Nuc",meta.data.Nuc$ident,sep = "_")

norm.data.SM1 = GetAssayData(object = BSS1, slot = "data")
norm.data.SM1 = as.matrix(norm.data.SM1)
norm.data.SM1 = as.data.frame(t(norm.data.SM1),stringsAsFactors=F)

meta.data.SM1 = FetchData(object = BSS1, vars = c('ident',"nFeature_RNA"))
meta.data.SM1$ident = paste("SM1",meta.data.SM1$ident,sep = "_")

data.Nuc = cbind(meta.data.Nuc[,c("nFeature_RNA","ident")], norm.data.Nuc, stringsAsFactors = F)
data.SM1 = cbind(meta.data.SM1[,c("nFeature_RNA","ident")], norm.data.SM1, stringsAsFactors = F)

data.Nuc = data.Nuc[,colnames(data.Nuc) %in% intersect(colnames(data.Nuc),colnames(data.SM1))]
data.SM1 = data.SM1[,colnames(data.SM1) %in% intersect(colnames(data.Nuc),colnames(data.SM1))]



data = rbind(data.Nuc, data.SM1, stringsAsFactors = F)

data.cor = data[,2:ncol(data)]

data.means <-as.data.frame(do.call("rbind", as.list(by(data.cor[,2:ncol(data.cor)], as.factor(data.cor$ident), colMeans))))

data.means<-as.data.frame(t(data.means),stringsAsFactors = F)

data.means.combined <- cbind(data.means, as.data.frame( t (data[,3:ncol(data)]) ,stringsAsFactors = F), stringsAsFactors=F)

saveRDS(data.means.combined, "Correlation_BSS1_YoungNucData_pseudobulk.RDS")
#data.means.combined = readRDS("Correlation_Axolotl_pseudobulk.RDS")

### only for cycling nuclei

#load young slime mold

cycleNuc = readRDS("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/Nuc_Data/Nuc_young_cc_anno_SeuratObj.RDS")

norm.data.Nuc = GetAssayData(object = cycleNuc, slot = "data")
norm.data.Nuc = as.matrix(norm.data.Nuc)
norm.data.Nuc = as.data.frame(t(norm.data.Nuc),stringsAsFactors=F)

meta.data.Nuc = FetchData(object = cycleNuc, vars = c('ident',"nFeature_RNA"))
meta.data.Nuc$ident = paste("Nuc",meta.data.Nuc$ident,sep = "_")

norm.data.SM1 = GetAssayData(object = BSS1, slot = "data")
norm.data.SM1 = as.matrix(norm.data.SM1)
norm.data.SM1 = as.data.frame(t(norm.data.SM1),stringsAsFactors=F)

meta.data.SM1 = FetchData(object = BSS1, vars = c('ident',"nFeature_RNA"))
meta.data.SM1$ident = paste("SM1",meta.data.SM1$ident,sep = "_")

data.Nuc = cbind(meta.data.Nuc[,c("nFeature_RNA","ident")], norm.data.Nuc, stringsAsFactors = F)
data.SM1 = cbind(meta.data.SM1[,c("nFeature_RNA","ident")], norm.data.SM1, stringsAsFactors = F)

data.Nuc = data.Nuc[,colnames(data.Nuc) %in% intersect(colnames(data.Nuc),colnames(data.SM1))]
data.SM1 = data.SM1[,colnames(data.SM1) %in% intersect(colnames(data.Nuc),colnames(data.SM1))]



data = rbind(data.Nuc, data.SM1, stringsAsFactors = F)

data.cor = data[,2:ncol(data)]

data.means <-as.data.frame(do.call("rbind", as.list(by(data.cor[,2:ncol(data.cor)], as.factor(data.cor$ident), colMeans))))

data.means<-as.data.frame(t(data.means),stringsAsFactors = F)

data.means.combined <- cbind(data.means, as.data.frame( t (data[,3:ncol(data)]) ,stringsAsFactors = F), stringsAsFactors=F)

saveRDS(data.means.combined, "Correlation_BSS1_YoungNucData_CConly_pseudobulk.RDS")
#data.means.combined = readRDS("Correlation_Axolotl_pseudobulk.RDS")

```

Including all nuclei

```{r}
#######BSS1 onto Nuc

rm(cor)
rm(cor.sc.bulk)

cor.sc.bulk<-data.frame()

for(i in 1:4) {
        print(i)
        cor<-cor(data.means.combined[,i], data.means.combined[,5300:5492], method = c("spearman"))
        cor.sc.bulk<-rbind(cor.sc.bulk,cor)
}

rownames(cor.sc.bulk)<-colnames(data.means.combined)[1:4]

saveRDS(cor.sc.bulk,"COR_matrix_BSS1_ON_YoungNucData.RDS")

cor.sc.bulk = readRDS("COR_matrix_BSS1_ON_YoungNucData.RDS")
library(gplots)
library(RColorBrewer)

col.regions=colorRampPalette(c(rep("black",1),brewer.pal(9,"Blues")[9:1]), space="Lab")

palette.breaks <- seq(min(cor.sc.bulk),max(cor.sc.bulk), 0.1)

hc <- hclust(as.dist(1-cor(cor.sc.bulk,method="pearson")), method="ward")
#hr <- hclust(as.dist(1-cor(t(cor.sc.bulk),method="pearson")), method="ward")

pdf("COR_heatmap_BSS1_ON_YoungNucData.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk),trace="none",density="none",col=col.regions,Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE, key=T,scale="col",ColSideColors=c("#A50026","#EFE9C3" , "#ABD9E9" ,"#5C91C2" , "#E54E35","#313695","#FDAE61" ,"dodgerblue","palegreen1","palegreen2","palegreen3","palegreen4")[as.factor(data[colnames(cor.sc.bulk),"ident"])])
dev.off()

#######Nuc onto BSS1

rm(cor)
rm(cor.sc.bulk)

cor.sc.bulk<-data.frame()

for(i in 5:7) {
        print(i)
        cor<-cor(data.means.combined[,i], data.means.combined[,8:5299], method = c("spearman"))
        cor.sc.bulk<-rbind(cor.sc.bulk,cor)
}

rownames(cor.sc.bulk)<-colnames(data.means.combined)[1:4]

saveRDS(cor.sc.bulk,"COR_matrix_YoungNucData_ON_BSS1.RDS")

cor.sc.bulk = readRDS("COR_matrix_YoungNucData_ON_BSS1.RDS")
library(gplots)
library(RColorBrewer)

col.regions=colorRampPalette(c(rep("black",1),brewer.pal(9,"Blues")[9:1]), space="Lab")

palette.breaks <- seq(min(cor.sc.bulk),max(cor.sc.bulk), 0.1)

hc <- hclust(as.dist(1-cor(cor.sc.bulk,method="pearson")), method="ward")
hr <- hclust(as.dist(1-cor(t(cor.sc.bulk),method="pearson")), method="ward")

pdf("COR_heatmap_YoungNucData_ON_BSS1.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk),trace="none",density="none",col=col.regions,Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE, key=T,scale="col",ColSideColors=c("#A50026","#EFE9C3" , "#ABD9E9" ,"#5C91C2" , "#E54E35","#313695","#FDAE61" ,"dodgerblue","palegreen1","palegreen2","palegreen3","palegreen4")[as.factor(data[colnames(cor.sc.bulk),"ident"])])
dev.off()

```

Only CC nuclei

```{r}
#######BSS1 onto Nuc

rm(cor)
rm(cor.sc.bulk)

cor.sc.bulk<-data.frame()

for(i in 1:6) {
        print(i)
        cor<-cor(data.means.combined[,i], data.means.combined[,2528:2720], method = c("spearman"))
        cor.sc.bulk<-rbind(cor.sc.bulk,cor)
}

rownames(cor.sc.bulk)<-colnames(data.means.combined)[1:6]

saveRDS(cor.sc.bulk,"COR_matrix_BSS1_ON_YoungNucData_CConly.RDS")

cor.sc.bulk = readRDS("COR_matrix_BSS1_ON_YoungNucData_CConly.RDS")
library(gplots)
library(RColorBrewer)

col.regions=colorRampPalette(c(rep("black",1),brewer.pal(9,"Blues")[9:1]), space="Lab")

palette.breaks <- seq(min(cor.sc.bulk),max(cor.sc.bulk), 0.1)

hc <- hclust(as.dist(1-cor(cor.sc.bulk,method="pearson")), method="ward")
#hr <- hclust(as.dist(1-cor(t(cor.sc.bulk),method="pearson")), method="ward")

pdf("COR_heatmap_BSS1_ON_YoungNucData_CConly.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk),trace="none",density="none",col=col.regions,Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE, key=T,scale="col",ColSideColors=c("#A50026","#EFE9C3" , "#ABD9E9" ,"#5C91C2" , "#E54E35","#313695","#FDAE61" ,"dodgerblue","palegreen1","palegreen2","palegreen3","palegreen4")[as.factor(data[colnames(cor.sc.bulk),"ident"])])
dev.off()

#######Nuc onto BSS1

rm(cor)
rm(cor.sc.bulk)

cor.sc.bulk<-data.frame()

for(i in 7:9) {
        print(i)
        cor<-cor(data.means.combined[,i], data.means.combined[,10:2527], method = c("spearman"))
        cor.sc.bulk<-rbind(cor.sc.bulk,cor)
}

rownames(cor.sc.bulk)<-colnames(data.means.combined)[7:9]

saveRDS(cor.sc.bulk,"COR_matrix_YoungNucData_CConly_ON_BSS1.RDS")

cor.sc.bulk = readRDS("COR_matrix_YoungNucData_CConly_ON_BSS1.RDS")
library(gplots)
library(RColorBrewer)

col.regions=colorRampPalette(c(rep("black",1),brewer.pal(9,"Blues")[9:1]), space="Lab")

palette.breaks <- seq(min(cor.sc.bulk),max(cor.sc.bulk), 0.1)

hc <- hclust(as.dist(1-cor(cor.sc.bulk,method="pearson")), method="ward")
hr <- hclust(as.dist(1-cor(t(cor.sc.bulk),method="pearson")), method="ward")

pdf("COR_heatmap_YoungNucData_CConly_ON_BSS1.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk),trace="none",density="none",col=col.regions,Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE, key=T,scale="col",ColSideColors=c("#A50026","#EFE9C3" , "#ABD9E9" ,"#5C91C2" , "#E54E35","#313695","#FDAE61" ,"dodgerblue","palegreen1","palegreen2","palegreen3","palegreen4")[as.factor(data[colnames(cor.sc.bulk),"ident"])])
dev.off()

```


##Correlate nuclei data to BSS4

```{r}

setwd("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/")


### for all nuclei

#load old slime mold

cycleNuc = readRDS("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/Nuc_Data/Nuc_old_anno_SeuratObj.RDS")

norm.data.Nuc = GetAssayData(object = cycleNuc, slot = "data")
norm.data.Nuc = as.matrix(norm.data.Nuc)
norm.data.Nuc = as.data.frame(t(norm.data.Nuc),stringsAsFactors=F)

meta.data.Nuc = FetchData(object = cycleNuc, vars = c('ident',"nFeature_RNA"))
meta.data.Nuc$ident = paste("Nuc",meta.data.Nuc$ident,sep = "_")

#get nuc var genes

#parallize scaling
library(future)
plan("multiprocess", workers = 8)
plan()

#set maximum to 10GB for each worker
options(future.globals.maxSize = 10000 * 1024^2)

plan("multiprocess", workers = 8)
library(tidyverse)

markers <- FindAllMarkers(cycleNuc, only.pos = TRUE,  logfc.threshold = 0.1)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno %>%
    group_by(cluster) %>%
    arrange(desc(avg_logFC)) %>%
    slice(1:12)


BSS4 = readRDS("/home/tobias_gerber/Projects/SingleCellGroup/SlimeMold/SpatialTranscriptomics/BSS4_Grid_SeuratObj.RDS")

norm.data.SM4 = GetAssayData(object = BSS4, slot = "data")
norm.data.SM4 = as.matrix(norm.data.SM4)
norm.data.SM4 = as.data.frame(t(norm.data.SM4),stringsAsFactors=F)

meta.data.SM4 = FetchData(object = BSS4, vars = c('ident',"nFeature_RNA"))
meta.data.SM4$ident = paste("SM4",meta.data.SM4$ident,sep = "_")


markers <- FindAllMarkers(BSS4, only.pos = TRUE,  logfc.threshold = 0.1)
markers.anno = markers
markers.anno = merge(markers.anno,anno.gene, by.x="gene" , by.y="transcript")
markers.anno$ID = markers.anno$gene

markers.anno = markers.anno %>%
    group_by(cluster) %>%
    arrange(desc(avg_logFC)) %>%
    slice(1:50)



data.Nuc = cbind(meta.data.Nuc[,c("nFeature_RNA","ident")], norm.data.Nuc, stringsAsFactors = F)
data.SM4 = cbind(meta.data.SM4[,c("nFeature_RNA","ident")], norm.data.SM4, stringsAsFactors = F)

data.Nuc = data.Nuc[,colnames(data.Nuc) %in% intersect(colnames(data.Nuc),colnames(data.SM4))]
data.SM4 = data.SM4[,colnames(data.SM4) %in% intersect(colnames(data.Nuc),colnames(data.SM4))]



data = rbind(data.Nuc, data.SM4, stringsAsFactors = F)

data.cor = data[,2:ncol(data)]

data.means <-as.data.frame(do.call("rbind", as.list(by(data.cor[,2:ncol(data.cor)], as.factor(data.cor$ident), colMeans))))

data.means<-as.data.frame(t(data.means),stringsAsFactors = F)

data.means.combined <- cbind(data.means, as.data.frame( t (data[,3:ncol(data)]) ,stringsAsFactors = F), stringsAsFactors=F)

saveRDS(data.means.combined, "Correlation_BSS4_OldNucData_pseudobulk.RDS")
#data.means.combined = readRDS("Correlation_Axolotl_pseudobulk.RDS")

```

Including all nuclei

```{r}
#######BSS4 onto Nuc

data.means.combined = readRDS("Correlation_BSS4_OldNucData_pseudobulk.RDS")

rm(cor)
rm(cor.sc.bulk)

cor.sc.bulk<-data.frame()

#use only nuclei cluster marker for correlation of grids to nuclei cluster bulks

for(i in 1:6) {
        print(i)
        cor<-cor(data.means.combined[markers.anno$gene,i], data.means.combined[markers.anno$gene,10340:10443], method = c("spearman"))
        cor.sc.bulk<-rbind(cor.sc.bulk,cor)
}

rownames(cor.sc.bulk)<-colnames(data.means.combined)[1:6]

saveRDS(cor.sc.bulk,"COR_matrix_BSS4_ON_OldNucData_VarGenes.RDS")

cor.sc.bulk = readRDS("COR_matrix_BSS4_ON_OldNucData_VarGenes.RDS")
library(gplots)
library(RColorBrewer)

col.regions=colorRampPalette(c(rep("black",1),brewer.pal(9,"Greys")[9:1],"white",brewer.pal(9,"Reds")[1:9]), space="Lab")

palette.breaks <- seq(min(cor.sc.bulk),max(cor.sc.bulk), 0.025)

hc <- hclust(as.dist(1-cor(cor.sc.bulk,method="pearson")), method="ward")
#hr <- hclust(as.dist(1-cor(t(cor.sc.bulk),method="pearson")), method="ward")

pdf("COR_heatmap_BSS4_ON_OldNucData_VarGenes.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk),trace="none",density="none",col=col.regions(100),Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE,key=T,scale="col",ColSideColors=c("orange","goldenrod1","red2")[as.factor(data[colnames(cor.sc.bulk),"ident"])])
dev.off()

col.regions=colorRampPalette(c("grey",brewer.pal(9,"Blues")[1:9]), space="Lab")

pdf("COR_heatmap_BSS4_ON_OldNucData_VarGenes_noScale.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk), trace="none",density="none",col=col.regions,Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE, key=T,scale="none",ColSideColors=c("orange","goldenrod1","red2")[as.factor(data[colnames(cor.sc.bulk),"ident"])])
dev.off()


#Do manual scaling over columns and then over rows.

cor.sc.bulk.scale = cor.sc.bulk
#data.plot[,-1:-2][data.plot[,-1:-2]==0]<-NA
cor.sc.bulk.scale<-scale(cor.sc.bulk.scale)
#data.plot[data.plot>3]<-3
#data.plot[data.plot<(-3)]<-(-3)
#data.plot[is.na(data.plot)]<-(-3)
#cor.sc.bulk.scale = as.data.frame(cor.sc.bulk.scale)
cor.sc.bulk.scale = t(scale(t(cor.sc.bulk.scale)))
cor.sc.bulk.scale[cor.sc.bulk.scale>1]<-1
cor.sc.bulk.scale[cor.sc.bulk.scale<(-1)]<-(-1)

col.regions=colorRampPalette(c(rep("black",1),brewer.pal(9,"Greys")[9:1],"white",brewer.pal(9,"Reds")[1:9]), space="Lab")

palette.breaks <- seq(min(cor.sc.bulk.scale),max(cor.sc.bulk.scale), 0.1)

hc <- hclust(as.dist(1-cor(cor.sc.bulk.scale,method="pearson")), method="ward")
hr <- hclust(as.dist(1-cor(t(cor.sc.bulk.scale),method="pearson")), method="ward")

pdf("COR_heatmap_BSS4_ON_OldNucData_VarGenes_scale2.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk),trace="none",density="none",col=col.regions(100),Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE,key=T,scale="none",ColSideColors=c("orange","goldenrod1","red2")[as.factor(data[colnames(cor.sc.bulk),"ident"])])
dev.off()




#######Nuc onto BSS4

rm(cor)
rm(cor.sc.bulk)

cor.sc.bulk<-data.frame()

#use only grid cluster marker for correlation of nuclei to grid cluster bulks

for(i in 7:9) {
        print(i)
        cor<-cor(data.means.combined[intersect(rownames(data.means.combined),markers.anno$gene),i], data.means.combined[intersect(rownames(data.means.combined),markers.anno$gene),10:10339], method = c("spearman"))
        cor.sc.bulk<-rbind(cor.sc.bulk,cor)
}

rownames(cor.sc.bulk)<-colnames(data.means.combined)[7:9]

saveRDS(cor.sc.bulk,"COR_matrix_OldNucData_ON_BSS4_varGenes.RDS")

cor.sc.bulk = readRDS("COR_matrix_OldNucData_ON_BSS4_varGenes.RDS")
library(gplots)
library(RColorBrewer)

cor.sc.bulk[is.na(cor.sc.bulk)] = -2

col.regions=colorRampPalette(c(rep("black",1),brewer.pal(9,"Greys")[9:1],"white",brewer.pal(9,"Reds")[1:9]), space="Lab")


palette.breaks <- seq(min(cor.sc.bulk),max(cor.sc.bulk), 0.1)

hc <- hclust(as.dist(1-cor(cor.sc.bulk,method="pearson")), method="ward")
hr <- hclust(as.dist(1-cor(t(cor.sc.bulk),method="pearson")), method="ward")

pdf("COR_heatmap_OldNucData_ON_BSS4_varGenes.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk),trace="none",density="none",col=col.regions,Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE, key=T,scale="col",ColSideColors=c('#88CCEE', '#44AA99', '#999933','#CC6677', '#882255', '#117733', '#332288', '#DDCC77', '#AA4499', '#DDDDDD')[as.factor(data[colnames(cor.sc.bulk),"ident"])])
dev.off()

#Do manual scaling over columns and then over rows.

cor.sc.bulk.scale = cor.sc.bulk
#data.plot[,-1:-2][data.plot[,-1:-2]==0]<-NA
cor.sc.bulk.scale<-scale(cor.sc.bulk.scale)
#cor.sc.bulk.scale[cor.sc.bulk.scale>3]<-3
#cor.sc.bulk.scale[cor.sc.bulk.scale<(-3)]<-(-3)
#data.plot[is.na(data.plot)]<-(-3)
#cor.sc.bulk.scale = as.data.frame(cor.sc.bulk.scale)
cor.sc.bulk.scale = t(scale(t(cor.sc.bulk.scale)))
cor.sc.bulk.scale[cor.sc.bulk.scale>1]<-1
cor.sc.bulk.scale[cor.sc.bulk.scale<(-1)]<-(-1)


col.regions=colorRampPalette(c(rep("black",1),brewer.pal(9,"Greys")[9:1],"white",brewer.pal(9,"Reds")[1:9]), space="Lab")

palette.breaks <- seq(min(cor.sc.bulk.scale),max(cor.sc.bulk.scale), 0.1)

hc <- hclust(as.dist(1-cor(cor.sc.bulk.scale,method="pearson")), method="ward")
hr <- hclust(as.dist(1-cor(t(cor.sc.bulk.scale),method="pearson")), method="ward")

pdf("COR_heatmap_OldNucData_ON_BSS4_varGenes_scale2.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk.scale),trace="none",density="none",col=col.regions(100),Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE, key=T,ColSideColors=c('#88CCEE', '#44AA99', '#999933','#CC6677', '#882255', '#117733', '#332288', '#DDCC77', '#AA4499', '#DDDDDD')[as.factor(data[colnames(cor.sc.bulk.scale),"ident"])])
dev.off()


pdf("COR_heatmap_OldNucData_ON_BSS4_varGenes_scale2.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk.scale),trace="none",density="none",col=col.regions(100),Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE, key=T,scale="col",ColSideColors=c('#88CCEE', '#44AA99', '#999933','#CC6677', '#882255', '#117733', '#332288', '#DDCC77', '#AA4499', '#DDDDDD')[as.factor(data[colnames(cor.sc.bulk.scale),"ident"])])
dev.off()





####correlate bulk samples with each other

tmp = data.means.combined[markers.anno$gene, 7:9]
tmp = tmp[complete.cases(tmp),]

rm(cor)
rm(cor.sc.bulk)

cor.sc.bulk<-data.frame()

for(i in 7:9) {
        print(i)
        cor<-cor(data.means.combined[rownames(tmp),i], data.means.combined[rownames(tmp),1:6], method = c("spearman"))
        cor.sc.bulk<-rbind(cor.sc.bulk,cor)
}

rownames(cor.sc.bulk)<-colnames(data.means.combined)[7:9]

saveRDS(cor.sc.bulk,"COR_matrix_OldNucData_ON_BSS4_bulkCor_varGenes.RDS")

cor.sc.bulk = readRDS("COR_matrix_OldNucData_ON_BSS4_bulkCor.RDS")
library(gplots)
library(RColorBrewer)


col.regions=colorRampPalette(c(rep("black",1),brewer.pal(9,"Greys")[9:1],"white",brewer.pal(9,"Reds")[1:9]), space="Lab")

palette.breaks <- seq(min(cor.sc.bulk),max(cor.sc.bulk), 0.1)

hc <- hclust(as.dist(1-cor(cor.sc.bulk,method="pearson")), method="ward")
hr <- hclust(as.dist(1-cor(t(cor.sc.bulk),method="pearson")), method="ward")

pdf("COR_heatmap_OldNucData_ON_BSS4_bulkCor_varGenes.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk),trace="none",density="none",col=col.regions(100),Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE,key=T,scale="col",ColSideColors=c('#88CCEE', '#44AA99', '#999933','#CC6677', '#882255', '#117733', '#332288', '#DDCC77', '#AA4499', '#DDDDDD')[1:ncol(cor.sc.bulk)])
dev.off()

col.regions=colorRampPalette(c("grey80",brewer.pal(9,"Blues")[3:9]), space="Lab")

pdf("COR_heatmap_OldNucData_ON_BSS4_bulkCor_noScale_varGenes.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk), trace="none",density="none",col=col.regions(100),Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE, key=T,scale="none",ColSideColors=c('#88CCEE', '#44AA99', '#999933','#CC6677', '#882255', '#117733', '#332288', '#DDCC77', '#AA4499', '#DDDDDD')[1:ncol(cor.sc.bulk)])
dev.off()


#Do manual scaling over columns and then over rows.

cor.sc.bulk.scale = cor.sc.bulk[,1:5]
#data.plot[,-1:-2][data.plot[,-1:-2]==0]<-NA
cor.sc.bulk.scale<-scale(cor.sc.bulk.scale)
#data.plot[data.plot>3]<-3
#data.plot[data.plot<(-3)]<-(-3)
#data.plot[is.na(data.plot)]<-(-3)
#cor.sc.bulk.scale = as.data.frame(cor.sc.bulk.scale)
cor.sc.bulk.scale = t(scale(t(cor.sc.bulk.scale)))
cor.sc.bulk.scale[cor.sc.bulk.scale>1]<-1
cor.sc.bulk.scale[cor.sc.bulk.scale<(-1)]<-(-1)

col.regions=colorRampPalette(c(rep("black",1),brewer.pal(9,"Greys")[9:1],"white",brewer.pal(9,"Reds")[1:9]), space="Lab")

palette.breaks <- seq(min(cor.sc.bulk.scale),max(cor.sc.bulk.scale), 0.1)

hc <- hclust(as.dist(1-cor(cor.sc.bulk.scale,method="pearson")), method="ward")
hr <- hclust(as.dist(1-cor(t(cor.sc.bulk.scale),method="pearson")), method="ward")

pdf("COR_heatmap_OldNucData_ON_BSS4_bulkCor_scale2_varGenes.pdf",width=15,height=10)
heatmap.celltypes<-heatmap.2(as.matrix(cor.sc.bulk.scale),trace="none",density="none",col=col.regions(100),Colv=as.dendrogram(hc), dendrogram="col",Rowv=FALSE,key=T,scale="none",ColSideColors=c('#88CCEE', '#44AA99', '#999933','#CC6677', '#882255', '#117733', '#332288', '#DDCC77', '#AA4499', '#DDDDDD')[1:ncol(cor.sc.bulk.scale)])
dev.off()

################ now select the top fan cluster (4) and the top network cluster (0) and correlate mock bulk to each individual grid spot


cor.sc.bulk<-data.frame()

for(i in 10340:10443) {
        print(i)
        cor<-cor(data.means.combined[rownames(tmp),i], data.means.combined[rownames(tmp),1:6], method = c("spearman"))
        cor.sc.bulk<-rbind(cor.sc.bulk,cor)
}

rownames(cor.sc.bulk)<-colnames(data.means.combined)[10340:10443]

BSS4@meta.data$NucC4score = cor.sc.bulk$Nuc_4
BSS4@meta.data$NucC0score = cor.sc.bulk$Nuc_0

pdf("BSS4_feature_NucleiScores.pdf",width=8,height=5)
FeaturePlot(BSS4, reduction = 'spatial', pt.size = 3,shape.by = "orig.ident", features = c("NucC4score","NucC0score"),order = T, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9])) + scale_shape_manual(values = c("BSS4" = 15))
dev.off()


gg_Fig <- FeaturePlot(BSS4, reduction = 'spatial', shape.by = "orig.ident" ,pt.size = 5, features = c("NucC4score","NucC0score"),order = F, cols = c(brewer.pal(9,"Greys")[3],brewer.pal(9,"Blues")[3:9]), repel=T)
gg_Fig <- lapply( 1:length(c("NucC4score","NucC0score")), function(x) { gg_Fig[[x]] + scale_shape_manual(values = c("BSS4" = 15)) })

pdf("BSS4_Grid_feature_NucleiScore.pdf",width=9,height=4)
CombinePlots( gg_Fig )
dev.off()
	
```
